{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BgI3oJZpteRP"
      },
      "source": [
        "# Editing analysis - Quick Start"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6BLjt3TYteRR"
      },
      "source": [
        "## #Denovo Call Editing - Prepare"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8KB52yp-teRR"
      },
      "source": [
        "### Imports"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "OeXh4DxkteRS"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import subprocess\n",
        "import os\n",
        "import glob\n",
        "import random\n",
        "import pybedtools\n",
        "import pysam\n",
        "import numpy as np\n",
        "from pathlib import Path\n",
        "from concurrent.futures import ProcessPoolExecutor as Pool\n",
        "from tqdm import tqdm\n",
        "from math import ceil\n",
        "import matplotlib.pyplot as plt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jrrXDjbhteRS"
      },
      "source": [
        "### Data\n",
        "##### 1. ADAR_KO  1sample  paired-end HEK293 RNA （SRR11304481）  call editing sites\n",
        "##### 2. ADAR_WT  1sample  paired-end HEK293 RNA （SRR11304482） call edting sites\n",
        "##### 3. WGS      1sample  paired-end HEK293 DNA （SRR2123657）  identify homozygous DNA sites"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dMedQR3WteRT"
      },
      "outputs": [],
      "source": [
        "# Directories\n",
        "BAM_DIR = './data/1_bam'\n",
        "dePCRduplicates_BAM_DIR = './data/2_dedup_bam'\n",
        "PILEUP_RAW_DIR = './data/3_pileup_raw'\n",
        "EDITING_DIR = './data/4_editing'\n",
        "\n",
        "# File paths\n",
        "REF_fa = './data/reference/hg38.fa'   # downloaded by yourself from gencode if there is not provided\n",
        "GTF = './data/reference/hg38.gtf'     # downloaded by yourself from gencode if there is not provided\n",
        "\n",
        "interested_genes = './data/interested_genes.txt'"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "y7uKhHS3teRT"
      },
      "source": [
        "### Analysis before BAM - preprocessed the raw FASTQ files in advance\n",
        "#### Step-1-trim - removed six bases to reduce the effect of random hexamer priming\n",
        "            cutadapt -q 20,20 --trim-n -m 15 -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT -u 6 -U 6\n",
        "#### Step-2-alignment\n",
        "##### WGS -\n",
        "            bwa mem -t 20 -M hg38.fa WGS_1.trim.fastq.gz WGS_2.trim.fastq.gz | samtools view -bS | samtools sort\n",
        "##### RNAseq -\n",
        "            hisat2\n",
        "#### Step-3-get intersting genes from bam\n",
        "            extracted reads from BAM files that map to around fifty gene regions"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_snukg-EteRT",
        "outputId": "8827e41e-4567-4025-a551-63fd1144f595"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "./data/1_bam/WGS.bam\n",
            "./data/1_bam/ADAR_WT.bam\n",
            "./data/1_bam/ADAR_KO.bam\n"
          ]
        }
      ],
      "source": [
        "# print bam files\n",
        "for bam in glob.glob(os.path.join(BAM_DIR, '*.bam')):\n",
        "    print(bam)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TD6K6k-YteRU"
      },
      "source": [
        "## #Denovo Call Editing - Start"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-3OTcsaWteRU"
      },
      "source": [
        "### Step 1: Remove PCR duplicates from bams\n",
        "        to remove duplicate reads generated during PCR amplification\n",
        "#### Used tools\n",
        "        Picard"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4zrmTNJqteRV",
        "outputId": "6537c6b6-adba-42e4-dfeb-9a208824d9e2"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "INFO\t2025-12-29 10:57:02\tMarkDuplicates\t\n",
            "\n",
            "********** NOTE: Picard's command line syntax is changing.\n",
            "**********\n",
            "********** For more information, please see:\n",
            "********** \n",
            "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
            "**********\n",
            "********** The command line looks like this in the new syntax:\n",
            "**********\n",
            "**********    MarkDuplicates -I ./data/1_bam/WGS.bam -O ./data/2_dedup_bam/WGS_dePCRdup.bam -M ./data/2_dedup_bam/WGS_dup_metrics.txt -REMOVE_DUPLICATES true -VALIDATION_STRINGENCY LENIENT -TMP_DIR ./data/2_dedup_bam/tmp\n",
            "**********\n",
            "\n",
            "\n",
            "10:57:02.398 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/public/data/lix/apps/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "[Mon Dec 29 10:57:02 CST 2025] MarkDuplicates INPUT=[./data/1_bam/WGS.bam] OUTPUT=./data/2_dedup_bam/WGS_dePCRdup.bam METRICS_FILE=./data/2_dedup_bam/WGS_dup_metrics.txt REMOVE_DUPLICATES=true TMP_DIR=[./data/2_dedup_bam/tmp] VALIDATION_STRINGENCY=LENIENT    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false FLOW_MODE=false FLOW_QUALITY_SUM_STRATEGY=false USE_END_IN_UNPAIRED_READS=false USE_UNPAIRED_CLIPPED_END=false UNPAIRED_END_UNCERTAINTY=0 FLOW_SKIP_FIRST_N_FLOWS=0 FLOW_Q_IS_KNOWN_END=false FLOW_EFFECTIVE_QUALITY_THRESHOLD=15 ADD_PG_TAG_TO_READS=true ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
            "[Mon Dec 29 10:57:02 CST 2025] Executing as lix@master on Linux 4.18.0-513.5.1.el8_9.x86_64 amd64; OpenJDK 64-Bit Server VM 1.8.0_392-b08; Deflater: Intel; Inflater: Intel; Provider GCS is not available; Picard version: 2.27.5\n",
            "INFO\t2025-12-29 10:57:02\tMarkDuplicates\tStart of doWork freeMemory: 1989873496; totalMemory: 2012217344; maxMemory: 28631367680\n",
            "INFO\t2025-12-29 10:57:02\tMarkDuplicates\tReading input file and constructing read end information.\n",
            "INFO\t2025-12-29 10:57:02\tMarkDuplicates\tWill retain up to 103736839 data points before spilling to disk.\n",
            "WARNING\t2025-12-29 10:57:02\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: SRR2123657.193553029. Cause: String 'SRR2123657.193553029' did not start with a parsable number.\n",
            "INFO\t2025-12-29 10:57:06\tMarkDuplicates\tRead 380526 records. 6867 pairs never matched.\n",
            "INFO\t2025-12-29 10:57:07\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 2145967152; totalMemory: 3039297536; maxMemory: 28631367680\n",
            "INFO\t2025-12-29 10:57:07\tMarkDuplicates\tWill retain up to 894730240 duplicate indices before spilling to disk.\n",
            "INFO\t2025-12-29 10:57:11\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
            "INFO\t2025-12-29 10:57:12\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
            "INFO\t2025-12-29 10:57:12\tMarkDuplicates\tSorting list of duplicate records.\n",
            "INFO\t2025-12-29 10:57:13\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 3014757560; totalMemory: 10197401600; maxMemory: 28631367680\n",
            "INFO\t2025-12-29 10:57:13\tMarkDuplicates\tMarking 9112 records as duplicates.\n",
            "INFO\t2025-12-29 10:57:13\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
            "INFO\t2025-12-29 10:57:13\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
            "INFO\t2025-12-29 10:57:16\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
            "INFO\t2025-12-29 10:57:16\tMarkDuplicates\tDuplicate Index cleanup.\n",
            "INFO\t2025-12-29 10:57:16\tMarkDuplicates\tGetting Memory Stats.\n",
            "INFO\t2025-12-29 10:57:16\tMarkDuplicates\tBefore output close freeMemory: 10172442504; totalMemory: 10197401600; maxMemory: 28631367680\n",
            "INFO\t2025-12-29 10:57:16\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
            "INFO\t2025-12-29 10:57:16\tMarkDuplicates\tAfter output close freeMemory: 10172523792; totalMemory: 10197401600; maxMemory: 28631367680\n",
            "[Mon Dec 29 10:57:16 CST 2025] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.24 minutes.\n",
            "Runtime.totalMemory()=10197401600\n",
            "INFO\t2025-12-29 10:57:17\tMarkDuplicates\t\n",
            "\n",
            "********** NOTE: Picard's command line syntax is changing.\n",
            "**********\n",
            "********** For more information, please see:\n",
            "********** \n",
            "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
            "**********\n",
            "********** The command line looks like this in the new syntax:\n",
            "**********\n",
            "**********    MarkDuplicates -I ./data/1_bam/ADAR_WT.bam -O ./data/2_dedup_bam/ADAR_WT_dePCRdup.bam -M ./data/2_dedup_bam/ADAR_WT_dup_metrics.txt -REMOVE_DUPLICATES true -VALIDATION_STRINGENCY LENIENT -TMP_DIR ./data/2_dedup_bam/tmp\n",
            "**********\n",
            "\n",
            "\n",
            "10:57:18.242 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/public/data/lix/apps/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "[Mon Dec 29 10:57:18 CST 2025] MarkDuplicates INPUT=[./data/1_bam/ADAR_WT.bam] OUTPUT=./data/2_dedup_bam/ADAR_WT_dePCRdup.bam METRICS_FILE=./data/2_dedup_bam/ADAR_WT_dup_metrics.txt REMOVE_DUPLICATES=true TMP_DIR=[./data/2_dedup_bam/tmp] VALIDATION_STRINGENCY=LENIENT    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false FLOW_MODE=false FLOW_QUALITY_SUM_STRATEGY=false USE_END_IN_UNPAIRED_READS=false USE_UNPAIRED_CLIPPED_END=false UNPAIRED_END_UNCERTAINTY=0 FLOW_SKIP_FIRST_N_FLOWS=0 FLOW_Q_IS_KNOWN_END=false FLOW_EFFECTIVE_QUALITY_THRESHOLD=15 ADD_PG_TAG_TO_READS=true ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
            "[Mon Dec 29 10:57:18 CST 2025] Executing as lix@master on Linux 4.18.0-513.5.1.el8_9.x86_64 amd64; OpenJDK 64-Bit Server VM 1.8.0_392-b08; Deflater: Intel; Inflater: Intel; Provider GCS is not available; Picard version: 2.27.5\n",
            "INFO\t2025-12-29 10:57:18\tMarkDuplicates\tStart of doWork freeMemory: 1989873424; totalMemory: 2012217344; maxMemory: 28631367680\n",
            "INFO\t2025-12-29 10:57:18\tMarkDuplicates\tReading input file and constructing read end information.\n",
            "INFO\t2025-12-29 10:57:18\tMarkDuplicates\tWill retain up to 103736839 data points before spilling to disk.\n",
            "WARNING\t2025-12-29 10:57:18\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: SRR11304482.3905544. Cause: String 'SRR11304482.3905544' did not start with a parsable number.\n",
            "INFO\t2025-12-29 10:57:22\tMarkDuplicates\tRead 329146 records. 8577 pairs never matched.\n",
            "INFO\t2025-12-29 10:57:23\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 2155933608; totalMemory: 3031957504; maxMemory: 28631367680\n",
            "INFO\t2025-12-29 10:57:23\tMarkDuplicates\tWill retain up to 894730240 duplicate indices before spilling to disk.\n",
            "INFO\t2025-12-29 10:57:27\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
            "INFO\t2025-12-29 10:57:27\tOpticalDuplicateFinder\tLarge duplicate set. size = 1036\n",
            "INFO\t2025-12-29 10:57:27\tOpticalDuplicateFinder\tcompared         1,000 ReadEnds to others.  Elapsed time: 00:00:00s.  Time for last 1,000:    0s.  Last read position: -1:-1\n",
            "INFO\t2025-12-29 10:57:27\tOpticalDuplicateFinder\tLarge duplicate set. size = 1103\n",
            "INFO\t2025-12-29 10:57:27\tOpticalDuplicateFinder\tcompared         1,000 ReadEnds to others.  Elapsed time: 00:00:00s.  Time for last 1,000:    0s.  Last read position: -1:-1\n",
            "INFO\t2025-12-29 10:57:27\tOpticalDuplicateFinder\tLarge duplicate set. size = 1182\n",
            "INFO\t2025-12-29 10:57:28\tOpticalDuplicateFinder\tcompared         1,000 ReadEnds to others.  Elapsed time: 00:00:00s.  Time for last 1,000:    0s.  Last read position: -1:-1\n",
            "INFO\t2025-12-29 10:57:28\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
            "INFO\t2025-12-29 10:57:28\tMarkDuplicates\tSorting list of duplicate records.\n",
            "INFO\t2025-12-29 10:57:29\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 3009274112; totalMemory: 10190061568; maxMemory: 28631367680\n",
            "INFO\t2025-12-29 10:57:29\tMarkDuplicates\tMarking 178310 records as duplicates.\n",
            "INFO\t2025-12-29 10:57:29\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
            "INFO\t2025-12-29 10:57:29\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
            "INFO\t2025-12-29 10:57:30\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
            "INFO\t2025-12-29 10:57:30\tMarkDuplicates\tDuplicate Index cleanup.\n",
            "INFO\t2025-12-29 10:57:30\tMarkDuplicates\tGetting Memory Stats.\n",
            "INFO\t2025-12-29 10:57:30\tMarkDuplicates\tBefore output close freeMemory: 10167090776; totalMemory: 10190061568; maxMemory: 28631367680\n",
            "INFO\t2025-12-29 10:57:30\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
            "INFO\t2025-12-29 10:57:30\tMarkDuplicates\tAfter output close freeMemory: 10167172064; totalMemory: 10190061568; maxMemory: 28631367680\n",
            "[Mon Dec 29 10:57:30 CST 2025] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.21 minutes.\n",
            "Runtime.totalMemory()=10190061568\n",
            "INFO\t2025-12-29 10:57:31\tMarkDuplicates\t\n",
            "\n",
            "********** NOTE: Picard's command line syntax is changing.\n",
            "**********\n",
            "********** For more information, please see:\n",
            "********** \n",
            "https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
            "**********\n",
            "********** The command line looks like this in the new syntax:\n",
            "**********\n",
            "**********    MarkDuplicates -I ./data/1_bam/ADAR_KO.bam -O ./data/2_dedup_bam/ADAR_KO_dePCRdup.bam -M ./data/2_dedup_bam/ADAR_KO_dup_metrics.txt -REMOVE_DUPLICATES true -VALIDATION_STRINGENCY LENIENT -TMP_DIR ./data/2_dedup_bam/tmp\n",
            "**********\n",
            "\n",
            "\n",
            "10:57:32.140 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/public/data/lix/apps/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "[Mon Dec 29 10:57:32 CST 2025] MarkDuplicates INPUT=[./data/1_bam/ADAR_KO.bam] OUTPUT=./data/2_dedup_bam/ADAR_KO_dePCRdup.bam METRICS_FILE=./data/2_dedup_bam/ADAR_KO_dup_metrics.txt REMOVE_DUPLICATES=true TMP_DIR=[./data/2_dedup_bam/tmp] VALIDATION_STRINGENCY=LENIENT    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false FLOW_MODE=false FLOW_QUALITY_SUM_STRATEGY=false USE_END_IN_UNPAIRED_READS=false USE_UNPAIRED_CLIPPED_END=false UNPAIRED_END_UNCERTAINTY=0 FLOW_SKIP_FIRST_N_FLOWS=0 FLOW_Q_IS_KNOWN_END=false FLOW_EFFECTIVE_QUALITY_THRESHOLD=15 ADD_PG_TAG_TO_READS=true ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
            "[Mon Dec 29 10:57:32 CST 2025] Executing as lix@master on Linux 4.18.0-513.5.1.el8_9.x86_64 amd64; OpenJDK 64-Bit Server VM 1.8.0_392-b08; Deflater: Intel; Inflater: Intel; Provider GCS is not available; Picard version: 2.27.5\n",
            "INFO\t2025-12-29 10:57:32\tMarkDuplicates\tStart of doWork freeMemory: 1989873432; totalMemory: 2012217344; maxMemory: 28631367680\n",
            "INFO\t2025-12-29 10:57:32\tMarkDuplicates\tReading input file and constructing read end information.\n",
            "INFO\t2025-12-29 10:57:32\tMarkDuplicates\tWill retain up to 103736839 data points before spilling to disk.\n",
            "WARNING\t2025-12-29 10:57:32\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: SRR11304481.1873917. Cause: String 'SRR11304481.1873917' did not start with a parsable number.\n",
            "INFO\t2025-12-29 10:57:35\tMarkDuplicates\tRead 329469 records. 7141 pairs never matched.\n",
            "INFO\t2025-12-29 10:57:36\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 2169257760; totalMemory: 3047161856; maxMemory: 28631367680\n",
            "INFO\t2025-12-29 10:57:36\tMarkDuplicates\tWill retain up to 894730240 duplicate indices before spilling to disk.\n",
            "INFO\t2025-12-29 10:57:40\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
            "INFO\t2025-12-29 10:57:40\tOpticalDuplicateFinder\tLarge duplicate set. size = 1072\n",
            "INFO\t2025-12-29 10:57:40\tOpticalDuplicateFinder\tcompared         1,000 ReadEnds to others.  Elapsed time: 00:00:00s.  Time for last 1,000:    0s.  Last read position: -1:-1\n",
            "INFO\t2025-12-29 10:57:40\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
            "INFO\t2025-12-29 10:57:41\tMarkDuplicates\tSorting list of duplicate records.\n",
            "INFO\t2025-12-29 10:57:42\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 3022621048; totalMemory: 10205265920; maxMemory: 28631367680\n",
            "INFO\t2025-12-29 10:57:42\tMarkDuplicates\tMarking 175562 records as duplicates.\n",
            "INFO\t2025-12-29 10:57:42\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
            "INFO\t2025-12-29 10:57:42\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
            "INFO\t2025-12-29 10:57:43\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
            "INFO\t2025-12-29 10:57:43\tMarkDuplicates\tDuplicate Index cleanup.\n",
            "INFO\t2025-12-29 10:57:43\tMarkDuplicates\tGetting Memory Stats.\n",
            "INFO\t2025-12-29 10:57:43\tMarkDuplicates\tBefore output close freeMemory: 10180571360; totalMemory: 10205265920; maxMemory: 28631367680\n",
            "INFO\t2025-12-29 10:57:43\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
            "INFO\t2025-12-29 10:57:43\tMarkDuplicates\tAfter output close freeMemory: 10180652648; totalMemory: 10205265920; maxMemory: 28631367680\n",
            "[Mon Dec 29 10:57:43 CST 2025] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.19 minutes.\n",
            "Runtime.totalMemory()=10205265920\n"
          ]
        }
      ],
      "source": [
        "# dePCR duplicates\n",
        "for raw_bam in glob.glob(os.path.join(BAM_DIR, '*.bam')):\n",
        "    sample_name = os.path.basename(raw_bam).split('.bam')[0]\n",
        "    out_bam = os.path.join(dePCRduplicates_BAM_DIR, f'{sample_name}_dePCRdup.bam')\n",
        "    subprocess.run(f'java -jar /public/home/lix/lix/apps/picard.jar MarkDuplicates I={raw_bam} O={out_bam} M={os.path.join(dePCRduplicates_BAM_DIR, sample_name + \"_dup_metrics.txt\")} REMOVE_DUPLICATES=true VALIDATION_STRINGENCY=LENIENT TMP_DIR={dePCRduplicates_BAM_DIR}/tmp', shell=True)\n",
        "    subprocess.run(f'/public/apps/bin/samtools index {out_bam}', shell=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yyhTOPQjteRV"
      },
      "source": [
        "### Step 2: Call and filter variants\n",
        "#### Filters:\n",
        "        Base quality >= 20\n",
        "        Mapping quelity >= 20\n",
        "        min_coverage >= 2\n",
        "        min_variant_frequency >= 0.02\n",
        "        min_variant_count >= 2\n",
        "#### Used tools：\n",
        "        Samtools mplieup"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GhJxIvdMteRV"
      },
      "source": [
        "### Step 2: Call variants\n",
        "##### Used tools：\n",
        "        Samtools mplieup"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4O9py3cuteRV",
        "outputId": "9c884d6f-500c-45e9-b37f-34c7aaaab86e"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "./data/2_dedup_bam/WGS_dePCRdup.bam\n",
            "./data/2_dedup_bam/ADAR_KO_dePCRdup.bam\n",
            "./data/2_dedup_bam/ADAR_WT_dePCRdup.bam\n"
          ]
        }
      ],
      "source": [
        "# print dePCR duplicates removed bam files\n",
        "for dedup_bam in glob.glob(os.path.join(dePCRduplicates_BAM_DIR, '*_dePCRdup.bam')):\n",
        "    print(dedup_bam)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "UjjeWTWLteRW",
        "outputId": "18d4f923-13e9-41b4-d46b-3d929277cbcb"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "[mpileup] 1 samples in 1 input files\n",
            "[mpileup] 1 samples in 1 input files\n",
            "[mpileup] 1 samples in 1 input files\n"
          ]
        }
      ],
      "source": [
        "# 1 - mpileup for all genes\n",
        "PILEUP_RAW_DIR = './data/3_pileup_raw'\n",
        "\n",
        "for bam in glob.glob(os.path.join(dePCRduplicates_BAM_DIR, '*_dePCRdup.bam')):\n",
        "    sample_name = os.path.basename(bam).split('_dePCRdup.bam')[0]\n",
        "    out_pileup = os.path.join(PILEUP_RAW_DIR, f'{sample_name}_pileup.txt')\n",
        "    subprocess.run(f'/public/apps/bin/samtools mpileup -A -B -f {REF_fa} -d 100000 -Q 20 -q 20 {bam} > {out_pileup}', shell=True)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ATqV6J_JteRW",
        "outputId": "02b64b36-ceb5-49a2-d792-fb987d5799de"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_KO mpileup results:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>pos</th>\n",
              "      <th>ref_base</th>\n",
              "      <th>coverage</th>\n",
              "      <th>bases</th>\n",
              "      <th>base_qualities</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16950361</td>\n",
              "      <td>C</td>\n",
              "      <td>2</td>\n",
              "      <td>^].^].</td>\n",
              "      <td>JJ</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16950362</td>\n",
              "      <td>C</td>\n",
              "      <td>2</td>\n",
              "      <td>..</td>\n",
              "      <td>JJ</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16950363</td>\n",
              "      <td>T</td>\n",
              "      <td>2</td>\n",
              "      <td>..</td>\n",
              "      <td>JJ</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16950364</td>\n",
              "      <td>A</td>\n",
              "      <td>2</td>\n",
              "      <td>..</td>\n",
              "      <td>JJ</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16950365</td>\n",
              "      <td>T</td>\n",
              "      <td>2</td>\n",
              "      <td>..</td>\n",
              "      <td>AA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4020032</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154400576</td>\n",
              "      <td>G</td>\n",
              "      <td>1</td>\n",
              "      <td>,</td>\n",
              "      <td>A</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4020033</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154400577</td>\n",
              "      <td>T</td>\n",
              "      <td>1</td>\n",
              "      <td>,</td>\n",
              "      <td>J</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4020034</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154400578</td>\n",
              "      <td>G</td>\n",
              "      <td>1</td>\n",
              "      <td>,</td>\n",
              "      <td>J</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4020035</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154400579</td>\n",
              "      <td>A</td>\n",
              "      <td>1</td>\n",
              "      <td>,</td>\n",
              "      <td>J</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4020036</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154400580</td>\n",
              "      <td>T</td>\n",
              "      <td>1</td>\n",
              "      <td>,$</td>\n",
              "      <td>J</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4020037 rows × 6 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "        chrom        pos ref_base  coverage   bases base_qualities\n",
              "0        chr1   16950361        C         2  ^].^].             JJ\n",
              "1        chr1   16950362        C         2      ..             JJ\n",
              "2        chr1   16950363        T         2      ..             JJ\n",
              "3        chr1   16950364        A         2      ..             JJ\n",
              "4        chr1   16950365        T         2      ..             AA\n",
              "...       ...        ...      ...       ...     ...            ...\n",
              "4020032  chrX  154400576        G         1       ,              A\n",
              "4020033  chrX  154400577        T         1       ,              J\n",
              "4020034  chrX  154400578        G         1       ,              J\n",
              "4020035  chrX  154400579        A         1       ,              J\n",
              "4020036  chrX  154400580        T         1      ,$              J\n",
              "\n",
              "[4020037 rows x 6 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Filtered A-to-G editing sites:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>pos</th>\n",
              "      <th>ref_base</th>\n",
              "      <th>coverage</th>\n",
              "      <th>bases</th>\n",
              "      <th>base_qualities</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>24924</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975285</td>\n",
              "      <td>A</td>\n",
              "      <td>123</td>\n",
              "      <td>&gt;&gt;GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG...</td>\n",
              "      <td>FFJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25611</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975972</td>\n",
              "      <td>A</td>\n",
              "      <td>190</td>\n",
              "      <td>&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;...</td>\n",
              "      <td>FFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJAAJJJJJJJJJJJJ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26750</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16977111</td>\n",
              "      <td>A</td>\n",
              "      <td>158</td>\n",
              "      <td>&gt;&gt;....&lt;&lt;......................&gt;&gt;.................</td>\n",
              "      <td>FFJJFFJJAAFFJJJJFFFFAAJJJJ77JJFFJJJJ&lt;&lt;FFJJFFJJ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>101890</th>\n",
              "      <td>chr1</td>\n",
              "      <td>92841775</td>\n",
              "      <td>T</td>\n",
              "      <td>267</td>\n",
              "      <td>&gt;&gt;&lt;&lt;&gt;&gt;,$,$,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,...</td>\n",
              "      <td>JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>159272</th>\n",
              "      <td>chr1</td>\n",
              "      <td>115831686</td>\n",
              "      <td>A</td>\n",
              "      <td>2</td>\n",
              "      <td>.-1G.-1G</td>\n",
              "      <td>JJ</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3990433</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154370977</td>\n",
              "      <td>T</td>\n",
              "      <td>211</td>\n",
              "      <td>&lt;.$.$............................................</td>\n",
              "      <td>FAAJJJJFFJJ&lt;&lt;JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3990457</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154371001</td>\n",
              "      <td>T</td>\n",
              "      <td>228</td>\n",
              "      <td>&lt;................................................</td>\n",
              "      <td>FJJJJFFJJJJJJJJJJJJJJFFJJJJ&lt;&lt;JJJJJJJJJJJJJJJJJ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3990461</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154371005</td>\n",
              "      <td>A</td>\n",
              "      <td>228</td>\n",
              "      <td>&lt;.$.$............................................</td>\n",
              "      <td>FJJJJJJJJJJJJFFJJJJFFJJJJJJFFJJJJFFJJJJJJJJJJJ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3990605</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154371149</td>\n",
              "      <td>T</td>\n",
              "      <td>145</td>\n",
              "      <td>&lt;.$.$................................CC..........</td>\n",
              "      <td>FJJJJJJJJJJJJJJFFJJJJJJJJJJJJJJJJJJJJ&lt;&lt;JJJJJJJ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3990733</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154371277</td>\n",
              "      <td>T</td>\n",
              "      <td>114</td>\n",
              "      <td>&lt;.$.$............................CC..............</td>\n",
              "      <td>FJJJJ&lt;&lt;JJJJJJJJJJJJJJFFJJJJJJJJFFJJJJFFJJ&lt;&lt;JJJ...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>286 rows × 6 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "        chrom        pos ref_base  coverage  \\\n",
              "24924    chr1   16975285        A       123   \n",
              "25611    chr1   16975972        A       190   \n",
              "26750    chr1   16977111        A       158   \n",
              "101890   chr1   92841775        T       267   \n",
              "159272   chr1  115831686        A         2   \n",
              "...       ...        ...      ...       ...   \n",
              "3990433  chrX  154370977        T       211   \n",
              "3990457  chrX  154371001        T       228   \n",
              "3990461  chrX  154371005        A       228   \n",
              "3990605  chrX  154371149        T       145   \n",
              "3990733  chrX  154371277        T       114   \n",
              "\n",
              "                                                     bases  \\\n",
              "24924    >>GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG...   \n",
              "25611    >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>...   \n",
              "26750    >>....<<......................>>.................   \n",
              "101890   >><<>>,$,$,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,...   \n",
              "159272                                            .-1G.-1G   \n",
              "...                                                    ...   \n",
              "3990433  <.$.$............................................   \n",
              "3990457  <................................................   \n",
              "3990461  <.$.$............................................   \n",
              "3990605  <.$.$................................CC..........   \n",
              "3990733  <.$.$............................CC..............   \n",
              "\n",
              "                                            base_qualities  \n",
              "24924    FFJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJ...  \n",
              "25611    FFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJAAJJJJJJJJJJJJ...  \n",
              "26750    FFJJFFJJAAFFJJJJFFFFAAJJJJ77JJFFJJJJ<<FFJJFFJJ...  \n",
              "101890   JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ...  \n",
              "159272                                                  JJ  \n",
              "...                                                    ...  \n",
              "3990433  FAAJJJJFFJJ<<JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ...  \n",
              "3990457  FJJJJFFJJJJJJJJJJJJJJFFJJJJ<<JJJJJJJJJJJJJJJJJ...  \n",
              "3990461  FJJJJJJJJJJJJFFJJJJFFJJJJJJFFJJJJFFJJJJJJJJJJJ...  \n",
              "3990605  FJJJJJJJJJJJJJJFFJJJJJJJJJJJJJJJJJJJJ<<JJJJJJJ...  \n",
              "3990733  FJJJJ<<JJJJJJJJJJJJJJFFJJJJJJJJFFJJJJFFJJ<<JJJ...  \n",
              "\n",
              "[286 rows x 6 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# 2 - display pileup df results (example for ADAR_KO)\n",
        "ADAR_KO_pileup = os.path.join(PILEUP_RAW_DIR, 'ADAR_KO_pileup.txt')\n",
        "pileup_df = pd.read_csv(ADAR_KO_pileup, sep='\\t', header=None,\n",
        "                         names=['chrom', 'pos', 'ref_base', 'coverage', 'bases', 'base_qualities'])\n",
        "\n",
        "print(\"ADAR_KO mpileup results:\")\n",
        "display(pileup_df)\n",
        "\n",
        "# filter A-to-G / T-to-C editing sites (example, not the final results)\n",
        "filtered_df = pileup_df[\n",
        "    ((pileup_df['ref_base'] == 'A') & (pileup_df['bases'].str.contains('G'))) |\n",
        "    ((pileup_df['ref_base'] == 'T') & (pileup_df['bases'].str.contains('C')))\n",
        "]\n",
        "print(\"Filtered A-to-G editing sites:\")\n",
        "display(filtered_df)\n",
        "\n",
        "# parse mpileup results\n",
        "# 1. match\n",
        "##   .\tBase matches the reference on the forward strand\n",
        "##   ,\tBase matches the reference on the reverse strand\n",
        "# 2. mismatch\n",
        "##   A C G T\tMismatch on the forward strand\n",
        "##   a c g t\tMismatch on the reverse strand\n",
        "# 3. indels\n",
        "##  +[length][sequence] example: +2AG\n",
        "##  -[length][sequence] example: -2AG\n",
        "# 4. Start and end of bases\n",
        "##  ^\tStart of a read segment\n",
        "##  $\tEnd of a read\n",
        "# 5. deletion\n",
        "##  *\tPlaceholder for a deletion\n",
        "# 6. Reference skips (spliced alignments)\n",
        "##  >\tReference skip on the forward strand\n",
        "##  <\tReference skip on the reverse strand"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LFcTfbnWteRW"
      },
      "source": [
        "### Step 3: filter variants"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bVvCXgvyteRX"
      },
      "source": [
        "#### Filters-1 -> get 'Alu RNA editing sites' and 'non-Alu mismatches':\n",
        "        1) require nonsingleton mismatch in RNA\n",
        "        Base quality >= 20\n",
        "        Mapping quelity >= 20\n",
        "        min_coverage >= 2\n",
        "        min_variant_ratio >= 0.02\n",
        "        min_variant_count >= 2\n",
        "        2)\n",
        "        Require homozygous DNA（min_cov=10, min_major_frac=0.95, max_minor_count=2, base_quality >= 20）\n",
        "        3)\n",
        "        Remove SNP\n",
        "        4)\n",
        "        Remove mismatches at 5' read ends(has done in \"cutadapt -u 6 -U 6\")\n",
        "#### Data\n",
        "        1) pileup results:\n",
        "                        1 - raw RNAseq 'samtools mpileup' results data\n",
        "                        2 - raw WGS 'samtools mpileup' results data -> to get homozygous DNA\n",
        "        2) required filter files:\n",
        "                        1 - Alu region data\n",
        "                        2 - SNP region data（data from 1000Genomes）\n",
        "##### Used tools：\n",
        "        bedtools\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_pSt2CNTteRX"
      },
      "source": [
        "#### 1 - prepare the homozygous DNA data and SNP data\n",
        "        A site was called homozygous if 10 or more reads contained the same base that represented more than 95% of the complete coverage and if only 2 or fewer alternative bases were present at the same position."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Zss9npOMteRX"
      },
      "outputs": [],
      "source": [
        "def get_homozygous_sites_from_WGSpileup_df(WGS_raw_pileup_df,min_coverage=10,min_majorBase_fraction=0.95,max_minorBase_count=2):\n",
        "\n",
        "    def parse_WGSpileup_bases_df(WGS_raw_pileup_df, min_base_qual=20):\n",
        "        df = WGS_raw_pileup_df.copy()\n",
        "        df[\"A\"] = 0\n",
        "        df[\"T\"] = 0\n",
        "        df[\"C\"] = 0\n",
        "        df[\"G\"] = 0\n",
        "        df[\"ref_count\"] = 0\n",
        "\n",
        "        for idx, row in df.iterrows():\n",
        "            bases = row[\"bases\"]\n",
        "            quals = row[\"quals\"]\n",
        "\n",
        "            i = 0\n",
        "            q = 0\n",
        "            n = len(bases)\n",
        "\n",
        "            while i < n:\n",
        "                b = bases[i]\n",
        "\n",
        "                # read start (^ + MAPQ)\n",
        "                if b == \"^\":\n",
        "                    i += 2\n",
        "                    continue\n",
        "\n",
        "                # read end\n",
        "                if b == \"$\":\n",
        "                    i += 1\n",
        "                    continue\n",
        "\n",
        "                # indel\n",
        "                if b in \"+-\":\n",
        "                    i += 1\n",
        "                    num = \"\"\n",
        "                    while i < n and bases[i].isdigit():\n",
        "                        num += bases[i]\n",
        "                        i += 1\n",
        "                    i += int(num)\n",
        "                    continue\n",
        "\n",
        "                # base quality\n",
        "                if q >= len(quals):\n",
        "                    break\n",
        "\n",
        "                base_qual = ord(quals[q]) - 33\n",
        "                q += 1\n",
        "\n",
        "                if base_qual < min_base_qual:\n",
        "                    i += 1\n",
        "                    continue\n",
        "\n",
        "                if b in \".,\":                 # reference\n",
        "                    df.at[idx, \"ref_count\"] += 1\n",
        "                elif b.upper() in [\"A\", \"T\", \"C\", \"G\"]:\n",
        "                    df.at[idx, b.upper()] += 1\n",
        "\n",
        "                i += 1\n",
        "\n",
        "        return df\n",
        "\n",
        "    pileup_df = parse_WGSpileup_bases_df(WGS_raw_pileup_df)\n",
        "\n",
        "    homo_sites = []\n",
        "    for _, row in pileup_df.iterrows():\n",
        "        A, T, C, G = row[\"A\"], row[\"T\"], row[\"C\"], row[\"G\"]\n",
        "        ref = row[\"ref\"].upper()\n",
        "        ref_count = row[\"ref_count\"]\n",
        "\n",
        "        coverage = A + T + C + G + ref_count\n",
        "        if coverage < min_coverage:\n",
        "            continue\n",
        "\n",
        "        base_counts = {\n",
        "            \"A\": A,\n",
        "            \"T\": T,\n",
        "            \"C\": C,\n",
        "            \"G\": G,\n",
        "            ref: ref_count\n",
        "        }\n",
        "\n",
        "        major_base = max(base_counts, key=base_counts.get)\n",
        "        major_count = base_counts[major_base]\n",
        "        minor_count = coverage - major_count\n",
        "\n",
        "        if (\n",
        "            major_count / coverage >= min_majorBase_fraction and\n",
        "            minor_count <= max_minorBase_count\n",
        "        ):\n",
        "            homo_sites.append({\n",
        "                \"chr\": row[\"chr\"],\n",
        "                \"pos\": row[\"pos\"],\n",
        "                \"ref\": ref,\n",
        "                \"major_base\": major_base,\n",
        "                \"coverage\": coverage,\n",
        "                \"major_frac\": round(major_count / coverage, 4)\n",
        "            })\n",
        "\n",
        "    return pd.DataFrame(homo_sites)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "b4YtLzR8teRX",
        "outputId": "68174fe4-fdf1-46b3-f576-5add5fab38f9"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "homozygous DNA sites:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chr</th>\n",
              "      <th>pos</th>\n",
              "      <th>ref</th>\n",
              "      <th>major_base</th>\n",
              "      <th>coverage</th>\n",
              "      <th>major_frac</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975070</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>10</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975071</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>10</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975072</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>10</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975075</td>\n",
              "      <td>C</td>\n",
              "      <td>C</td>\n",
              "      <td>10</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975076</td>\n",
              "      <td>C</td>\n",
              "      <td>C</td>\n",
              "      <td>10</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1777154</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154374502</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "      <td>12</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1777155</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154374503</td>\n",
              "      <td>C</td>\n",
              "      <td>C</td>\n",
              "      <td>12</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1777156</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154374504</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>12</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1777157</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154374505</td>\n",
              "      <td>C</td>\n",
              "      <td>C</td>\n",
              "      <td>12</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1777158</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154374506</td>\n",
              "      <td>C</td>\n",
              "      <td>C</td>\n",
              "      <td>10</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1777159 rows × 6 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "          chr        pos ref major_base  coverage  major_frac\n",
              "0        chr1   16975070   G          G        10         1.0\n",
              "1        chr1   16975071   G          G        10         1.0\n",
              "2        chr1   16975072   G          G        10         1.0\n",
              "3        chr1   16975075   C          C        10         1.0\n",
              "4        chr1   16975076   C          C        10         1.0\n",
              "...       ...        ...  ..        ...       ...         ...\n",
              "1777154  chrX  154374502   T          T        12         1.0\n",
              "1777155  chrX  154374503   C          C        12         1.0\n",
              "1777156  chrX  154374504   A          A        12         1.0\n",
              "1777157  chrX  154374505   C          C        12         1.0\n",
              "1777158  chrX  154374506   C          C        10         1.0\n",
              "\n",
              "[1777159 rows x 6 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# invoke 'get_homozygous_sites_from_WGSpileup_df' function on WGS pileup data\n",
        "WGS_pileup = './data/3_pileup_raw/WGS_pileup.txt'\n",
        "WGS_pileup_df = pd.read_csv(WGS_pileup, sep='\\t', header=None, names=['chr','pos','ref','coverage','bases','quals'])\n",
        "homozygous_DNA_sites_df = get_homozygous_sites_from_WGSpileup_df(WGS_pileup_df)\n",
        "homozygous_DNA_sites_df.to_csv(os.path.join(EDITING_DIR, 'homozygous_DNA_sites.txt'), sep='\\t', index=False)\n",
        "\n",
        "print('homozygous DNA sites:')\n",
        "display(homozygous_DNA_sites_df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "tfgL4RovteRY",
        "outputId": "3a3359ec-b09e-47af-9ec0-cffd63c04f98"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SNP sites：\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>10176</td>\n",
              "      <td>10177</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>10351</td>\n",
              "      <td>10352</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>10615</td>\n",
              "      <td>10616</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>11007</td>\n",
              "      <td>11008</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>11011</td>\n",
              "      <td>11012</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046910</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>82899</td>\n",
              "      <td>82900</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046911</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>82921</td>\n",
              "      <td>82922</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046912</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>82929</td>\n",
              "      <td>82930</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046913</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>82936</td>\n",
              "      <td>82937</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046914</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>85465</td>\n",
              "      <td>85466</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>16046915 rows × 3 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                            chrom  start    end\n",
              "0                            chr1  10176  10177\n",
              "1                            chr1  10351  10352\n",
              "2                            chr1  10615  10616\n",
              "3                            chr1  11007  11008\n",
              "4                            chr1  11011  11012\n",
              "...                           ...    ...    ...\n",
              "16046910  chr22_KI270738v1_random  82899  82900\n",
              "16046911  chr22_KI270738v1_random  82921  82922\n",
              "16046912  chr22_KI270738v1_random  82929  82930\n",
              "16046913  chr22_KI270738v1_random  82936  82937\n",
              "16046914  chr22_KI270738v1_random  85465  85466\n",
              "\n",
              "[16046915 rows x 3 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Alu region：\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>10176</td>\n",
              "      <td>10177</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>10351</td>\n",
              "      <td>10352</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>10615</td>\n",
              "      <td>10616</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>11007</td>\n",
              "      <td>11008</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>11011</td>\n",
              "      <td>11012</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046910</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>82899</td>\n",
              "      <td>82900</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046911</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>82921</td>\n",
              "      <td>82922</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046912</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>82929</td>\n",
              "      <td>82930</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046913</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>82936</td>\n",
              "      <td>82937</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046914</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>85465</td>\n",
              "      <td>85466</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>16046915 rows × 3 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                            chrom  start    end\n",
              "0                            chr1  10176  10177\n",
              "1                            chr1  10351  10352\n",
              "2                            chr1  10615  10616\n",
              "3                            chr1  11007  11008\n",
              "4                            chr1  11011  11012\n",
              "...                           ...    ...    ...\n",
              "16046910  chr22_KI270738v1_random  82899  82900\n",
              "16046911  chr22_KI270738v1_random  82921  82922\n",
              "16046912  chr22_KI270738v1_random  82929  82930\n",
              "16046913  chr22_KI270738v1_random  82936  82937\n",
              "16046914  chr22_KI270738v1_random  85465  85466\n",
              "\n",
              "[16046915 rows x 3 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "nonAlu_repeat region：\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>10176</td>\n",
              "      <td>10177</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>10351</td>\n",
              "      <td>10352</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>10615</td>\n",
              "      <td>10616</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>11007</td>\n",
              "      <td>11008</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>11011</td>\n",
              "      <td>11012</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046910</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>82899</td>\n",
              "      <td>82900</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046911</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>82921</td>\n",
              "      <td>82922</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046912</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>82929</td>\n",
              "      <td>82930</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046913</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>82936</td>\n",
              "      <td>82937</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16046914</th>\n",
              "      <td>chr22_KI270738v1_random</td>\n",
              "      <td>85465</td>\n",
              "      <td>85466</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>16046915 rows × 3 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                            chrom  start    end\n",
              "0                            chr1  10176  10177\n",
              "1                            chr1  10351  10352\n",
              "2                            chr1  10615  10616\n",
              "3                            chr1  11007  11008\n",
              "4                            chr1  11011  11012\n",
              "...                           ...    ...    ...\n",
              "16046910  chr22_KI270738v1_random  82899  82900\n",
              "16046911  chr22_KI270738v1_random  82921  82922\n",
              "16046912  chr22_KI270738v1_random  82929  82930\n",
              "16046913  chr22_KI270738v1_random  82936  82937\n",
              "16046914  chr22_KI270738v1_random  85465  85466\n",
              "\n",
              "[16046915 rows x 3 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# show the prepared SNP data, Alu region data\n",
        "SNP_df = pd.read_csv('./data/reference/hg38_SNP.bed', sep = '\\t', header = None, names=['chrom', 'start', 'end'])\n",
        "Alu_df = pd.read_csv('./data/reference/hg38_Alu.bed', sep = '\\t', header = None, names=['chrom', 'start', 'end'])\n",
        "\n",
        "print('SNP sites：')\n",
        "display(SNP_df)\n",
        "print('Alu region：')\n",
        "display(SNP_df)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mtCqDl0ZteRY"
      },
      "source": [
        "#### 2 - Filter:\n",
        "        1)\n",
        "        require nonsingleton mismatch in RNA\n",
        "        2)\n",
        "        Require homozygous DNA（min_cov=10, min_major_frac=0.95, max_minor_count=2, base_quality >= 20）\n",
        "        3)\n",
        "        Remove SNP\n",
        "        4)\n",
        "        Remove mismatches at 5' read ends(has done in \"cutadapt -u 6 -U 6\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "g5xYU5rNteRY"
      },
      "source": [
        "##### 1） require nonsingleton mismatch in RNA\n",
        "        - base quality >= 20\n",
        "        - reads coverage >= 2\n",
        "        - variant count >= 2\n",
        "        - variant ratio >= 0.02"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "iJpSnwj_teRY"
      },
      "outputs": [],
      "source": [
        "def get_nonsingleton_mismatch_df(raw_pileup_df, min_base_quality=20, min_coverage=2, max_coverage=1_000_000,\n",
        "                                    min_variant_frequency=0.02, min_variant_count=2):\n",
        "    \"\"\"\n",
        "    Parameters\n",
        "    ----------\n",
        "    raw_pileup_df : pd.DataFrame\n",
        "    min_base_quality : int\n",
        "    min_cov : int\n",
        "    max_cov : int\n",
        "    min_var_freq : float\n",
        "    min_var_count : int\n",
        "\n",
        "    \"\"\"\n",
        "    results = []\n",
        "\n",
        "    for _, row in raw_pileup_df.iterrows():\n",
        "        chr_ = row['chr']\n",
        "        pos = row['pos']\n",
        "        ref_base = row['ref']\n",
        "        coverage = row['coverage']\n",
        "        bases = row['bases']\n",
        "        quality = row['quality']\n",
        "\n",
        "        if coverage < min_coverage or coverage > max_coverage:\n",
        "            continue\n",
        "\n",
        "        # parse bases and qualities\n",
        "        pile_bases = list(bases)\n",
        "        quality_scores = [ord(q)-33 for q in quality]  # Illumina phred+33\n",
        "\n",
        "        # 初始化计数\n",
        "        counts = {'A':0,'T':0,'C':0,'G':0,'ref':0}\n",
        "\n",
        "        i = 0\n",
        "        read_idx = 0\n",
        "        while i < len(pile_bases):\n",
        "            base = pile_bases[i]\n",
        "\n",
        "            # ignore start and end of read\n",
        "            if base == '^':  # skip the next character (mapping quality)\n",
        "                i += 2\n",
        "                continue\n",
        "            elif base == '$':\n",
        "                i += 1\n",
        "                continue\n",
        "\n",
        "            # 忽略 indel\n",
        "            if base in '+-':\n",
        "                i += 1\n",
        "                indel_len = ''\n",
        "                while pile_bases[i].isdigit():\n",
        "                    indel_len += pile_bases[i]\n",
        "                    i += 1\n",
        "                indel_len = int(indel_len)\n",
        "                i += indel_len\n",
        "                continue\n",
        "\n",
        "            # 碱基质量\n",
        "            if read_idx < len(quality_scores) and quality_scores[read_idx] >= min_base_quality:\n",
        "                if base in '.,':  # match reference\n",
        "                    counts['ref'] += 1\n",
        "                elif base.upper() == 'A':\n",
        "                    counts['A'] += 1\n",
        "                elif base.upper() == 'T':\n",
        "                    counts['T'] += 1\n",
        "                elif base.upper() == 'C':\n",
        "                    counts['C'] += 1\n",
        "                elif base.upper() == 'G':\n",
        "                    counts['G'] += 1\n",
        "            read_idx += 1\n",
        "            i += 1\n",
        "\n",
        "        total_count = counts['ref'] + counts['A'] + counts['T'] + counts['C'] + counts['G']\n",
        "\n",
        "        # 筛选可能变异\n",
        "        for nuc in ['A','T','C','G']:\n",
        "            alt_count = counts[nuc]\n",
        "            if alt_count >= min_variant_count and (alt_count / total_count) >= min_variant_frequency:\n",
        "                var_freq = alt_count / total_count\n",
        "                results.append([chr_, pos, total_count, ref_base, nuc, round(var_freq,3), alt_count])\n",
        "\n",
        "    result_df = pd.DataFrame(results, columns=['chr','pos','coverage','ref','alt','var_freq','alt_count'])\n",
        "    return result_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "q4Get6bFteRZ",
        "outputId": "1051aca4-bae4-43e9-d8ae-3df04db725ce"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_KO editing sites (raw):\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chr</th>\n",
              "      <th>pos</th>\n",
              "      <th>coverage</th>\n",
              "      <th>ref</th>\n",
              "      <th>alt</th>\n",
              "      <th>var_freq</th>\n",
              "      <th>alt_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16974964</td>\n",
              "      <td>72</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.028</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975285</td>\n",
              "      <td>121</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>1.000</td>\n",
              "      <td>121</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975628</td>\n",
              "      <td>6</td>\n",
              "      <td>C</td>\n",
              "      <td>A</td>\n",
              "      <td>0.333</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16976085</td>\n",
              "      <td>26</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.077</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16977094</td>\n",
              "      <td>3</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>1.000</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>902</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154357212</td>\n",
              "      <td>31</td>\n",
              "      <td>C</td>\n",
              "      <td>G</td>\n",
              "      <td>0.065</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>903</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365712</td>\n",
              "      <td>6</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>0.667</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>904</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154366167</td>\n",
              "      <td>221</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.982</td>\n",
              "      <td>217</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>905</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368109</td>\n",
              "      <td>7</td>\n",
              "      <td>C</td>\n",
              "      <td>G</td>\n",
              "      <td>0.286</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>906</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368228</td>\n",
              "      <td>10</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.200</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>907 rows × 7 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      chr        pos  coverage ref alt  var_freq  alt_count\n",
              "0    chr1   16974964        72   G   A     0.028          2\n",
              "1    chr1   16975285       121   A   G     1.000        121\n",
              "2    chr1   16975628         6   C   A     0.333          2\n",
              "3    chr1   16976085        26   G   A     0.077          2\n",
              "4    chr1   16977094         3   C   T     1.000          3\n",
              "..    ...        ...       ...  ..  ..       ...        ...\n",
              "902  chrX  154357212        31   C   G     0.065          2\n",
              "903  chrX  154365712         6   T   C     0.667          4\n",
              "904  chrX  154366167       221   G   A     0.982        217\n",
              "905  chrX  154368109         7   C   G     0.286          2\n",
              "906  chrX  154368228        10   G   A     0.200          2\n",
              "\n",
              "[907 rows x 7 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_WT editing sites (raw):\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chr</th>\n",
              "      <th>pos</th>\n",
              "      <th>coverage</th>\n",
              "      <th>ref</th>\n",
              "      <th>alt</th>\n",
              "      <th>var_freq</th>\n",
              "      <th>alt_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16974730</td>\n",
              "      <td>2</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>1.000</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975108</td>\n",
              "      <td>2</td>\n",
              "      <td>T</td>\n",
              "      <td>G</td>\n",
              "      <td>1.000</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975285</td>\n",
              "      <td>107</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>1.000</td>\n",
              "      <td>107</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16976221</td>\n",
              "      <td>16</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>0.125</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16977094</td>\n",
              "      <td>4</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>1.000</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1114</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365712</td>\n",
              "      <td>4</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>0.500</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1115</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365722</td>\n",
              "      <td>4</td>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>0.500</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1116</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154366167</td>\n",
              "      <td>211</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>1.000</td>\n",
              "      <td>211</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1117</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368266</td>\n",
              "      <td>7</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.286</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1118</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368348</td>\n",
              "      <td>5</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.400</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1119 rows × 7 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "       chr        pos  coverage ref alt  var_freq  alt_count\n",
              "0     chr1   16974730         2   T   C     1.000          2\n",
              "1     chr1   16975108         2   T   G     1.000          2\n",
              "2     chr1   16975285       107   A   G     1.000        107\n",
              "3     chr1   16976221        16   A   G     0.125          2\n",
              "4     chr1   16977094         4   C   T     1.000          4\n",
              "...    ...        ...       ...  ..  ..       ...        ...\n",
              "1114  chrX  154365712         4   T   C     0.500          2\n",
              "1115  chrX  154365722         4   A   C     0.500          2\n",
              "1116  chrX  154366167       211   G   A     1.000        211\n",
              "1117  chrX  154368266         7   G   A     0.286          2\n",
              "1118  chrX  154368348         5   G   A     0.400          2\n",
              "\n",
              "[1119 rows x 7 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# invoke the function 'get_nonsingleton_mismatch_df' on ADAR_KO and ADAR_WT pileup data to get editing sites\n",
        "ADAR_KO_pileup = os.path.join(PILEUP_RAW_DIR, 'ADAR_KO_pileup.txt')\n",
        "ADAR_WT_pileup = os.path.join(PILEUP_RAW_DIR, 'ADAR_WT_pileup.txt')\n",
        "pileup_df_KO = pd.read_csv(ADAR_KO_pileup, sep='\\t', header=None,\n",
        "                         names=['chr','pos','ref','coverage','bases','quality'])\n",
        "pileup_df_WT = pd.read_csv(ADAR_WT_pileup, sep='\\t', header=None,\n",
        "                            names=['chr','pos','ref','coverage','bases','quality'])\n",
        "\n",
        "ADAR_KO_editing_sites_raw = get_nonsingleton_mismatch_df(pileup_df_KO)\n",
        "ADAR_WT_editing_sites_raw = get_nonsingleton_mismatch_df(pileup_df_WT)\n",
        "\n",
        "print(\"ADAR_KO editing sites (raw):\")\n",
        "display(ADAR_KO_editing_sites_raw)\n",
        "print(\"ADAR_WT editing sites (raw):\")\n",
        "display(ADAR_WT_editing_sites_raw)\n",
        "\n",
        "# save results\n",
        "ADAR_KO_editing_sites_raw.to_csv(os.path.join(EDITING_DIR, 'ADAR_KO_editing_sites_raw.txt'), sep='\\t', index=False)\n",
        "ADAR_WT_editing_sites_raw.to_csv(os.path.join(EDITING_DIR, 'ADAR_WT_editing_sites_raw.txt'), sep='\\t', index=False)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OMOh5Kk3teRZ"
      },
      "source": [
        "##### 2） require homozygous DNA"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "f_2soHphteRZ",
        "outputId": "9eb25ba6-f67a-4e10-8086-73d812b7e27e"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_WT variantsites in homozygous DNA sites:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chr</th>\n",
              "      <th>pos</th>\n",
              "      <th>coverage</th>\n",
              "      <th>ref</th>\n",
              "      <th>alt</th>\n",
              "      <th>var_freq</th>\n",
              "      <th>alt_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16974730</td>\n",
              "      <td>2</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>1.000</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975108</td>\n",
              "      <td>2</td>\n",
              "      <td>T</td>\n",
              "      <td>G</td>\n",
              "      <td>1.000</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975285</td>\n",
              "      <td>107</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>1.000</td>\n",
              "      <td>107</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16976221</td>\n",
              "      <td>16</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>0.125</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16977094</td>\n",
              "      <td>4</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>1.000</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1114</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365712</td>\n",
              "      <td>4</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>0.500</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1115</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365722</td>\n",
              "      <td>4</td>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>0.500</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1116</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154366167</td>\n",
              "      <td>211</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>1.000</td>\n",
              "      <td>211</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1117</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368266</td>\n",
              "      <td>7</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.286</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1118</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368348</td>\n",
              "      <td>5</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.400</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1119 rows × 7 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "       chr        pos  coverage ref alt  var_freq  alt_count\n",
              "0     chr1   16974730         2   T   C     1.000          2\n",
              "1     chr1   16975108         2   T   G     1.000          2\n",
              "2     chr1   16975285       107   A   G     1.000        107\n",
              "3     chr1   16976221        16   A   G     0.125          2\n",
              "4     chr1   16977094         4   C   T     1.000          4\n",
              "...    ...        ...       ...  ..  ..       ...        ...\n",
              "1114  chrX  154365712         4   T   C     0.500          2\n",
              "1115  chrX  154365722         4   A   C     0.500          2\n",
              "1116  chrX  154366167       211   G   A     1.000        211\n",
              "1117  chrX  154368266         7   G   A     0.286          2\n",
              "1118  chrX  154368348         5   G   A     0.400          2\n",
              "\n",
              "[1119 rows x 7 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_KO variantsites in homozygous DNA sites:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chr</th>\n",
              "      <th>pos</th>\n",
              "      <th>coverage</th>\n",
              "      <th>ref</th>\n",
              "      <th>alt</th>\n",
              "      <th>var_freq</th>\n",
              "      <th>alt_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16974964</td>\n",
              "      <td>72</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.028</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975285</td>\n",
              "      <td>121</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>1.000</td>\n",
              "      <td>121</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975628</td>\n",
              "      <td>6</td>\n",
              "      <td>C</td>\n",
              "      <td>A</td>\n",
              "      <td>0.333</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16976085</td>\n",
              "      <td>26</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.077</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16977094</td>\n",
              "      <td>3</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>1.000</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>902</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154357212</td>\n",
              "      <td>31</td>\n",
              "      <td>C</td>\n",
              "      <td>G</td>\n",
              "      <td>0.065</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>903</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365712</td>\n",
              "      <td>6</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>0.667</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>904</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154366167</td>\n",
              "      <td>221</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.982</td>\n",
              "      <td>217</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>905</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368109</td>\n",
              "      <td>7</td>\n",
              "      <td>C</td>\n",
              "      <td>G</td>\n",
              "      <td>0.286</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>906</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368228</td>\n",
              "      <td>10</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.200</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>907 rows × 7 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      chr        pos  coverage ref alt  var_freq  alt_count\n",
              "0    chr1   16974964        72   G   A     0.028          2\n",
              "1    chr1   16975285       121   A   G     1.000        121\n",
              "2    chr1   16975628         6   C   A     0.333          2\n",
              "3    chr1   16976085        26   G   A     0.077          2\n",
              "4    chr1   16977094         3   C   T     1.000          3\n",
              "..    ...        ...       ...  ..  ..       ...        ...\n",
              "902  chrX  154357212        31   C   G     0.065          2\n",
              "903  chrX  154365712         6   T   C     0.667          4\n",
              "904  chrX  154366167       221   G   A     0.982        217\n",
              "905  chrX  154368109         7   C   G     0.286          2\n",
              "906  chrX  154368228        10   G   A     0.200          2\n",
              "\n",
              "[907 rows x 7 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# filter variant sites in homozygous DNA sites\n",
        "\n",
        "def get_variantsites_in_homozygous_sites(variant_df, homozygous_df):\n",
        "    merged_df = variant_df.merge(homozygous_df, on = ['chr', 'pos', 'ref'], how = 'left')\n",
        "    columns = variant_df.columns\n",
        "    new_variant_df = variant_df[columns]\n",
        "    new_variant_df.drop_duplicates(inplace=True)\n",
        "    return new_variant_df\n",
        "\n",
        "df_ADAR_WT_variantsites_homozygous = get_variantsites_in_homozygous_sites(ADAR_WT_editing_sites_raw, homozygous_DNA_sites_df)\n",
        "df_ADAR_KO_variantsites_homozygous = get_variantsites_in_homozygous_sites(ADAR_KO_editing_sites_raw, homozygous_DNA_sites_df)\n",
        "\n",
        "print(\"ADAR_WT variantsites in homozygous DNA sites:\")\n",
        "display(df_ADAR_WT_variantsites_homozygous)\n",
        "print(\"ADAR_KO variantsites in homozygous DNA sites:\")\n",
        "display(df_ADAR_KO_variantsites_homozygous)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zyvoM_frteRZ"
      },
      "source": [
        "##### 3） remove SNP"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "faZ3Br-mteRa",
        "outputId": "669eed13-81d9-4ce3-a4fc-69ee936aac21"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_WT variant sites in BED format:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16974729</td>\n",
              "      <td>16974730</td>\n",
              "      <td>T@C@2@2@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975107</td>\n",
              "      <td>16975108</td>\n",
              "      <td>T@G@2@2@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975284</td>\n",
              "      <td>16975285</td>\n",
              "      <td>A@G@107@107@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16976220</td>\n",
              "      <td>16976221</td>\n",
              "      <td>A@G@16@2@0.125</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16977093</td>\n",
              "      <td>16977094</td>\n",
              "      <td>C@T@4@4@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1114</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365711</td>\n",
              "      <td>154365712</td>\n",
              "      <td>T@C@4@2@0.5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1115</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365721</td>\n",
              "      <td>154365722</td>\n",
              "      <td>A@C@4@2@0.5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1116</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154366166</td>\n",
              "      <td>154366167</td>\n",
              "      <td>G@A@211@211@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1117</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368265</td>\n",
              "      <td>154368266</td>\n",
              "      <td>G@A@7@2@0.286</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1118</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368347</td>\n",
              "      <td>154368348</td>\n",
              "      <td>G@A@5@2@0.4</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1119 rows × 4 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     chrom      start        end             name\n",
              "0     chr1   16974729   16974730      T@C@2@2@1.0\n",
              "1     chr1   16975107   16975108      T@G@2@2@1.0\n",
              "2     chr1   16975284   16975285  A@G@107@107@1.0\n",
              "3     chr1   16976220   16976221   A@G@16@2@0.125\n",
              "4     chr1   16977093   16977094      C@T@4@4@1.0\n",
              "...    ...        ...        ...              ...\n",
              "1114  chrX  154365711  154365712      T@C@4@2@0.5\n",
              "1115  chrX  154365721  154365722      A@C@4@2@0.5\n",
              "1116  chrX  154366166  154366167  G@A@211@211@1.0\n",
              "1117  chrX  154368265  154368266    G@A@7@2@0.286\n",
              "1118  chrX  154368347  154368348      G@A@5@2@0.4\n",
              "\n",
              "[1119 rows x 4 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_KO variant sites in BED format:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16974963</td>\n",
              "      <td>16974964</td>\n",
              "      <td>G@A@72@2@0.028</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975284</td>\n",
              "      <td>16975285</td>\n",
              "      <td>A@G@121@121@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975627</td>\n",
              "      <td>16975628</td>\n",
              "      <td>C@A@6@2@0.333</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16976084</td>\n",
              "      <td>16976085</td>\n",
              "      <td>G@A@26@2@0.077</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16977093</td>\n",
              "      <td>16977094</td>\n",
              "      <td>C@T@3@3@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>902</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154357211</td>\n",
              "      <td>154357212</td>\n",
              "      <td>C@G@31@2@0.065</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>903</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365711</td>\n",
              "      <td>154365712</td>\n",
              "      <td>T@C@6@4@0.667</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>904</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154366166</td>\n",
              "      <td>154366167</td>\n",
              "      <td>G@A@221@217@0.982</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>905</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368108</td>\n",
              "      <td>154368109</td>\n",
              "      <td>C@G@7@2@0.286</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>906</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368227</td>\n",
              "      <td>154368228</td>\n",
              "      <td>G@A@10@2@0.2</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>907 rows × 4 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "    chrom      start        end               name\n",
              "0    chr1   16974963   16974964     G@A@72@2@0.028\n",
              "1    chr1   16975284   16975285    A@G@121@121@1.0\n",
              "2    chr1   16975627   16975628      C@A@6@2@0.333\n",
              "3    chr1   16976084   16976085     G@A@26@2@0.077\n",
              "4    chr1   16977093   16977094        C@T@3@3@1.0\n",
              "..    ...        ...        ...                ...\n",
              "902  chrX  154357211  154357212     C@G@31@2@0.065\n",
              "903  chrX  154365711  154365712      T@C@6@4@0.667\n",
              "904  chrX  154366166  154366167  G@A@221@217@0.982\n",
              "905  chrX  154368108  154368109      C@G@7@2@0.286\n",
              "906  chrX  154368227  154368228       G@A@10@2@0.2\n",
              "\n",
              "[907 rows x 4 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# convert the candidate sites into a BED file\n",
        "def get_variant_df_bed_format(variant_df):\n",
        "    bed_records = []\n",
        "    for _, row in variant_df.iterrows():\n",
        "        chrom = row['chr']\n",
        "        start = row['pos'] - 1  # BED format is 0-based\n",
        "        end = row['pos']        # BED format end is exclusive\n",
        "        name = '@'.join([str(row['ref']), str(row['alt']), str(row['coverage']), str(row['alt_count']), str(row['var_freq'])])\n",
        "        bed_records.append([chrom, start, end, name])\n",
        "    bed_df = pd.DataFrame(bed_records, columns=['chrom', 'start', 'end', 'name'])\n",
        "    return bed_df\n",
        "\n",
        "bed_df_ADAR_WT = get_variant_df_bed_format(df_ADAR_WT_variantsites_homozygous)\n",
        "bed_df_ADAR_KO = get_variant_df_bed_format(df_ADAR_KO_variantsites_homozygous)\n",
        "\n",
        "print(\"ADAR_WT variant sites in BED format:\")\n",
        "display(bed_df_ADAR_WT)\n",
        "\n",
        "print(\"ADAR_KO variant sites in BED format:\")\n",
        "display(bed_df_ADAR_KO)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_uS4gZAQteRa"
      },
      "outputs": [],
      "source": [
        "def remove_SNP(variant_df, snp_bed):\n",
        "\n",
        "    # load SNP bed\n",
        "    variant_bed = pybedtools.BedTool.from_dataframe(variant_df[['chrom','start','end','name']])\n",
        "\n",
        "    # bedtools intersect to exclude SNP sites\n",
        "    variant_sites_noSNP_bed = variant_bed.intersect(snp_bed, v=True) # v=True: exclude overlapping\n",
        "    variant_sites_noSNP_df = variant_sites_noSNP_bed.to_dataframe(names=['chrom','start','end','name'])\n",
        "\n",
        "    return variant_sites_noSNP_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "GE3KNTLiteRa",
        "outputId": "0818c988-9827-4c20-f47a-ae3714f0e35b"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Number of ADAR_WT variant sites before removing SNPs: 1119, after removing SNPs: 835\n",
            "Number of ADAR_KO variant sites before removing SNPs: 907, after removing SNPs: 599\n",
            "\n",
            "ADAR_WT variant sites after removing SNPs:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16974729</td>\n",
              "      <td>16974730</td>\n",
              "      <td>T@C@2@2@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975107</td>\n",
              "      <td>16975108</td>\n",
              "      <td>T@G@2@2@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16976220</td>\n",
              "      <td>16976221</td>\n",
              "      <td>A@G@16@2@0.125</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16977636</td>\n",
              "      <td>16977637</td>\n",
              "      <td>T@G@2@2@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16977787</td>\n",
              "      <td>16977788</td>\n",
              "      <td>T@G@2@2@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>830</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365622</td>\n",
              "      <td>154365623</td>\n",
              "      <td>A@G@5@2@0.4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>831</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365711</td>\n",
              "      <td>154365712</td>\n",
              "      <td>T@C@4@2@0.5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>832</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365721</td>\n",
              "      <td>154365722</td>\n",
              "      <td>A@C@4@2@0.5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>833</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368265</td>\n",
              "      <td>154368266</td>\n",
              "      <td>G@A@7@2@0.286</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>834</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368347</td>\n",
              "      <td>154368348</td>\n",
              "      <td>G@A@5@2@0.4</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>835 rows × 4 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "    chrom      start        end            name\n",
              "0    chr1   16974729   16974730     T@C@2@2@1.0\n",
              "1    chr1   16975107   16975108     T@G@2@2@1.0\n",
              "2    chr1   16976220   16976221  A@G@16@2@0.125\n",
              "3    chr1   16977636   16977637     T@G@2@2@1.0\n",
              "4    chr1   16977787   16977788     T@G@2@2@1.0\n",
              "..    ...        ...        ...             ...\n",
              "830  chrX  154365622  154365623     A@G@5@2@0.4\n",
              "831  chrX  154365711  154365712     T@C@4@2@0.5\n",
              "832  chrX  154365721  154365722     A@C@4@2@0.5\n",
              "833  chrX  154368265  154368266   G@A@7@2@0.286\n",
              "834  chrX  154368347  154368348     G@A@5@2@0.4\n",
              "\n",
              "[835 rows x 4 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_KO variant sites after removing SNPs:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16974963</td>\n",
              "      <td>16974964</td>\n",
              "      <td>G@A@72@2@0.028</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16975627</td>\n",
              "      <td>16975628</td>\n",
              "      <td>C@A@6@2@0.333</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16976084</td>\n",
              "      <td>16976085</td>\n",
              "      <td>G@A@26@2@0.077</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>16977442</td>\n",
              "      <td>16977443</td>\n",
              "      <td>A@T@4@2@0.5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>115815227</td>\n",
              "      <td>115815228</td>\n",
              "      <td>G@A@2@2@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>594</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154356863</td>\n",
              "      <td>154356864</td>\n",
              "      <td>T@C@25@2@0.08</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>595</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154356959</td>\n",
              "      <td>154356960</td>\n",
              "      <td>T@A@14@2@0.143</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>596</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154357211</td>\n",
              "      <td>154357212</td>\n",
              "      <td>C@G@31@2@0.065</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>597</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154365711</td>\n",
              "      <td>154365712</td>\n",
              "      <td>T@C@6@4@0.667</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>598</th>\n",
              "      <td>chrX</td>\n",
              "      <td>154368227</td>\n",
              "      <td>154368228</td>\n",
              "      <td>G@A@10@2@0.2</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>599 rows × 4 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "    chrom      start        end            name\n",
              "0    chr1   16974963   16974964  G@A@72@2@0.028\n",
              "1    chr1   16975627   16975628   C@A@6@2@0.333\n",
              "2    chr1   16976084   16976085  G@A@26@2@0.077\n",
              "3    chr1   16977442   16977443     A@T@4@2@0.5\n",
              "4    chr1  115815227  115815228     G@A@2@2@1.0\n",
              "..    ...        ...        ...             ...\n",
              "594  chrX  154356863  154356864   T@C@25@2@0.08\n",
              "595  chrX  154356959  154356960  T@A@14@2@0.143\n",
              "596  chrX  154357211  154357212  C@G@31@2@0.065\n",
              "597  chrX  154365711  154365712   T@C@6@4@0.667\n",
              "598  chrX  154368227  154368228    G@A@10@2@0.2\n",
              "\n",
              "[599 rows x 4 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# invoke 'remove_SNP' function on ADAR_WT and ADAR_KO variant bed data\n",
        "snp_bed_path = pybedtools.BedTool('./data/reference/hg38_SNP.bed')\n",
        "\n",
        "ADAR_WT_variantsites_noSNP_df = remove_SNP(bed_df_ADAR_WT, snp_bed_path)\n",
        "ADAR_KO_variantsites_noSNP_df = remove_SNP(bed_df_ADAR_KO, snp_bed_path)\n",
        "\n",
        "# number: before and after removing SNPs\n",
        "print(\"Number of ADAR_WT variant sites before removing SNPs: %s, after removing SNPs: %s\" % (len(bed_df_ADAR_WT), len(ADAR_WT_variantsites_noSNP_df)))\n",
        "print(\"Number of ADAR_KO variant sites before removing SNPs: %s, after removing SNPs: %s\" % (len(bed_df_ADAR_KO), len(ADAR_KO_variantsites_noSNP_df)))\n",
        "print()\n",
        "\n",
        "print(\"ADAR_WT variant sites after removing SNPs:\")\n",
        "display(ADAR_WT_variantsites_noSNP_df)\n",
        "\n",
        "print(\"ADAR_KO variant sites after removing SNPs:\")\n",
        "display(ADAR_KO_variantsites_noSNP_df)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VGgSPJMdteRa"
      },
      "source": [
        "##### 4） Separate Alu and non-Alu mismatches"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "k0v_BfumteRa"
      },
      "outputs": [],
      "source": [
        "def saparate_Alu_nonAlu_sites(variant_df, alu_bed):\n",
        "    # load variant bed\n",
        "    variant_bed = pybedtools.BedTool.from_dataframe(variant_df[['chrom','start','end','name']])\n",
        "\n",
        "    # bedtools intersect to get Alu sites\n",
        "    variant_sites_Alu_bed = variant_bed.intersect(alu_bed, u=True) # u=True: report only those that have overlaps\n",
        "    variant_sites_Alu_df = variant_sites_Alu_bed.to_dataframe(names=['chrom','start','end','name'])\n",
        "\n",
        "    # bedtools intersect to get non-Alu sites\n",
        "    variant_sites_nonAlu_bed = variant_bed.intersect(alu_bed, v=True) # v=True: exclude overlapping\n",
        "    variant_sites_nonAlu_df = variant_sites_nonAlu_bed.to_dataframe(names=['chrom','start','end','name'])\n",
        "\n",
        "    return variant_sites_Alu_df, variant_sites_nonAlu_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "e04TeuXmteRb",
        "outputId": "87dcf446-d09e-4097-9582-05ad5dd607b2"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_WT variant sites in Alu regions: 241, in non-Alu regions: 594\n",
            "ADAR_KO variant sites in Alu regions: 61, in non-Alu regions: 538\n"
          ]
        }
      ],
      "source": [
        "# invoke 'saparate_Alu_nonAlu_sites' function on ADAR_WT and ADAR_KO variant bed data after removing SNPs\n",
        "\n",
        "Alu_bed_path = pybedtools.BedTool('./data/reference/hg38_Alu.bed')\n",
        "ADAR_WT_variantsites_Alu, ADAR_WT_variantsites_nonAlu = saparate_Alu_nonAlu_sites(ADAR_WT_variantsites_noSNP_df, Alu_bed_path)\n",
        "ADAR_KO_variantsites_Alu, ADAR_KO_variantsites_nonAlu = saparate_Alu_nonAlu_sites(ADAR_KO_variantsites_noSNP_df, Alu_bed_path)\n",
        "\n",
        "# number: Alu and non-Alu sites\n",
        "print(\"ADAR_WT variant sites in Alu regions: %s, in non-Alu regions: %s\" % (len(ADAR_WT_variantsites_Alu), len(ADAR_WT_variantsites_nonAlu)))\n",
        "print(\"ADAR_KO variant sites in Alu regions: %s, in non-Alu regions: %s\" % (len(ADAR_KO_variantsites_Alu), len(ADAR_KO_variantsites_nonAlu)))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "9DgAh4u-teRb"
      },
      "outputs": [],
      "source": [
        "# save\n",
        "ADAR_WT_variantsites_Alu.to_csv(os.path.join(EDITING_DIR, 'ADAR_WT_Alu_sites.txt'), sep='\\t', header=False, index=False)\n",
        "ADAR_KO_variantsites_Alu.to_csv(os.path.join(EDITING_DIR, 'ADAR_KO_Alu_sites.txt'), sep='\\t', header=False, index=False)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mWNMO6qWteRb"
      },
      "source": [
        "#### Filters-2 -> for 'non-Alu mismatches' : get 'Repetitive non-Alu editing sites' and 'Nonrepetitive editing sites'\n",
        "        1)\n",
        "        min_variant_count >=3; min_variant_frequency >= 0.1\n",
        "        2)\n",
        "        remove sites in simple repeats\n",
        "        3)\n",
        "        Filter intronic sites within 4 bp of splice junctions\n",
        "        4)\n",
        "        remove sites in homopolymer runs of >= 5bp\n",
        "        5)\n",
        "        BLAT reads with mismatch against reference genome to ensure unique mapping\n",
        "#### Data\n",
        "        1)  Non-Alu mismatches data\n",
        "        2)  required filter files:\n",
        "                        1 - NonAlu-repeat region data\n",
        "                        2 - simple repeat region data (data from repeatmasker)\n",
        "                        3 - splice junctions region data (get from GTF)\n",
        "                        4 - reference fa filte - to BLAT\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9t_pFG_bteRb"
      },
      "source": [
        "#### 1 - prepare the splice junctions data、 simple repeat data、 NonAlu-repeat region data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "aVEqzlpJteRb",
        "outputId": "bff34b25-55af-489d-b5fa-47dcb6948c5d"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Splice site regions with 4bp flanking sequences:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>11116</td>\n",
              "      <td>11124</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>11207</td>\n",
              "      <td>11215</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>12005</td>\n",
              "      <td>12013</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>12223</td>\n",
              "      <td>12231</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>12608</td>\n",
              "      <td>12616</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7347893</th>\n",
              "      <td>chrM</td>\n",
              "      <td>15883</td>\n",
              "      <td>15891</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7347894</th>\n",
              "      <td>chrM</td>\n",
              "      <td>15883</td>\n",
              "      <td>15891</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7347895</th>\n",
              "      <td>chrM</td>\n",
              "      <td>15949</td>\n",
              "      <td>15957</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7347896</th>\n",
              "      <td>chrM</td>\n",
              "      <td>15951</td>\n",
              "      <td>15959</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7347897</th>\n",
              "      <td>chrM</td>\n",
              "      <td>16019</td>\n",
              "      <td>16027</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>7347898 rows × 3 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "        chrom  start    end\n",
              "0        chr1  11116  11124\n",
              "1        chr1  11207  11215\n",
              "2        chr1  12005  12013\n",
              "3        chr1  12223  12231\n",
              "4        chr1  12608  12616\n",
              "...       ...    ...    ...\n",
              "7347893  chrM  15883  15891\n",
              "7347894  chrM  15883  15891\n",
              "7347895  chrM  15949  15957\n",
              "7347896  chrM  15951  15959\n",
              "7347897  chrM  16019  16027\n",
              "\n",
              "[7347898 rows x 3 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "def generate_splice_site_bed(gtf_file, flank=4):\n",
        "    # 1. read GTF file and extract exon features\n",
        "    gtf_df = pd.read_csv(\n",
        "        gtf_file, sep='\\t', comment='#', header=None,\n",
        "        names=['chr','source','feature','start','end','score','strand','frame','attribute']\n",
        "    )\n",
        "    exons = gtf_df[gtf_df['feature'] == 'exon'].copy()\n",
        "\n",
        "    # BED is 0-based, half-open\n",
        "    exons['start0'] = exons['start'] - 1\n",
        "\n",
        "    # 2. generate splice site regions with flanking sequences\n",
        "    splice_regions = []\n",
        "    for _, row in exons.iterrows():\n",
        "        # 5' splice site\n",
        "        splice_regions.append([row['chr'], max(0,row['start0'] - flank), row['start0'] + flank])\n",
        "        # 3' splice site\n",
        "        splice_regions.append([row['chr'], max(0,row['end'] - flank), row['end'] + flank])\n",
        "\n",
        "    splice_bed = pd.DataFrame(splice_regions, columns=['chrom','start','end'])\n",
        "    splice_bed.to_csv(f'./data/reference/hg38_splice_sites_flank{flank}bp.bed', sep='\\t', index=False, header=False)\n",
        "\n",
        "    return splice_bed\n",
        "\n",
        "# generate splice site bed file with 4bp flanking sequences\n",
        "splice_bed = generate_splice_site_bed(GTF, flank=4)\n",
        "print(\"Splice site regions with 4bp flanking sequences:\")\n",
        "display(splice_bed)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ob5Iz-FmteRg",
        "outputId": "334d7f7c-2ba7-4e38-cc78-c7e887afe820"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Simple repeat region：\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>12582784</td>\n",
              "      <td>12583841</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>12582839</td>\n",
              "      <td>12582993</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>12582910</td>\n",
              "      <td>12583333</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>22020082</td>\n",
              "      <td>22020671</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>121609878</td>\n",
              "      <td>121734316</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1049710</th>\n",
              "      <td>chr22_KI270739v1_random</td>\n",
              "      <td>42428</td>\n",
              "      <td>42470</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1049711</th>\n",
              "      <td>chr22_KI270739v1_random</td>\n",
              "      <td>42529</td>\n",
              "      <td>60923</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1049712</th>\n",
              "      <td>chr22_KI270739v1_random</td>\n",
              "      <td>42644</td>\n",
              "      <td>60923</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1049713</th>\n",
              "      <td>chr22_KI270739v1_random</td>\n",
              "      <td>50068</td>\n",
              "      <td>61010</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1049714</th>\n",
              "      <td>chr22_KI270739v1_random</td>\n",
              "      <td>61202</td>\n",
              "      <td>73979</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1049715 rows × 3 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                           chrom      start        end\n",
              "0                           chr1   12582784   12583841\n",
              "1                           chr1   12582839   12582993\n",
              "2                           chr1   12582910   12583333\n",
              "3                           chr1   22020082   22020671\n",
              "4                           chr1  121609878  121734316\n",
              "...                          ...        ...        ...\n",
              "1049710  chr22_KI270739v1_random      42428      42470\n",
              "1049711  chr22_KI270739v1_random      42529      60923\n",
              "1049712  chr22_KI270739v1_random      42644      60923\n",
              "1049713  chr22_KI270739v1_random      50068      61010\n",
              "1049714  chr22_KI270739v1_random      61202      73979\n",
              "\n",
              "[1049715 rows x 3 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Non-Alu repeat region：\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>67108753</td>\n",
              "      <td>67109046</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>25165803</td>\n",
              "      <td>25166380</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>50331336</td>\n",
              "      <td>50332274</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>58719764</td>\n",
              "      <td>58720546</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>75496057</td>\n",
              "      <td>75497775</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4400864</th>\n",
              "      <td>chr22_KI270739v1_random</td>\n",
              "      <td>42471</td>\n",
              "      <td>42520</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4400865</th>\n",
              "      <td>chr22_KI270739v1_random</td>\n",
              "      <td>42530</td>\n",
              "      <td>48656</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4400866</th>\n",
              "      <td>chr22_KI270739v1_random</td>\n",
              "      <td>48656</td>\n",
              "      <td>49096</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4400867</th>\n",
              "      <td>chr22_KI270739v1_random</td>\n",
              "      <td>49091</td>\n",
              "      <td>50070</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4400868</th>\n",
              "      <td>chr22_KI270739v1_random</td>\n",
              "      <td>50070</td>\n",
              "      <td>73985</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4400869 rows × 3 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                           chrom     start       end\n",
              "0                           chr1  67108753  67109046\n",
              "1                           chr1  25165803  25166380\n",
              "2                           chr1  50331336  50332274\n",
              "3                           chr1  58719764  58720546\n",
              "4                           chr1  75496057  75497775\n",
              "...                          ...       ...       ...\n",
              "4400864  chr22_KI270739v1_random     42471     42520\n",
              "4400865  chr22_KI270739v1_random     42530     48656\n",
              "4400866  chr22_KI270739v1_random     48656     49096\n",
              "4400867  chr22_KI270739v1_random     49091     50070\n",
              "4400868  chr22_KI270739v1_random     50070     73985\n",
              "\n",
              "[4400869 rows x 3 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# show simple repeat region data\n",
        "\n",
        "SimpleRepeat_df = pd.read_csv('./data/reference/hg38_simple_repeat.bed', sep = '\\t', header = None, names=['chrom', 'start', 'end'])\n",
        "print('Simple repeat region：')\n",
        "display(SimpleRepeat_df)\n",
        "\n",
        "nonAlu_repeat_df = pd.read_csv('./data/reference/hg38_nonAlu_repeat.bed', sep = '\\t', header = None, names=['chrom', 'start', 'end'])\n",
        "print('Non-Alu repeat region：')\n",
        "display(nonAlu_repeat_df)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LmrjB-pZteRg"
      },
      "source": [
        "#### 2 - Filter:\n",
        "        1)\n",
        "        more stringent variant call criteria:min_variant_count >=3; min_variant_frequency >= 0.1\n",
        "        2)\n",
        "        remove sites in simple repeats\n",
        "        3)\n",
        "        Filter intronic sites within 4 bp of splice junctions\n",
        "        4)\n",
        "        remove sites in homopolymer runs of >= 5bp\n",
        "        5)\n",
        "        BLAT reads with mismatch against reference genome to ensure unique mapping"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QoRtT-BPteRh"
      },
      "source": [
        "##### 1） more stringent variant call criteria:min_variant_count >=3; min_variant_frequency >= 0.1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ITpLw9OJteRh"
      },
      "outputs": [],
      "source": [
        "def stringent_criteria_filter(variant_df, min_variant_count=3, min_variant_frequency=0.1):\n",
        "    variant_df['var_count'] = variant_df['name'].str.split('@').str[3].astype(int)\n",
        "    variant_df['varFreq'] = variant_df['name'].str.split('@').str[4].astype(float)\n",
        "    variant_df = variant_df[\n",
        "        (variant_df['var_count'] >= 3) &\n",
        "        (variant_df['varFreq'] >= 0.1)\n",
        "    ]\n",
        "\n",
        "    variant_df = variant_df.drop(columns=['var_count', 'varFreq'])\n",
        "    return variant_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ra06lM1-teRh",
        "outputId": "c727eeef-e533-4d74-ee7a-67416458a0f0"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_WT non-Alu variant sites before stringent filtering: 594, after stringent filtering: 30\n",
            "ADAR_KO non-Alu variant sites before stringent filtering: 538, after stringent filtering: 19\n",
            "ADAR_WT non-Alu variant sites after stringent filtering:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144421987</td>\n",
              "      <td>144421988</td>\n",
              "      <td>T@C@30@4@0.133</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144426380</td>\n",
              "      <td>144426381</td>\n",
              "      <td>C@T@8@8@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>42</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144447182</td>\n",
              "      <td>144447183</td>\n",
              "      <td>G@A@7@4@0.571</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>60</th>\n",
              "      <td>chr1</td>\n",
              "      <td>160317393</td>\n",
              "      <td>160317394</td>\n",
              "      <td>T@C@5@4@0.8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>72</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100518</td>\n",
              "      <td>37100519</td>\n",
              "      <td>T@C@67@21@0.313</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   chrom      start        end             name\n",
              "20  chr1  144421987  144421988   T@C@30@4@0.133\n",
              "27  chr1  144426380  144426381      C@T@8@8@1.0\n",
              "42  chr1  144447182  144447183    G@A@7@4@0.571\n",
              "60  chr1  160317393  160317394      T@C@5@4@0.8\n",
              "72  chr2   37100518   37100519  T@C@67@21@0.313"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_KO non-Alu variant sites after stringent filtering:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144426380</td>\n",
              "      <td>144426381</td>\n",
              "      <td>C@T@5@5@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>71</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100842</td>\n",
              "      <td>37100843</td>\n",
              "      <td>G@A@89@42@0.472</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>139</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90838991</td>\n",
              "      <td>90838992</td>\n",
              "      <td>G@A@41@41@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>141</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90839054</td>\n",
              "      <td>90839055</td>\n",
              "      <td>G@A@44@44@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>142</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90839055</td>\n",
              "      <td>90839056</td>\n",
              "      <td>G@A@44@44@1.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    chrom      start        end             name\n",
              "22   chr1  144426380  144426381      C@T@5@5@1.0\n",
              "71   chr2   37100842   37100843  G@A@89@42@0.472\n",
              "139  chr4   90838991   90838992    G@A@41@41@1.0\n",
              "141  chr4   90839054   90839055    G@A@44@44@1.0\n",
              "142  chr4   90839055   90839056    G@A@44@44@1.0"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# invoke 'stringent_criteria_filter' function on ADAR_WT and ADAR_KO non-Alu variant bed data\n",
        "\n",
        "ADAR_WT_variantsites_nonAlu_stringent = stringent_criteria_filter(ADAR_WT_variantsites_nonAlu)\n",
        "ADAR_KO_variantsites_nonAlu_stringent = stringent_criteria_filter(ADAR_KO_variantsites_nonAlu)\n",
        "\n",
        "# number: before and after stringent criteria filtering\n",
        "print(\"ADAR_WT non-Alu variant sites before stringent filtering: %s, after stringent filtering: %s\" % (len(ADAR_WT_variantsites_nonAlu), len(ADAR_WT_variantsites_nonAlu_stringent)))\n",
        "print(\"ADAR_KO non-Alu variant sites before stringent filtering: %s, after stringent filtering: %s\" % (len(ADAR_KO_variantsites_nonAlu), len(ADAR_KO_variantsites_nonAlu_stringent)))\n",
        "\n",
        "print(\"ADAR_WT non-Alu variant sites after stringent filtering:\")\n",
        "display(ADAR_WT_variantsites_nonAlu_stringent.head())\n",
        "\n",
        "print(\"ADAR_KO non-Alu variant sites after stringent filtering:\")\n",
        "display(ADAR_KO_variantsites_nonAlu_stringent.head())"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cwBrwDRWteRh"
      },
      "source": [
        "##### 2） remove sites in simple repeats"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7FNtsi57teRh"
      },
      "outputs": [],
      "source": [
        "def remove_simple_repeat_regions(nonAlu_variant_df, simple_repeat_bed):\n",
        "    # load variant bed\n",
        "    variant_bed = pybedtools.BedTool.from_dataframe(nonAlu_variant_df[['chrom','start','end','name']])\n",
        "\n",
        "    # bedtools intersect to exclude simple repeat regions\n",
        "    variant_sites_noSimpleRepeat_bed = variant_bed.intersect(simple_repeat_bed, v=True) # v=True: exclude overlapping\n",
        "    variant_sites_noSimpleRepeat_df = variant_sites_noSimpleRepeat_bed.to_dataframe(names=['chrom','start','end','name'])\n",
        "\n",
        "    return variant_sites_noSimpleRepeat_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Xdg9NKQzteRh",
        "outputId": "b5eecfc5-b247-445c-95e7-09a974361fb5"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_WT non-Alu variant sites before removing simple repeat regions: 30, after removing simple repeat regions: 25\n",
            "ADAR_KO non-Alu variant sites before removing simple repeat regions: 19, after removing simple repeat regions: 17\n",
            "ADAR_WT non-Alu variant sites after removing simple repeat regions:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144421987</td>\n",
              "      <td>144421988</td>\n",
              "      <td>T@C@30@4@0.133</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144447182</td>\n",
              "      <td>144447183</td>\n",
              "      <td>G@A@7@4@0.571</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>160317393</td>\n",
              "      <td>160317394</td>\n",
              "      <td>T@C@5@4@0.8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100518</td>\n",
              "      <td>37100519</td>\n",
              "      <td>T@C@67@21@0.313</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100558</td>\n",
              "      <td>37100559</td>\n",
              "      <td>T@C@73@8@0.11</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  chrom      start        end             name\n",
              "0  chr1  144421987  144421988   T@C@30@4@0.133\n",
              "1  chr1  144447182  144447183    G@A@7@4@0.571\n",
              "2  chr1  160317393  160317394      T@C@5@4@0.8\n",
              "3  chr2   37100518   37100519  T@C@67@21@0.313\n",
              "4  chr2   37100558   37100559    T@C@73@8@0.11"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_KO non-Alu variant sites after removing simple repeat regions:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100842</td>\n",
              "      <td>37100843</td>\n",
              "      <td>G@A@89@42@0.472</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90838991</td>\n",
              "      <td>90838992</td>\n",
              "      <td>G@A@41@41@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90839054</td>\n",
              "      <td>90839055</td>\n",
              "      <td>G@A@44@44@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90839055</td>\n",
              "      <td>90839056</td>\n",
              "      <td>G@A@44@44@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90839066</td>\n",
              "      <td>90839067</td>\n",
              "      <td>G@A@42@42@1.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  chrom     start       end             name\n",
              "0  chr2  37100842  37100843  G@A@89@42@0.472\n",
              "1  chr4  90838991  90838992    G@A@41@41@1.0\n",
              "2  chr4  90839054  90839055    G@A@44@44@1.0\n",
              "3  chr4  90839055  90839056    G@A@44@44@1.0\n",
              "4  chr4  90839066  90839067    G@A@42@42@1.0"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# invoke 'remove_simple_repeat_regions' function on ADAR_WT and ADAR_KO non-Alu stringent variant bed data\n",
        "simple_repeat_bed_path = pybedtools.BedTool('./data/reference/hg38_simple_repeat.bed')\n",
        "\n",
        "ADAR_WT_variantsites_nonAlu_noSimpleRepeat = remove_simple_repeat_regions(ADAR_WT_variantsites_nonAlu_stringent, simple_repeat_bed_path)\n",
        "ADAR_KO_variantsites_nonAlu_noSimpleRepeat = remove_simple_repeat_regions(ADAR_KO_variantsites_nonAlu_stringent, simple_repeat_bed_path)\n",
        "\n",
        "# number: before and after removing simple repeat regions\n",
        "print(\"ADAR_WT non-Alu variant sites before removing simple repeat regions: %s, after removing simple repeat regions: %s\" % (len(ADAR_WT_variantsites_nonAlu_stringent), len(ADAR_WT_variantsites_nonAlu_noSimpleRepeat)))\n",
        "print(\"ADAR_KO non-Alu variant sites before removing simple repeat regions: %s, after removing simple repeat regions: %s\" % (len(ADAR_KO_variantsites_nonAlu_stringent), len(ADAR_KO_variantsites_nonAlu_noSimpleRepeat)))\n",
        "\n",
        "print(\"ADAR_WT non-Alu variant sites after removing simple repeat regions:\")\n",
        "display(ADAR_WT_variantsites_nonAlu_noSimpleRepeat.head())\n",
        "print(\"ADAR_KO non-Alu variant sites after removing simple repeat regions:\")\n",
        "display(ADAR_KO_variantsites_nonAlu_noSimpleRepeat.head())"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sg1wgp6_teRi"
      },
      "source": [
        "##### 3） filter intronic sites within 4bp of splice junctions"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "s3bc4vl9teRi"
      },
      "outputs": [],
      "source": [
        "def filter_near_splice_sites(variant_df, splice_bed):\n",
        "    # load variant bed\n",
        "    variant_bed = pybedtools.BedTool.from_dataframe(variant_df[['chrom','start','end','name']])\n",
        "\n",
        "    # bedtools intersect to exclude sites near splice sites\n",
        "    variant_sites_noSplice_bed = variant_bed.intersect(splice_bed, v=True) # v=True: exclude overlapping\n",
        "    variant_sites_noSplice_df = variant_sites_noSplice_bed.to_dataframe(names=['chrom','start','end','name'])\n",
        "\n",
        "    return variant_sites_noSplice_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "gTL3rgHcteRi",
        "outputId": "527cf6d4-31b6-4d38-c08c-a5950a4c0082"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_WT non-Alu variant sites before removing splice site nearby: 25, after removing splice site nearby: 24\n",
            "ADAR_KO non-Alu variant sites before removing splice site nearby: 17, after removing splice site nearby: 16\n",
            "ADAR_WT non-Alu variant sites after removing splice site nearby:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144421987</td>\n",
              "      <td>144421988</td>\n",
              "      <td>T@C@30@4@0.133</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144447182</td>\n",
              "      <td>144447183</td>\n",
              "      <td>G@A@7@4@0.571</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>160317393</td>\n",
              "      <td>160317394</td>\n",
              "      <td>T@C@5@4@0.8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100518</td>\n",
              "      <td>37100519</td>\n",
              "      <td>T@C@67@21@0.313</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100558</td>\n",
              "      <td>37100559</td>\n",
              "      <td>T@C@73@8@0.11</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  chrom      start        end             name\n",
              "0  chr1  144421987  144421988   T@C@30@4@0.133\n",
              "1  chr1  144447182  144447183    G@A@7@4@0.571\n",
              "2  chr1  160317393  160317394      T@C@5@4@0.8\n",
              "3  chr2   37100518   37100519  T@C@67@21@0.313\n",
              "4  chr2   37100558   37100559    T@C@73@8@0.11"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_KO non-Alu variant sites after removing splice site nearby:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100842</td>\n",
              "      <td>37100843</td>\n",
              "      <td>G@A@89@42@0.472</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90838991</td>\n",
              "      <td>90838992</td>\n",
              "      <td>G@A@41@41@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90839054</td>\n",
              "      <td>90839055</td>\n",
              "      <td>G@A@44@44@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90839055</td>\n",
              "      <td>90839056</td>\n",
              "      <td>G@A@44@44@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90839066</td>\n",
              "      <td>90839067</td>\n",
              "      <td>G@A@42@42@1.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  chrom     start       end             name\n",
              "0  chr2  37100842  37100843  G@A@89@42@0.472\n",
              "1  chr4  90838991  90838992    G@A@41@41@1.0\n",
              "2  chr4  90839054  90839055    G@A@44@44@1.0\n",
              "3  chr4  90839055  90839056    G@A@44@44@1.0\n",
              "4  chr4  90839066  90839067    G@A@42@42@1.0"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# invoke 'filter_near_splice_sites' function on ADAR_WT and ADAR_KO non-Alu stringent variant bed data after removing simple repeat regions\n",
        "\n",
        "splice_bed_path = pybedtools.BedTool('./data/reference/hg38_splice_sites_flank4bp.bed')\n",
        "ADAR_WT_nonAlu_rm_splice = filter_near_splice_sites(ADAR_WT_variantsites_nonAlu_noSimpleRepeat, splice_bed_path)\n",
        "ADAR_KO_nonAlu_rm_splice = filter_near_splice_sites(ADAR_KO_variantsites_nonAlu_noSimpleRepeat, splice_bed_path)\n",
        "\n",
        "# number: before and after removing sites near splice sites\n",
        "print(\"ADAR_WT non-Alu variant sites before removing splice site nearby: %s, after removing splice site nearby: %s\" % (len(ADAR_WT_variantsites_nonAlu_noSimpleRepeat), len(ADAR_WT_nonAlu_rm_splice)))\n",
        "print(\"ADAR_KO non-Alu variant sites before removing splice site nearby: %s, after removing splice site nearby: %s\" % (len(ADAR_KO_variantsites_nonAlu_noSimpleRepeat), len(ADAR_KO_nonAlu_rm_splice)))\n",
        "\n",
        "print(\"ADAR_WT non-Alu variant sites after removing splice site nearby:\")\n",
        "display(ADAR_WT_nonAlu_rm_splice.head())\n",
        "print(\"ADAR_KO non-Alu variant sites after removing splice site nearby:\")\n",
        "display(ADAR_KO_nonAlu_rm_splice.head())"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cp_XQw4UteRi"
      },
      "source": [
        "##### 4） remove sites in homopolymer runs of >= 5bp"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "2bE_hKc_teRi"
      },
      "outputs": [],
      "source": [
        "def remove_homoplymer_regions(variant_df, ref_fa):\n",
        "\n",
        "    def remove_homoplymer(candidate_site_bed_row, ref_fa, homo_length):\n",
        "        line = candidate_site_bed_row\n",
        "        chrom = line.iloc[0]\n",
        "        pos = line.iloc[2]   # 1-based\n",
        "\n",
        "        genome = pysam.FastaFile(ref_fa)\n",
        "        up_homo =  genome.fetch( chrom, pos - homo_length - 1, pos - 1 ).upper()\n",
        "        down_homo = genome.fetch( chrom, pos , pos + homo_length ).upper()\n",
        "        inter_homo = up_homo[-int(homo_length/2)-1:] + down_homo[0:int(homo_length/2)+1]\n",
        "\n",
        "        bool_homo = (len(set(up_homo)) == 1) or (len(set(down_homo)) == 1) or (len(set(inter_homo)) == 1)\n",
        "        if bool_homo:\n",
        "            return True\n",
        "        else:\n",
        "            return False\n",
        "\n",
        "    variant_df['isHomopolymer'] = variant_df.apply(lambda x:remove_homoplymer(x, ref_fa, 5), axis=1)\n",
        "    variant_rm_homo_df = variant_df[ variant_df['isHomopolymer'] == False ].iloc[:, :4]\n",
        "\n",
        "    return variant_rm_homo_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Mp-_PovhteRi",
        "outputId": "4af9f178-3ef0-482d-ced2-717fa9316a8b"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_WT non-Alu variant sites before removing homopolymer: 24, after removing homopolymer: 24\n",
            "ADAR_KO non-Alu variant sites before removing homopolymer: 16, after removing homopolymer: 15\n",
            "\n",
            "ADAR_WT non-Alu variant sites after removing homopolymer:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144421987</td>\n",
              "      <td>144421988</td>\n",
              "      <td>T@C@30@4@0.133</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144447182</td>\n",
              "      <td>144447183</td>\n",
              "      <td>G@A@7@4@0.571</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>160317393</td>\n",
              "      <td>160317394</td>\n",
              "      <td>T@C@5@4@0.8</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100518</td>\n",
              "      <td>37100519</td>\n",
              "      <td>T@C@67@21@0.313</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100558</td>\n",
              "      <td>37100559</td>\n",
              "      <td>T@C@73@8@0.11</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  chrom      start        end             name\n",
              "0  chr1  144421987  144421988   T@C@30@4@0.133\n",
              "1  chr1  144447182  144447183    G@A@7@4@0.571\n",
              "2  chr1  160317393  160317394      T@C@5@4@0.8\n",
              "3  chr2   37100518   37100519  T@C@67@21@0.313\n",
              "4  chr2   37100558   37100559    T@C@73@8@0.11"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_KO non-Alu variant sites after removing homopolymer:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100842</td>\n",
              "      <td>37100843</td>\n",
              "      <td>G@A@89@42@0.472</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90838991</td>\n",
              "      <td>90838992</td>\n",
              "      <td>G@A@41@41@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90839054</td>\n",
              "      <td>90839055</td>\n",
              "      <td>G@A@44@44@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90839055</td>\n",
              "      <td>90839056</td>\n",
              "      <td>G@A@44@44@1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr4</td>\n",
              "      <td>90839066</td>\n",
              "      <td>90839067</td>\n",
              "      <td>G@A@42@42@1.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  chrom     start       end             name\n",
              "0  chr2  37100842  37100843  G@A@89@42@0.472\n",
              "1  chr4  90838991  90838992    G@A@41@41@1.0\n",
              "2  chr4  90839054  90839055    G@A@44@44@1.0\n",
              "3  chr4  90839055  90839056    G@A@44@44@1.0\n",
              "4  chr4  90839066  90839067    G@A@42@42@1.0"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# invoke 'remove_homoplymer_regions' function on ADAR_WT and ADAR_KO non-Alu variant bed data\n",
        "\n",
        "ref_fa = './data/reference/hg38.fa'\n",
        "ADAR_WT_nonAlu_rm_homo = remove_homoplymer_regions(ADAR_WT_nonAlu_rm_splice, REF_fa)\n",
        "ADAR_KO_nonAlu_rm_homo = remove_homoplymer_regions(ADAR_KO_nonAlu_rm_splice, REF_fa)\n",
        "\n",
        "# number: before and after removing homopolymer\n",
        "print(\"ADAR_WT non-Alu variant sites before removing homopolymer: %s, after removing homopolymer: %s\" % (len(ADAR_WT_nonAlu_rm_splice), len(ADAR_WT_nonAlu_rm_homo)))\n",
        "print(\"ADAR_KO non-Alu variant sites before removing homopolymer: %s, after removing homopolymer: %s\" % (len(ADAR_KO_nonAlu_rm_splice), len(ADAR_KO_nonAlu_rm_homo)))\n",
        "\n",
        "print()\n",
        "\n",
        "print(\"ADAR_WT non-Alu variant sites after removing homopolymer:\")\n",
        "display(ADAR_WT_nonAlu_rm_homo.head())\n",
        "print(\"ADAR_KO non-Alu variant sites after removing homopolymer:\")\n",
        "display(ADAR_KO_nonAlu_rm_homo.head())\n",
        "\n",
        "# save\n",
        "ADAR_WT_nonAlu_rm_homo.to_csv(os.path.join(EDITING_DIR, 'ADAR_WT_nonAlu_sites.txt'), sep='\\t', header=False, index=False)\n",
        "ADAR_KO_nonAlu_rm_homo.to_csv(os.path.join(EDITING_DIR, 'ADAR_KO_nonAlu_sites.txt'), sep='\\t', header=False, index=False)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bZAkQrwSteRi"
      },
      "source": [
        "##### 5） BLAT reads with mismatch against reference genome to ensure unique mapping\n",
        "            1. Requirement (i): best hit overlaps the candidate site\n",
        "            2. Requirement (ii): second-best hit score < 95% of best hit\n",
        "            3. Keep sites where passing reads > failing reads"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "2ZqCnYXDteRj"
      },
      "outputs": [],
      "source": [
        "def blat_filter_nonAlu_sites(site_file, bam_file, ref_fa, output_file,\n",
        "                                  min_basequal=25, min_mismatch=2, score_limit=0.95,\n",
        "                                  blat_bin=\"./data/blat\"):\n",
        "    \"\"\"\n",
        "    1. Extract reads covering each site from the BAM file\n",
        "    2. Filter reads based on base quality and edited nucleotides\n",
        "    3. Generate the fatmp FASTA file\n",
        "    4. Align reads to the reference genome using BLAT\n",
        "    5. Parse the PSL file and filter reads\n",
        "    6. Count editing events at each site and output the final table\n",
        "    \"\"\"\n",
        "    fatmp_file = Path(site_file + \".fatmp\")\n",
        "    psl_file = Path(site_file + \".psltmp\")\n",
        "\n",
        "    bam = pysam.AlignmentFile(bam_file, \"rb\")\n",
        "\n",
        "    site_df = pd.read_csv(site_file, sep=\"\\t\", header=None, dtype={0:str, 1:int, 2:int, 3:str})\n",
        "\n",
        "    # -----------------------------\n",
        "    # 1. get reads, generate fatmp\n",
        "    # -----------------------------\n",
        "    with open(fatmp_file, \"w\") as fafile:\n",
        "        for idx, row in site_df.iterrows():\n",
        "            chrom = row[0]\n",
        "            pos = int(row[1])\n",
        "            # column 4 format: ref@alt@cov@alt_count@ratio\n",
        "            fields = row[3].split(\"@\")\n",
        "            ref_base = fields[0]\n",
        "            alt_base = fields[1]\n",
        "            tmp_reads_file = f\"{output_file}_tmp\"\n",
        "\n",
        "            reads = list(bam.fetch(chrom, pos-1, pos))\n",
        "            mismatch_readcount = 0\n",
        "\n",
        "            for r in reads:\n",
        "                ref_pos = r.reference_start\n",
        "                seq = r.query_sequence\n",
        "                qual = r.query_qualities\n",
        "                cigar_tuples = r.cigartuples\n",
        "                if cigar_tuples is None:\n",
        "                    continue\n",
        "\n",
        "                ref_cursor = ref_pos\n",
        "                read_cursor = 0\n",
        "                base_match = False\n",
        "\n",
        "                for op, length in cigar_tuples:\n",
        "                    if op in (0, 7, 8):  # M, =, X\n",
        "                        for i in range(length):\n",
        "                            if ref_cursor == pos-1 and seq[read_cursor].upper() == alt_base and qual[read_cursor] >= min_basequal:\n",
        "                                base_match = True\n",
        "                            ref_cursor += 1\n",
        "                            read_cursor += 1\n",
        "                    elif op in (1, 4, 5):  # I, S, H\n",
        "                        read_cursor += length\n",
        "                    elif op in (2, 3):  # D, N\n",
        "                        ref_cursor += length\n",
        "\n",
        "                if base_match:\n",
        "                    fafile.write(f\">{chrom}-{pos}-{mismatch_readcount}\\n{seq}\\n\")\n",
        "                    mismatch_readcount += 1\n",
        "\n",
        "    # -----------------------------\n",
        "    # 2. BLAT\n",
        "    # -----------------------------\n",
        "    subprocess.run(f\"{blat_bin} -stepSize=5 -repMatch=2253 -minScore=20 -minIdentity=0 -noHead {ref_fa} {fatmp_file} {psl_file}\",\n",
        "                   shell=True, check=True)\n",
        "\n",
        "    # -----------------------------\n",
        "    # 3. parse PSL\n",
        "    # -----------------------------\n",
        "    pslhash = {}\n",
        "    with open(psl_file) as f:\n",
        "        for line in f:\n",
        "            fields = line.strip().split()\n",
        "            name = fields[9]\n",
        "            blatscore = \"@\".join([fields[0], fields[13], fields[17], fields[18], fields[20]])\n",
        "            if name in pslhash:\n",
        "                pslhash[name] = f\"{pslhash[name]}-{blatscore}\"\n",
        "            else:\n",
        "                pslhash[name] = blatscore\n",
        "\n",
        "    # -----------------------------\n",
        "    # 4. filter PSL results\n",
        "    # -----------------------------\n",
        "    sitehash = {}\n",
        "    discardhash = {}\n",
        "\n",
        "    for key, val in pslhash.items():\n",
        "        key_parts = key.split(\"-\")\n",
        "        site_key = f\"{key_parts[0]}_{key_parts[1]}\"\n",
        "        psllines = val.split(\"-\")\n",
        "\n",
        "        scores = []\n",
        "        largest_score = 0\n",
        "        largest_line = psllines[0]\n",
        "        for line in psllines:\n",
        "            s = int(line.split(\"@\")[0])\n",
        "            scores.append(s)\n",
        "            if s > largest_score:\n",
        "                largest_line = line\n",
        "                largest_score = s\n",
        "        scores.sort(reverse=True)\n",
        "        second_score = scores[1] if len(scores) > 1 else 0\n",
        "\n",
        "        split_line = largest_line.split(\"@\")\n",
        "        overlap_found = False\n",
        "        if split_line[1] == key_parts[0] and second_score < largest_score * score_limit:\n",
        "            numblocks = int(split_line[2])\n",
        "            blocksizes = list(map(int, split_line[3].strip(\",\").split(\",\")))\n",
        "            blockstarts = list(map(int, split_line[4].strip(\",\").split(\",\")))\n",
        "            for bs, start in zip(blocksizes, blockstarts):\n",
        "                startpos = start + 1\n",
        "                endpos = start + bs\n",
        "                if int(key_parts[1]) >= startpos and int(key_parts[1]) <= endpos:\n",
        "                    overlap_found = True\n",
        "                    break\n",
        "        if overlap_found:\n",
        "            sitehash[site_key] = sitehash.get(site_key, 0) + 1\n",
        "        else:\n",
        "            discardhash[site_key] = discardhash.get(site_key, 0) + 1\n",
        "\n",
        "    # -----------------------------\n",
        "    # 5. output final results\n",
        "    # -----------------------------\n",
        "    with open(output_file, \"w\") as out_f:\n",
        "        for idx, row in site_df.iterrows():\n",
        "            chrom = row[0]\n",
        "            pos = str(row[1])\n",
        "            fields = row[3].split(\"@\")\n",
        "            cov = int(fields[2])\n",
        "            oldalter = int(fields[3])\n",
        "            site_key = f\"{chrom}_{pos}\"\n",
        "            newalter = sitehash.get(site_key, 0)\n",
        "            discardnum = discardhash.get(site_key, 0)\n",
        "            newcov = cov - (oldalter - newalter)\n",
        "            neweditfreq = round(newalter / newcov, 3) if newcov > 0 else 0\n",
        "            if newalter >= min_mismatch and newalter > discardnum:\n",
        "                out_f.write(f\"{chrom}\\t{int(pos)-1}\\t{pos}\\t{fields[0]}@{fields[1]}@{newcov}@{newalter}@{neweditfreq}\\n\")\n",
        "\n",
        "    # os.remove(fatmp_file)\n",
        "    # os.remove(psl_file)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "w2QMQAh4teRj",
        "outputId": "8e5cb657-4b55-4b1b-fcb8-7310d639587f"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Loaded 3099750718 letters in 194 sequences\n",
            "Searched 47802 bases in 337 sequences\n",
            "Loaded 3099750718 letters in 194 sequences\n",
            "Searched 6570 bases in 48 sequences\n",
            "\n",
            "ADAR_WT non-Alu variant sites before Blat: 24, after Blat: 4\n",
            "ADAR_KO non-Alu variant sites before Blat: 15, after Blat: 0\n",
            "\n",
            "ADAR_WT non-Alu variant sites after Blat:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>0</th>\n",
              "      <th>1</th>\n",
              "      <th>2</th>\n",
              "      <th>3</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100517</td>\n",
              "      <td>37100518</td>\n",
              "      <td>T@C@115@69@0.6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100557</td>\n",
              "      <td>37100558</td>\n",
              "      <td>T@C@140@75@0.536</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100952</td>\n",
              "      <td>37100953</td>\n",
              "      <td>T@C@138@74@0.536</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr20</td>\n",
              "      <td>37519159</td>\n",
              "      <td>37519160</td>\n",
              "      <td>T@C@146@77@0.527</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "       0         1         2                 3\n",
              "0   chr2  37100517  37100518    T@C@115@69@0.6\n",
              "1   chr2  37100557  37100558  T@C@140@75@0.536\n",
              "2   chr2  37100952  37100953  T@C@138@74@0.536\n",
              "3  chr20  37519159  37519160  T@C@146@77@0.527"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_KO non-Alu variant sites after Blat:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "Empty DataFrame\n",
              "Columns: []\n",
              "Index: []"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# invoke 'blat_filter_nonAlu_sites' function on ADAR_WT and ADAR_KO non-Alu sites\n",
        "blat_filter_nonAlu_sites(\n",
        "    site_file='./data/4_editing/ADAR_WT_nonAlu_sites.txt',\n",
        "    bam_file='./data/2_dedup_bam/ADAR_WT_dePCRdup.bam',\n",
        "    ref_fa=REF_fa,\n",
        "    output_file='./data/4_editing/ADAR_WT_nonAlu_blat_filtered.txt',\n",
        "    min_basequal=25,\n",
        "    min_mismatch=2,\n",
        "    score_limit=0.95,\n",
        "    blat_bin=\"/public/home/lix/bin/x86_64/blat\"\n",
        ")\n",
        "\n",
        "blat_filter_nonAlu_sites(\n",
        "    site_file='./data/4_editing/ADAR_KO_nonAlu_sites.txt',\n",
        "    bam_file='./data/2_dedup_bam/ADAR_KO_dePCRdup.bam',\n",
        "    ref_fa=REF_fa,\n",
        "    output_file='./data/4_editing/ADAR_KO_nonAlu_blat_filtered.txt',\n",
        "    min_basequal=25,\n",
        "    min_mismatch=2,\n",
        "    score_limit=0.95,\n",
        "    blat_bin=\"/public/home/lix/bin/x86_64/blat\"\n",
        ")\n",
        "\n",
        "# number: before and after BLAT filtering\n",
        "try:\n",
        "    ADAR_WT_nonAlu_blat_filtered_df = pd.read_csv(\n",
        "        './data/4_editing/ADAR_WT_nonAlu_blat_filtered.txt',\n",
        "        sep='\\t', header=None\n",
        "    )\n",
        "except pd.errors.EmptyDataError:\n",
        "    ADAR_WT_nonAlu_blat_filtered_df = pd.DataFrame()\n",
        "\n",
        "try:\n",
        "    ADAR_KO_nonAlu_blat_filtered_df = pd.read_csv(\n",
        "        './data/4_editing/ADAR_KO_nonAlu_blat_filtered.txt',\n",
        "        sep='\\t', header=None\n",
        "    )\n",
        "except pd.errors.EmptyDataError:\n",
        "    ADAR_KO_nonAlu_blat_filtered_df = pd.DataFrame()\n",
        "\n",
        "print()\n",
        "\n",
        "print(\"ADAR_WT non-Alu variant sites before Blat: %s, after Blat: %s\" % (len(ADAR_WT_nonAlu_rm_homo), len(ADAR_WT_nonAlu_blat_filtered_df)))\n",
        "print(\"ADAR_KO non-Alu variant sites before Blat: %s, after Blat: %s\" % (len(ADAR_KO_nonAlu_rm_homo), len(ADAR_KO_nonAlu_blat_filtered_df)))\n",
        "\n",
        "print()\n",
        "\n",
        "print(\"ADAR_WT non-Alu variant sites after Blat:\")\n",
        "display(ADAR_WT_nonAlu_blat_filtered_df.head())\n",
        "print(\"ADAR_KO non-Alu variant sites after Blat:\")\n",
        "display(ADAR_KO_nonAlu_blat_filtered_df.head())"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "aAm9jXJ5teRj"
      },
      "source": [
        "##### 6） Separate editing sites in repetitive and nonrepetitive regions"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4HzDyykEteRj"
      },
      "outputs": [],
      "source": [
        "def Separate_rep_nonrep_sites(variant_file, nonAlu_repeat_bed, output_rep_file, output_nonrep_file):\n",
        "    # load variant bed\n",
        "    variant_df = pd.read_csv(variant_file, sep='\\t', header=None, names=['chrom','start','end','name'])\n",
        "    variant_bed = pybedtools.BedTool.from_dataframe(variant_df)\n",
        "\n",
        "    # bedtools intersect to get repeat sites\n",
        "    variant_sites_rep_bed = variant_bed.intersect(nonAlu_repeat_bed, u=True) # u=True: report only those that have overlaps\n",
        "    variant_sites_rep_df = variant_sites_rep_bed.to_dataframe(names=['chrom','start','end','name'])\n",
        "\n",
        "    # bedtools intersect to get non-repeat sites\n",
        "    variant_sites_nonrep_bed = variant_bed.intersect(nonAlu_repeat_bed, v=True) # v=True: exclude overlapping\n",
        "    variant_sites_nonrep_df = variant_sites_nonrep_bed.to_dataframe(names=['chrom','start','end','name'])\n",
        "\n",
        "    # save\n",
        "    variant_sites_rep_df.to_csv(output_rep_file, sep='\\t', header=False, index=False)\n",
        "    variant_sites_nonrep_df.to_csv(output_nonrep_file, sep='\\t', header=False, index=False)\n",
        "\n",
        "    return variant_sites_rep_df, variant_sites_nonrep_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "RJptxMoAteRk",
        "outputId": "83bcb339-e708-4e62-87df-48c7908277dc"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ADAR_WT non-Alu variant sites in repeat regions: 0, in non-repeat regions: 4\n",
            "ADAR_KO non-Alu variant sites in repeat regions: 0, in non-repeat regions: 0\n"
          ]
        }
      ],
      "source": [
        "# invoke 'Separate_rep_nonrep_sites' function on ADAR_WT and ADAR_KO non-Alu blat filtered variant bed data\n",
        "\n",
        "nonAlu_repeat_bed_path = pybedtools.BedTool('./data/reference/hg38_nonAlu_repeat.bed')\n",
        "ADAR_WT_nonAlu_rep, ADAR_WT_nonAlu_nonrep = Separate_rep_nonrep_sites(\n",
        "    variant_file='./data/4_editing/ADAR_WT_nonAlu_blat_filtered.txt',\n",
        "    nonAlu_repeat_bed=nonAlu_repeat_bed_path,\n",
        "    output_rep_file='./data/4_editing/ADAR_WT_nonAlu_repeat_sites.txt',\n",
        "    output_nonrep_file='./data/4_editing/ADAR_WT_nonAlu_nonrepeat_sites.txt'\n",
        ")\n",
        "\n",
        "ADAR_KO_nonAlu_rep, ADAR_KO_nonAlu_nonrep = Separate_rep_nonrep_sites(\n",
        "    variant_file='./data/4_editing/ADAR_KO_nonAlu_blat_filtered.txt',\n",
        "    nonAlu_repeat_bed=nonAlu_repeat_bed_path,\n",
        "    output_rep_file='./data/4_editing/ADAR_KO_nonAlu_repeat_sites.txt',\n",
        "    output_nonrep_file='./data/4_editing/ADAR_KO_nonAlu_nonrepeat_sites.txt'\n",
        ")\n",
        "\n",
        "# number: repeat and non-repeat sites\n",
        "print(\"ADAR_WT non-Alu variant sites in repeat regions: %s, in non-repeat regions: %s\" % (len(ADAR_WT_nonAlu_rep), len(ADAR_WT_nonAlu_nonrep)))\n",
        "print(\"ADAR_KO non-Alu variant sites in repeat regions: %s, in non-repeat regions: %s\" % (len(ADAR_KO_nonAlu_rep), len(ADAR_KO_nonAlu_nonrep)))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DnFbJNcjteRk"
      },
      "source": [
        "### Step 4 - Results Analysis\n",
        "#### Files -\n",
        "        1. Alu.sites\n",
        "        2. nonAlu_Rep.sites\n",
        "        3. nonAlu_nonRep.sites"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ka1MGWl9teRk"
      },
      "source": [
        "##### 1 - show the numbers of editing sites"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dbYYBMg_teRk",
        "outputId": "ed3f99a6-55d2-4537-9f98-d903936e0fbb"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Final ADAR_WT sites: Alu: 241, nonAlu repeat: 0, nonAlu non-repeat: 4\n",
            "Final ADAR_KO sites: Alu: 61, nonAlu repeat: 0, nonAlu non-repeat: 0\n"
          ]
        }
      ],
      "source": [
        "# ADAR_KO all final results\n",
        "ADAR_KO_Alu = pd.read_csv(os.path.join(EDITING_DIR, 'ADAR_KO_Alu_sites.txt'), sep='\\t', header=None, names=['chrom','start','end','name'])\n",
        "ADAR_KO_nonAlu_repeat = pd.read_csv(os.path.join(EDITING_DIR, 'ADAR_KO_nonAlu_repeat_sites.txt'), sep='\\t', header=None, names=['chrom','start','end','name'])\n",
        "ADAR_KO_nonAlu_nonrepeat = pd.read_csv(os.path.join(EDITING_DIR, 'ADAR_KO_nonAlu_nonrepeat_sites.txt'), sep='\\t', header=None, names=['chrom','start','end','name'])\n",
        "\n",
        "# ADAR_WT all final results\n",
        "ADAR_WT_Alu = pd.read_csv(os.path.join(EDITING_DIR, 'ADAR_WT_Alu_sites.txt'), sep='\\t', header=None, names=['chrom','start','end','name'])\n",
        "ADAR_WT_nonAlu_repeat = pd.read_csv(os.path.join(EDITING_DIR, 'ADAR_WT_nonAlu_repeat_sites.txt'), sep='\\t', header=None, names=['chrom','start','end','name'])\n",
        "ADAR_WT_nonAlu_nonrepeat = pd.read_csv(os.path.join(EDITING_DIR, 'ADAR_WT_nonAlu_nonrepeat_sites.txt'), sep='\\t', header=None, names=['chrom','start','end','name'])\n",
        "\n",
        "print(f\"Final ADAR_WT sites: Alu: {len(ADAR_WT_Alu)}, nonAlu repeat: {len(ADAR_WT_nonAlu_repeat)}, nonAlu non-repeat: {len(ADAR_WT_nonAlu_nonrepeat)}\")\n",
        "print(f\"Final ADAR_KO sites: Alu: {len(ADAR_KO_Alu)}, nonAlu repeat: {len(ADAR_KO_nonAlu_repeat)}, nonAlu non-repeat: {len(ADAR_KO_nonAlu_nonrepeat)}\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xuA0V9YCteRk"
      },
      "source": [
        "##### 2 - merge Alu, non-Alu repeat, non-Alu non-repeat sites"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "cAtb0bJuteRk",
        "outputId": "441d4279-4821-4f7f-c7be-ed18184af268"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Final ADAR_WT editing sites:\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/tmp/ipykernel_1750080/488196005.py:31: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
            "  ADAR_KO_all = pd.concat([\n",
            "/tmp/ipykernel_1750080/488196005.py:36: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
            "  ADAR_WT_all = pd.concat([\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>region_type</th>\n",
              "      <th>ref</th>\n",
              "      <th>alt</th>\n",
              "      <th>cov</th>\n",
              "      <th>mis</th>\n",
              "      <th>ediLevel</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144440666</td>\n",
              "      <td>144440667</td>\n",
              "      <td>Alu</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>1.000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144440683</td>\n",
              "      <td>144440684</td>\n",
              "      <td>Alu</td>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>7</td>\n",
              "      <td>2</td>\n",
              "      <td>0.286</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144446011</td>\n",
              "      <td>144446012</td>\n",
              "      <td>Alu</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>8</td>\n",
              "      <td>8</td>\n",
              "      <td>1.000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37102068</td>\n",
              "      <td>37102069</td>\n",
              "      <td>Alu</td>\n",
              "      <td>A</td>\n",
              "      <td>T</td>\n",
              "      <td>53</td>\n",
              "      <td>2</td>\n",
              "      <td>0.038</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37102070</td>\n",
              "      <td>37102071</td>\n",
              "      <td>Alu</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>51</td>\n",
              "      <td>11</td>\n",
              "      <td>0.216</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>240</th>\n",
              "      <td>chr20</td>\n",
              "      <td>37505014</td>\n",
              "      <td>37505015</td>\n",
              "      <td>Alu</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>4</td>\n",
              "      <td>2</td>\n",
              "      <td>0.500</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>241</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100517</td>\n",
              "      <td>37100518</td>\n",
              "      <td>nonAlu_nonrepeat</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>115</td>\n",
              "      <td>69</td>\n",
              "      <td>0.600</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>242</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100557</td>\n",
              "      <td>37100558</td>\n",
              "      <td>nonAlu_nonrepeat</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>140</td>\n",
              "      <td>75</td>\n",
              "      <td>0.536</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>243</th>\n",
              "      <td>chr2</td>\n",
              "      <td>37100952</td>\n",
              "      <td>37100953</td>\n",
              "      <td>nonAlu_nonrepeat</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>138</td>\n",
              "      <td>74</td>\n",
              "      <td>0.536</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>244</th>\n",
              "      <td>chr20</td>\n",
              "      <td>37519159</td>\n",
              "      <td>37519160</td>\n",
              "      <td>nonAlu_nonrepeat</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>146</td>\n",
              "      <td>77</td>\n",
              "      <td>0.527</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>245 rows × 9 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     chrom      start        end       region_type ref alt  cov mis  ediLevel\n",
              "0     chr1  144440666  144440667               Alu   G   T    2   2     1.000\n",
              "1     chr1  144440683  144440684               Alu   A   C    7   2     0.286\n",
              "2     chr1  144446011  144446012               Alu   T   C    8   8     1.000\n",
              "3     chr2   37102068   37102069               Alu   A   T   53   2     0.038\n",
              "4     chr2   37102070   37102071               Alu   T   C   51  11     0.216\n",
              "..     ...        ...        ...               ...  ..  ..  ...  ..       ...\n",
              "240  chr20   37505014   37505015               Alu   T   C    4   2     0.500\n",
              "241   chr2   37100517   37100518  nonAlu_nonrepeat   T   C  115  69     0.600\n",
              "242   chr2   37100557   37100558  nonAlu_nonrepeat   T   C  140  75     0.536\n",
              "243   chr2   37100952   37100953  nonAlu_nonrepeat   T   C  138  74     0.536\n",
              "244  chr20   37519159   37519160  nonAlu_nonrepeat   T   C  146  77     0.527\n",
              "\n",
              "[245 rows x 9 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Final ADAR_KO editing sites:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>region_type</th>\n",
              "      <th>ref</th>\n",
              "      <th>alt</th>\n",
              "      <th>cov</th>\n",
              "      <th>mis</th>\n",
              "      <th>ediLevel</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>144446011</td>\n",
              "      <td>144446012</td>\n",
              "      <td>Alu</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>4</td>\n",
              "      <td>4</td>\n",
              "      <td>1.000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>160269601</td>\n",
              "      <td>160269602</td>\n",
              "      <td>Alu</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>1.000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>160289057</td>\n",
              "      <td>160289058</td>\n",
              "      <td>Alu</td>\n",
              "      <td>C</td>\n",
              "      <td>A</td>\n",
              "      <td>27</td>\n",
              "      <td>2</td>\n",
              "      <td>0.074</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>160303121</td>\n",
              "      <td>160303122</td>\n",
              "      <td>Alu</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>4</td>\n",
              "      <td>2</td>\n",
              "      <td>0.500</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>160311472</td>\n",
              "      <td>160311473</td>\n",
              "      <td>Alu</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>1.000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>56</th>\n",
              "      <td>chr19</td>\n",
              "      <td>53578963</td>\n",
              "      <td>53578964</td>\n",
              "      <td>Alu</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>94</td>\n",
              "      <td>2</td>\n",
              "      <td>0.021</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>57</th>\n",
              "      <td>chr19</td>\n",
              "      <td>53579032</td>\n",
              "      <td>53579033</td>\n",
              "      <td>Alu</td>\n",
              "      <td>A</td>\n",
              "      <td>T</td>\n",
              "      <td>66</td>\n",
              "      <td>45</td>\n",
              "      <td>0.682</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>58</th>\n",
              "      <td>chr19</td>\n",
              "      <td>53579047</td>\n",
              "      <td>53579048</td>\n",
              "      <td>Alu</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>63</td>\n",
              "      <td>2</td>\n",
              "      <td>0.032</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>59</th>\n",
              "      <td>chr19</td>\n",
              "      <td>53579091</td>\n",
              "      <td>53579092</td>\n",
              "      <td>Alu</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>47</td>\n",
              "      <td>2</td>\n",
              "      <td>0.043</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>60</th>\n",
              "      <td>chr20</td>\n",
              "      <td>37504962</td>\n",
              "      <td>37504963</td>\n",
              "      <td>Alu</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>6</td>\n",
              "      <td>6</td>\n",
              "      <td>1.000</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>61 rows × 9 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "    chrom      start        end region_type ref alt cov mis  ediLevel\n",
              "0    chr1  144446011  144446012         Alu   T   C   4   4     1.000\n",
              "1    chr1  160269601  160269602         Alu   G   T   2   2     1.000\n",
              "2    chr1  160289057  160289058         Alu   C   A  27   2     0.074\n",
              "3    chr1  160303121  160303122         Alu   C   T   4   2     0.500\n",
              "4    chr1  160311472  160311473         Alu   G   A   2   2     1.000\n",
              "..    ...        ...        ...         ...  ..  ..  ..  ..       ...\n",
              "56  chr19   53578963   53578964         Alu   C   T  94   2     0.021\n",
              "57  chr19   53579032   53579033         Alu   A   T  66  45     0.682\n",
              "58  chr19   53579047   53579048         Alu   C   T  63   2     0.032\n",
              "59  chr19   53579091   53579092         Alu   C   T  47   2     0.043\n",
              "60  chr20   37504962   37504963         Alu   T   C   6   6     1.000\n",
              "\n",
              "[61 rows x 9 columns]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# merge Alu, non-Alu repeat, non-Alu non-repeat sites into one table for ADAR_KO and ADAR_WT respectively\n",
        "def load_and_parse_sites(df, region_type):\n",
        "    df = df.copy()\n",
        "\n",
        "    if df.empty:\n",
        "        return pd.DataFrame(columns=[\n",
        "            'chrom', 'start', 'end', 'region_type',\n",
        "            'ref', 'alt', 'cov', 'mis', 'ediLevel'\n",
        "        ])\n",
        "\n",
        "    # region type\n",
        "    df['region_type'] = region_type\n",
        "\n",
        "    # parse name 列：ref@alt@cov@mis@ediLevel\n",
        "    parsed = df['name'].astype(str).str.split('@', expand=True)\n",
        "\n",
        "    if parsed.shape[1] != 5:\n",
        "        raise ValueError(f\"Unexpected name format in {region_type}\")\n",
        "\n",
        "    parsed.columns = ['ref', 'alt', 'cov', 'mis', 'ediLevel']\n",
        "\n",
        "    parsed['cov'] = parsed['cov'].astype(int)\n",
        "    parsed['mis'] = parsed['mis'].astype(int)\n",
        "    parsed['ediLevel'] = parsed['ediLevel'].astype(float)\n",
        "\n",
        "    # merge\n",
        "    df = pd.concat([df.drop(columns='name'), parsed], axis=1)\n",
        "\n",
        "    return df\n",
        "\n",
        "ADAR_KO_all = pd.concat([\n",
        "    load_and_parse_sites(ADAR_KO_Alu, 'Alu'),\n",
        "    load_and_parse_sites(ADAR_KO_nonAlu_repeat, 'nonAlu_repeat'),\n",
        "    load_and_parse_sites(ADAR_KO_nonAlu_nonrepeat, 'nonAlu_nonrepeat')\n",
        "], ignore_index=True)\n",
        "ADAR_WT_all = pd.concat([\n",
        "    load_and_parse_sites(ADAR_WT_Alu, 'Alu'),\n",
        "    load_and_parse_sites(ADAR_WT_nonAlu_repeat, 'nonAlu_repeat'),\n",
        "    load_and_parse_sites(ADAR_WT_nonAlu_nonrepeat, 'nonAlu_nonrepeat')\n",
        "], ignore_index=True)\n",
        "\n",
        "print(\"Final ADAR_WT editing sites:\")\n",
        "display(ADAR_WT_all)\n",
        "\n",
        "print(\"Final ADAR_KO editing sites:\")\n",
        "display(ADAR_KO_all)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "rjgB5hATteRk"
      },
      "source": [
        "##### 3 - compare with the editing sites in REDIrepotal database"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ydNuXcCDteRl",
        "outputId": "200059af-3a43-498e-b90e-e6a9b8cfec2a"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Number of total ADAR_KO A-to-G editing sites: 27\n",
            "Number of ADAR_KO A-to-G editing sites overlapping with REDIportal known sites: 20\n",
            "\n",
            "Number of total ADAR_WT A-to-G editing sites: 215\n",
            "Number of ADAR_WT A-to-G editing sites overlapping with REDIportal known sites: 204\n"
          ]
        }
      ],
      "source": [
        "ADAR_KO_AG = ADAR_KO_all[\n",
        "    (ADAR_KO_all['ref'] == 'A') & (ADAR_KO_all['alt'] == 'G') |\n",
        "    (ADAR_KO_all['ref'] == 'T') & (ADAR_KO_all['alt'] == 'C')\n",
        "]\n",
        "\n",
        "ADAR_WT_AG = ADAR_WT_all[\n",
        "    (ADAR_WT_all['ref'] == 'A') & (ADAR_WT_all['alt'] == 'G') |\n",
        "    (ADAR_WT_all['ref'] == 'T') & (ADAR_WT_all['alt'] == 'C')\n",
        "]\n",
        "\n",
        "REDIrepotal_data = pd.read_csv('./data/reference/REDIportal_known_AG_sites_selected_genes.bed', sep='\\t', header=None, usecols=[0,1,2], names=['chrom', 'start', 'end'])\n",
        "\n",
        "# calculate overlap with REDIportal known A-to-G editing sites\n",
        "def calculate_REDIportal_overlap(variant_df, REDIportal_df):\n",
        "    merged_df = variant_df.merge(\n",
        "        REDIportal_df,\n",
        "        left_on=['chrom', 'start', 'end'],\n",
        "        right_on=['chrom', 'start', 'end'],\n",
        "        how='inner'\n",
        "    )\n",
        "    return merged_df\n",
        "\n",
        "ADAR_KO_AG_REDIportal_overlap = calculate_REDIportal_overlap(ADAR_KO_AG, REDIrepotal_data)\n",
        "ADAR_WT_AG_REDIportal_overlap = calculate_REDIportal_overlap(ADAR_WT_AG, REDIrepotal_data)\n",
        "\n",
        "# number of overlapping sites\n",
        "print(\"Number of total ADAR_KO A-to-G editing sites: %s\" % len(ADAR_KO_AG))\n",
        "print(\"Number of ADAR_KO A-to-G editing sites overlapping with REDIportal known sites: %s\" % len(ADAR_KO_AG_REDIportal_overlap))\n",
        "print()\n",
        "print(\"Number of total ADAR_WT A-to-G editing sites: %s\" % len(ADAR_WT_AG))\n",
        "print(\"Number of ADAR_WT A-to-G editing sites overlapping with REDIportal known sites: %s\" % len(ADAR_WT_AG_REDIportal_overlap))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DzFAR1EsteRl"
      },
      "source": [
        "##### 4 - calculate the A-G fraction and A-G count in different regions(Alu、 nonAlu、 nonRep)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dDu9IsObteRl",
        "outputId": "228bc98c-9f3a-4c8d-b487-47e85ef1d549"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "A-to-G / T-to-C editing fraction in different regions:\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/tmp/ipykernel_1750080/284331363.py:20: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
            "  .apply(lambda x: x['is_AG_TC'].sum() / len(x) if len(x) > 0 else 0)\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>region_type</th>\n",
              "      <th>condition</th>\n",
              "      <th>AG_fraction</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Alu</td>\n",
              "      <td>KO</td>\n",
              "      <td>0.442623</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Alu</td>\n",
              "      <td>WT</td>\n",
              "      <td>0.875519</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>nonAlu_nonrepeat</td>\n",
              "      <td>WT</td>\n",
              "      <td>1.000000</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "        region_type condition  AG_fraction\n",
              "0               Alu        KO     0.442623\n",
              "1               Alu        WT     0.875519\n",
              "2  nonAlu_nonrepeat        WT     1.000000"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "A-to-G / T-to-C editing counts in different regions:\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>region_type</th>\n",
              "      <th>condition</th>\n",
              "      <th>AG_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Alu</td>\n",
              "      <td>KO</td>\n",
              "      <td>27</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Alu</td>\n",
              "      <td>WT</td>\n",
              "      <td>211</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>nonAlu_nonrepeat</td>\n",
              "      <td>WT</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "        region_type condition  AG_count\n",
              "0               Alu        KO        27\n",
              "1               Alu        WT       211\n",
              "2  nonAlu_nonrepeat        WT         4"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# calculate A-to-G / T-to-C editing fraction and counts in different regions for ADAR_KO and ADAR_WT\n",
        "\n",
        "ADAR_KO_all['condition'] = 'KO'\n",
        "ADAR_WT_all['condition'] = 'WT'\n",
        "\n",
        "all_sites = pd.concat([ADAR_KO_all, ADAR_WT_all], ignore_index=True)\n",
        "\n",
        "regions = ['Alu', 'nonAlu_repeat', 'nonAlu_nonrepeat']\n",
        "conditions = ['WT', 'KO']\n",
        "\n",
        "is_AG_TC = (\n",
        "    ((all_sites['ref'] == 'A') & (all_sites['alt'] == 'G')) |\n",
        "    ((all_sites['ref'] == 'T') & (all_sites['alt'] == 'C'))\n",
        ")\n",
        "\n",
        "AG_fraction_df = (\n",
        "    all_sites\n",
        "    .assign(is_AG_TC=is_AG_TC)\n",
        "    .groupby(['region_type', 'condition'])\n",
        "    .apply(lambda x: x['is_AG_TC'].sum() / len(x) if len(x) > 0 else 0)\n",
        "    .reset_index(name='AG_fraction')\n",
        ")\n",
        "\n",
        "AG_count_df = (\n",
        "    all_sites[is_AG_TC]\n",
        "    .groupby(['region_type', 'condition'])\n",
        "    .size()\n",
        "    .reset_index(name='AG_count')\n",
        ")\n",
        "\n",
        "print(\"A-to-G / T-to-C editing fraction in different regions:\")\n",
        "display(AG_fraction_df)\n",
        "\n",
        "print(\"A-to-G / T-to-C editing counts in different regions:\")\n",
        "display(AG_count_df)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sIBca4_AteRl"
      },
      "source": [
        "### Step 5 -Plot\n",
        "#### 1 - A-G fraction\n",
        "#### 2 - A-G numbers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "k8TcXXt7teRl"
      },
      "outputs": [],
      "source": [
        "# prepare data\n",
        "\n",
        "AG_fraction_plot = (\n",
        "    AG_fraction_df\n",
        "    .set_index(['region_type', 'condition'])\n",
        "    .reindex(\n",
        "        pd.MultiIndex.from_product(\n",
        "            [regions, conditions],\n",
        "            names=['region_type', 'condition']\n",
        "        ),\n",
        "        fill_value=0\n",
        "    )\n",
        "    .reset_index()\n",
        ")\n",
        "\n",
        "AG_count_plot = (\n",
        "    AG_count_df\n",
        "    .set_index(['region_type', 'condition'])\n",
        "    .reindex(\n",
        "        pd.MultiIndex.from_product(\n",
        "            [regions, conditions],\n",
        "            names=['region_type', 'condition']\n",
        "        ),\n",
        "        fill_value=0\n",
        "    )\n",
        "    .reset_index()\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "mtH-UbhhteRl",
        "outputId": "178b4b88-1378-4143-9b20-03340d3d19b3"
      },
      "outputs": [
        {
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAk4AAAGGCAYAAACNCg6xAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAY/JJREFUeJzt3XdYFNf7NvB76R0VARsqdlGsWMAoYEGw915Q0FgSo2gsQWyoqFEkRjEx9q8Ne29EAXvvgr2ABkVRqYLCnvcPX+bnCuquArvK/bmuvXTPnpl5ZpjdffacM2dkQggBIiIiIvosLXUHQERERPStYOJEREREpCQmTkRERERKYuJEREREpCQmTkRERERKYuJEREREpCQmTkRERERKYuJEREREpCQmTkRERERKYuKkBgsWLIBMJkP16tVVWi4yMhJTpkzBgwcP8iaw/+/YsWPo2bMnSpcuDX19fRgbG6NatWoYPXo0bty4ofR6MjMzYWVlhfnz52d77cGDB5DJZEo9Pra/69atQ1BQ0BfupXKOHj2Kbt26oWTJktDT04O5uTmcnJywePFipKSkfHJZT0/PT+7X54SHh0MmkyE8PFxhnWXLllWoN3PmTGzfvl2p5fNL1rZzenTp0iVfYwkODsbKlSuzlWedgzm9ltc8PT1hYmKS59txcXFROPYGBgaws7PD9OnT8ebNmzzdtjqPb34qW7YsPD09laqbmJiIGTNmwMHBAWZmZtDX10fZsmUxcOBAXLhwIW8D/YgP48/pc2Pv3r2YMmWKUssXBDrqDqAgWr58OQDg+vXrOH36NBo0aKDUcpGRkZg6dSpcXFyyfXnmlokTJ2LGjBlwdHTExIkTUbFiRWRkZODKlStYtWoVAgMDkZGRAW1t7c+u68iRI3j27Bk6deqU7bXixYvj5MmTCmXDhg1DQkIC1q5dm61uTtatW4dr165h5MiRyu+gCiZPnoxp06bByckJ/v7+KF++PFJTU3HixAlMmTIFt27dyjEpfJ+hoSEOHz6cazH5+fnhl19+USibOXMmunTpgg4dOiiU16lTBydPnoSdnV2ubV9VM2fOhKurq0KZhYVFvsYQHByMokWLZvtwzzoHy5cvn6/x5Ldy5cpJ76lnz55h6dKl8PPzQ3R0NJYsWZJn2y0ox1dZd+/ehZubG+Li4jBkyBBMnToVJiYmePDgATZu3Ii6devi1atXMDc3V2ucOX1u7N27F4sWLcoxedq2bRvMzMzyMUL1Y+KUz86dO4fLly+jdevW2LNnD5YtW6Z04pTX1q9fjxkzZmDIkCEIDg5WaBVp0aIFfHx8EBwcrPT6Nm/eDAcHB5QpUybba/r6+mjYsKFCmZmZGd68eZOtXB02bdqEadOmwcvLC//884/CsfDw8MDYsWOzJX450dLSytX9UeVLyMzMTO3HsmLFikrHkJmZiYyMDOjr6+dxVO/kdA5+jwwNDRX208PDA3Z2dli1ahUWLFgAAwODPNluQTm+ysjMzETHjh3x/PlznDx5UqG3wdnZGf3798e+ffugq6urxijfUfVzo3bt2nkYjWZiV10+W7ZsGQBg1qxZcHJywoYNG5CamvrZ5VauXImuXbsCAFxdXaWm9/ebwZcvX46aNWvCwMAARYoUQceOHREVFaV0bNOnT0fRokUxf/78HLuSZDIZhg8frlRrkxAC27ZtQ+fOnZXevipcXFywZ88ePHz4MMfurxcvXmDYsGFSF1u5cuXg6+uL9PR0pdY/bdo0FC5cWOpW/ZCpqSnc3NxybX9u3LgBd3d3GBkZoWjRohgyZAiSkpKy1fuwq04mkyElJQWrVq2SjoGLiwuAj3f1mZiY4M6dO2jVqhVMTExgY2OD0aNHZzs2jx49QpcuXWBqaopChQqhd+/eOHv2bK50v2R148yZMwfTp0+Hra0t9PX1ERYWhrS0NIwePRq1atWCubk5ihQpAkdHR+zYsSPbeuRyOf7880/UqlULhoaGKFSoEBo2bIidO3cCeNeNcP36dUREREjHJ+v4fawr6dixY2jWrBlMTU1hZGQEJycn7NmzR6HOypUrIZPJEBYWhqFDh6Jo0aKwsLBAp06d8N9//yl9HK5fv45mzZrB2NgYlpaW+OmnnxQ+D5o1a4YqVargw3uxCyFQoUIFtG7dWultZdHR0UGtWrXw5s0bvHr1SmGdwcHB0rEsXLgwunTpgnv37mXb9syZM1GmTBkYGBjAwcEBoaGhcHFxkc49QL3H99y5c+jRowfKli0LQ0NDlC1bFj179sTDhw+/eDtv377F2LFjUaxYMRgZGeGHH37AmTNnPhsLAGzfvh1Xr17FhAkTPjpEw8PDA0ZGRnl2nJSN/8PPDU9PTyxatAgAchxCkVNXXXR0NPr06QMrKyvo6+ujatWqmDdvHuRyuVQn6/yYO3cuAgMDYWtrCxMTEzg6OuLUqVNKHVd1YeKUj16/fo3169ejXr16qF69OgYOHIikpCRs2rTps8u2bt0aM2fOBAAsWrQIJ0+exMmTJ6UPzoCAAHh5eaFatWrYunUr/vjjD1y5cgWOjo64ffv2Z9f/33//ITIyEi1atMiVX6AnTpxAbGxsniVOwcHBaNSoEYoVKyYdi6wWoLS0NLi6umL16tXw8fHBnj170KdPH8yZMyfHbsMPxcbG4tq1a3Bzc1P4IPtSGRkZ2R7vf4A8ffoUzs7OuHbtGoKDg/G///0PycnJ+Omnnz677pMnT8LQ0BCtWrWSjsHnWgXfvn2Ldu3aoVmzZtixYwcGDhyI+fPnY/bs2VKdlJQUuLq6IiwsDLNnz8bGjRthbW2N7t27q7Tvcrk8276/b8GCBTh8+DDmzp2Lffv2oUqVKkhPT8eLFy8wZswYbN++HevXr8cPP/yATp06YfXq1QrLe3p64pdffkG9evUQEhKCDRs2oF27dtKH+rZt21CuXDnUrl1bOj7btm37aLwRERFo2rQpEhISsGzZMqxfvx6mpqZo27YtQkJCstX39vaGrq4u1q1bhzlz5iA8PBx9+vRR6ti8ffsWrVq1QrNmzbB9+3b89NNP+PvvvxWO8S+//IKbN2/i0KFDCsvu27cPd+/exfDhw5Xa1ofu37+PQoUKwdLSUir78ccfMXLkSDRv3hzbt29HcHAwrl+/DicnJzx9+lSq5+vrC19fX7i7u2PHjh0YMmQIvL29cevWrc9uN7+O74MHD1C5cmUEBQXhwIEDmD17NmJjY1GvXj08f/78i7YzaNAgzJ07F/369cOOHTvQuXNndOrUCS9fvvxsPAcPHgSAbN3pH5MXx+lL4/fz85PGJb7/WfuxIRTPnj2Dk5MTDh48CH9/f+zcuRPNmzfHmDFjcvxMW7RoEUJDQxEUFIS1a9ciJSUFrVq1QkJCglLHSi0E5ZvVq1cLAOKvv/4SQgiRlJQkTExMROPGjZVaftOmTQKACAsLUyh/+fKlMDQ0FK1atVIoj46OFvr6+qJXr16fXfepU6cEADF+/Phsr2VkZIi3b99KD7lc/tn1jRw5Utjb23+23vucnZ1FtWrVlK7funVrUaZMmWzlf/31lwAgNm7cqFA+e/ZsAUAcPHjwk+v91LFQRf/+/QWAHB/NmjWT6o0bN07IZDJx6dIlheVbtGiR7e/dv3//bPtsbGws+vfvn237YWFhOS6f07Fp1aqVqFy5svR80aJFAoDYt2+fQr0ff/xRABArVqz45L5nbTunx+3bt8X9+/cFAFG+fHnx5s2bT64r6/zz8vIStWvXlsqPHDkiAAhfX99PLl+tWjXh7OycrTwrhvf3pWHDhsLKykokJSUpbL969eqiVKlS0rm/YsUKAUAMGzZMYZ1z5swRAERsbOwnY8r6O/zxxx8K5TNmzBAAxLFjx4QQQmRmZopy5cqJ9u3bK9Tz8PAQ5cuX/+x7Mes9lfXejY2NFZMmTVL4HBJCiJMnTwoAYt68eQrLx8TECENDQzF27FghhBAvXrwQ+vr6onv37gr1spZ//zir8/h+KCMjQyQnJwtjY2OFY67sdqKiogQAMWrUKIV6a9euFQByfP+9z93dXQAQaWlpSsWb28dJlfhz+twYPny4+Fi6UKZMGYXlx48fLwCI06dPK9QbOnSokMlk4ubNm0KI/zs/7O3tRUZGhlTvzJkzAoBYv379Z46S+rDFKR8tW7YMhoaG6NGjBwDAxMQEXbt2xdGjRxVahT78hS4+aKb/0MmTJ/H69etszaU2NjZo2rSpwq9VVdcNvBvMq6urKz22bNny2WW2bt2aK61NH7ZYZGZmfnaZw4cPw9jYONvVW1nHJ+t4CCE+2RqSGwwNDXH27Nlsj/dbhcLCwlCtWjXUrFlTYdlevXrlejzAu+b2tm3bKpTVqFFDoRsjIiICpqamcHd3V6jXs2dPlbY1e/bsbPtuY2Mjvd6uXbscx3Vs2rQJjRo1gomJCXR0dKCrq4tly5YpdD3v27cPAL641eVDKSkpOH36NLp06aJwxZu2tjb69u2LR48e4ebNmwrLtGvXTuF5jRo1ACBbl9DH9O7dW+F51t88LCwMwLsxcj/99BN2796N6OhoAO8GGe/fvx/Dhg1T6urM69evS+/d4sWLY9q0aZgwYQJ+/PFHqc7u3bshk8nQp08fhfdDsWLFULNmTanb5tSpU0hPT0e3bt0UttGwYcPPXrCSn8c3OTkZ48aNQ4UKFaCjowMdHR2YmJggJSUlx+ELn9tO1t/jw79Xt27doKOTu0OF8+I45Wf8hw8fhp2dHerXr69Q7unpCSFEtotlWrdurTD8Q9X3kDowccond+7cwZEjR9C6dWsIIfDq1Su8evVK+nLPutLuwYMHCkmKrq4uIiIiPrnu+Ph4ADlffVaiRAnpdQDZ1r1q1SoAkL7McjpZw8PDcfbsWfz1119K7euZM2cQHR2dK4nTtGnTFOJVZnB0fHw8ihUrlu1LxcrKCjo6OtLxWLVqVbbjAQClS5cG8K4742tpaWnBwcEh26NSpUrZ4v1QTmW5wcjIKFt3rL6+PtLS0hRisra2zrZsTmWfUq5cuWz7/v7g75zO2a1bt0pTQKxZswYnT57E2bNnMXDgQIUYnz17Bm1t7Vw7Ti9fvoQQ4qPvIwAK7yUg+xWCWfv2+vXrz25PR0cn2/JZ+/L+dgYOHAhDQ0Pp/bdo0SIYGhpi4MCBn90G8O6CgrNnz+LMmTPYtGkTatasiYCAAGzYsEGq8/TpUwghYG1tne09cerUKal7KyuuLzk38vP49urVCwsXLoS3tzcOHDiAM2fO4OzZs7C0tMxx2c9tJyuuD8+1nP6GOVHlMyUvjtPXxq+K+Pj4fHsPqQuvqssny5cvhxACmzdvxubNm7O9vmrVKkyfPh0lSpTA2bNnFV6rXLnyJ9eddeLFxsZme+2///5D0aJFpecfrtvW1hbAu5O6WrVqCA0NRVpamsIXa61atQC8+xWnjC1btqBSpUoqz1OVk8GDB6NNmzbSc2WuuLKwsMDp06chhFBInuLi4pCRkSEdj7Zt22Y7HsC7L3N7e3scPHgQqampuTLO6XPxPnnyJFt5TmX5xcLCIseBo7kdU04tJmvWrIGtrS1CQkIUXv9w8LqlpSUyMzPx5MmTj463UEXhwoWhpaX10fcRAIX30tfKyMhAfHy8whdH1vF9v8zc3Bz9+/fH0qVLMWbMGKxYsQK9evVCoUKFlNpO1gBuAKhXrx5cXV1RrVo1jBw5Em3atIGJiQmKFi0KmUyGo0eP5vgeyyrLiuv9MU/vx/6pVqf8Or4JCQnYvXs3Jk+ejPHjx0vlWWPnvkTWfj958gQlS5aUyrP+hp/TsmVLLFmyBNu3b1eIKSd5cZy+Nn5Vt5Vf7yF1YYtTPsjMzMSqVatQvnx5hIWFZXuMHj0asbGx2LdvH/T09LL9Qjc1NQXw8Uzc0dERhoaGWLNmjUL5o0ePcPjwYTRr1kwq+3Dd739A+/r64vnz5/Dx8VGqC+9jtmzZkmuDwkuUKKEQr729vfSavr5+jr9KmjVrhuTk5GyTQmYNLM46HhYWFtmORxY/Pz+8fPkSI0aMyPFYJCcnSwM+v5arqyuuX7+Oy5cvK5SvW7dOqeU/dhy+hrOzM5KSkqTusCzvt1LkFZlMBj09PYWk6cmTJ9muqvPw8AAALF68+JPrU/b4GBsbo0GDBti6datCfblcjjVr1qBUqVIKLYW54cM5y7L+5u9fnQYAI0aMwPPnz9GlSxe8evVKqQsHPsbCwgKzZs3C06dP8eeffwIA2rRpAyEEHj9+nGMLadb7rkGDBtDX1882QPnUqVOf7VrJr+Mrk8kghMiWAC5dulSprv6cZP09Pvx7bdy4Uaku/vbt28Pe3h4BAQG4du1ajnUOHDiA1NTUPDlOXxu/Kq1AzZo1Q2RkZLYJPVevXg2ZTJZtXrdvEVuc8sG+ffvw33//Yfbs2dk+EAGgevXqWLhwIZYtW6bQupJTPQBYsmQJTE1NYWBgAFtbW1hYWMDPzw+//fYb+vXrh549eyI+Ph5Tp06FgYEBJk+erFScPXv2xPXr1zFjxgxcvnwZnp6eqFixIuRyOWJiYvC///0PAKRELieXLl3C3bt38+xquvfZ29tj69atWLx4MerWrSt1i/Xr1w+LFi1C//798eDBA9jb2+PYsWOYOXMmWrVqhebNm3923V27doWfnx/8/f1x48YNeHl5SRNgnj59Wrr66XNTEsjl8o9eWlu7dm3o6+tj5MiRWL58OVq3bo3p06fD2toaa9euVXqWdnt7e4SHh2PXrl0oXrw4TE1NP9tK+Tn9+/fH/Pnz0adPH0yfPh0VKlTAvn37cODAAQDvuiDzSps2bbB161YMGzYMXbp0QUxMDPz9/VG8eHGFsYCNGzdG3759MX36dDx9+hRt2rSBvr4+Ll68CCMjI/z8888A3h2fDRs2ICQkBOXKlYOBgYFCAv6+gIAAtGjRAq6urhgzZgz09PQQHByMa9euYf369UqNKVKWnp4e5s2bh+TkZNSrVw8nTpzA9OnT4eHhgR9++EGhbqVKleDu7o59+/bhhx9+yDYeTlX9+vVDYGAg5s6di+HDh6NRo0YYPHgwBgwYgHPnzqFJkyYwNjZGbGwsjh07Bnt7ewwdOhRFihSBj48PAgICULhwYXTs2BGPHj3C1KlTUbx48c+eF/lxfM3MzNCkSRP8/vvvKFq0KMqWLYuIiAgsW7ZM6Va6D1WtWhV9+vRBUFAQdHV10bx5c1y7dg1z585VavJHbW1tbNu2DW5ubnB0dMTQoUPh6uoKY2NjPHz4EJs3b8auXbukK9xy+zh9bfxZ75fZs2fDw8MD2traqFGjBvT09LLVHTVqFFavXo3WrVtj2rRpKFOmDPbs2YPg4GAMHTo01398qIUaBqQXOB06dBB6enoiLi7uo3V69OghdHR0xJMnTz65rqCgIGFrayu0tbWzXbGydOlSUaNGDaGnpyfMzc1F+/btxfXr11WO98iRI6J79+6iVKlSQldXVxgZGQk7OzsxdOhQce7cuU8uO3HixByvdFOGqlfVvXjxQnTp0kUUKlRIyGQyhas+4uPjxZAhQ0Tx4sWFjo6OKFOmjJgwYYLSV7VkiYiIEF26dBHFixcXurq6wszMTDg6Oorff/9dJCYmfnLZT11Vh/9/dVmWyMhI0aJFC2FgYCCKFCkivLy8xI4dO5S6qu7SpUuiUaNGwsjISOHKpo9dVWdsbJwt1smTJ2e7aiY6Olp06tRJmJiYCFNTU9G5c2exd+9eAUDs2LHjk/uete1Nmzbl+HrWFTW///57jq/PmjVLlC1bVujr64uqVauKf/75J8cYMzMzxfz580X16tWl897R0VHs2rVLqvPgwQPh5uYmTE1NBQDp+OV01ZcQQhw9elQ0bdpUGBsbC0NDQ9GwYUOF9Qnxf1cznT17Nsf9/vDK1w9l/R2uXLkiXFxchKGhoShSpIgYOnSoSE5OznGZlStXCgBiw4YNn1z3+z71ntqzZ48AIKZOnSqVLV++XDRo0EDa9/Lly4t+/fopvO/lcrmYPn26KFWqlNDT0xM1atQQu3fvFjVr1hQdO3aU6qnz+D569Eh07txZFC5cWJiamgp3d3dx7dq1bFeAqbKd9PR0MXr0aGFlZSUMDAxEw4YNxcmTJ7Ot81NevXol/P39RZ06dYSJiYnQ1dUVpUuXFn369BHHjx/P0+OkbPwfW9bb21tYWlpKn7X3798XQmS/qk4IIR4+fCh69eolLCwshK6urqhcubL4/fffRWZmplTnU58BAMTkyZM/f0DVRCbEV/TJEH3Azs4OHh4emDdvnrpDoTwwc+ZMTJw4EdHR0ShVqpS6wylQOnfujFOnTkkXkGiS+/fvo0qVKpg8eTJ+++03dYdDlKfYVUe5KjIyUt0hUC5ZuHAhAKBKlSp4+/YtDh8+jAULFqBPnz5MmvJJeno6Lly4gDNnzmDbtm0IDAxUe9J0+fJlrF+/Hk5OTjAzM8PNmzcxZ84cmJmZwcvLS62xEeUHJk5ElCMjIyPMnz8fDx48QHp6OkqXLo1x48Zh4sSJ6g6twIiNjZUSlB9//FEat6VOxsbGOHfuHJYtWybdlNbFxQUzZsxQeboKom8Ru+qIiIiIlMTpCIiIiIiUxMSJiIiISElMnIiIiIiUVOAGh8vlcvz3338wNTXN1cnsiIiI6NskhEBSUhJKlCjx2YlcC1zi9N9//yncnZ2IiIgIAGJiYj473UqBS5yybhcSExOj1FTzRERE9H1LTEyEjY3NJ28plqXAJU5Z3XNmZmZMnIiIiEiizBAeDg4nIiIiUhITJyIiIiIlMXEiIiIiUlKBG+NERET0PZPL5Xjz5o26w9Aourq60NbWzpV1qTVxOnLkCH7//XecP38esbGx2LZtGzp06PDJZSIiIuDj44Pr16+jRIkSGDt2LIYMGZI/ARMREWmwN2/e4P79+5DL5eoOReMUKlQIxYoV++o5HNWaOKWkpKBmzZoYMGAAOnfu/Nn69+/fR6tWrTBo0CCsWbMGx48fx7Bhw2BpaanU8kRERN8rIQRiY2Ohra0NGxubz07kWFAIIZCamoq4uDgAQPHixb9qfWpNnDw8PODh4aF0/b/++gulS5dGUFAQAKBq1ao4d+4c5s6dy8SJiIgKtIyMDKSmpqJEiRIwMjJSdzgaxdDQEAAQFxcHKyurr+q2+6bS0ZMnT8LNzU2hrGXLljh37hzevn2rpqiIiIjULzMzEwCgp6en5kg0U1Yy+bX5wjc1OPzJkyewtrZWKLO2tkZGRgaeP3+eY/Nbeno60tPTpeeJiYl5HicREZG68D6sOcut4/JNJU5A9h0XQuRYniUgIABTp07N87iIvgdlx+9Rdwj57sGs1uoOgYi+Id9UV12xYsXw5MkThbK4uDjo6OjAwsIix2UmTJiAhIQE6RETE5MfoRIREdF36JtqcXJ0dMSuXbsUyg4ePAgHBwfo6urmuIy+vj709fXzIzwiIiKNk98tyaq24v7111/49ddf8fLlS+jovEtLkpOTUbhwYTRs2BBHjx6V6h49ehRNmjT57DpXrFgBT09PleJQllpbnJKTk3Hp0iVcunQJwLvpBi5duoTo6GgA71qL+vXrJ9UfMmQIHj58CB8fH0RFRWH58uVYtmwZxowZo47wiYiI6Cu5uroiOTkZ586dk8qOHj2KYsWK4ezZs0hNTZXKw8PDUaxYMcTGxkqPbt26wd3dXaGse/fueRavWhOnc+fOoXbt2qhduzYAwMfHB7Vr18akSZMAALGxsVISBQC2trbYu3cvwsPDUatWLfj7+2PBggWcioCIiOgbVblyZZQoUQLh4eFSWXh4ONq3b4/y5cvjxIkTCuXNmjVDsWLFpIehoSH09fWzleUVtXbVubi4SIO7c7Jy5cpsZc7Ozrhw4UIeRkVERET5ycXFBWFhYRg/fjwAICwsDGPHjoVcLkdYWBiaN2+ON2/e4OTJk/jzzz/VGus3NTiciIiIvj8uLi44fvw4MjIykJSUhIsXL6JJkyZwdnaWWqJOnTqF169fw9XVVa2xflODw4mIiOj74+rqipSUFJw9exYvX75EpUqVYGVlBWdnZ/Tt2xcpKSkIDw9H6dKlUa5cObXGysSJiIiI1KpChQooVaoUwsLC8PLlSzg7OwN4Nw2Rra0tjh8/jrCwMDRt2lTNkbKrjoiIiDSAq6srwsPDER4eDhcXF6nc2dkZBw4cwKlTp9TeTQcwcSIiIiIN4OrqimPHjuHSpUtSixPwLnH6559/kJaWxsSJiIiICHiXOL1+/RoVKlRQuC+ts7MzkpKSUL58edjY2Kgxwnc4xomIiOg79q3cj7Fs2bI5TlFUqlQplacuyktscSIiIiJSEhMnIiIiIiUxcSIiIiJSEhMnIiIiIiUxcSIiIiJSEhMnIiIiIiUxcSIiIiJSEhMnIiIiIiUxcSIiIiJSEhMnIiIiIiXxlitERETfsynm+by9BJWqe3p64tWrV9i+fbtUtnnzZvTp0wfTpk3D6NGjsWDBAqxYsQK3bt2CgYEBHB0dMXHiRDRq1CiXg/88tjgRERGRxli6dCl69+6NhQsX4tdff0WPHj0wbdo0jBgxAlFRUYiIiICNjQ1cXFwUkq38whYnIiIi0ghz5szBpEmTsG7dOnTu3BkhISHYvHkzdu7cibZt20r1lixZgvj4eHh7e6NFixYwNjbOtxjZ4kRERERqN378ePj7+2P37t3o3LkzAGDdunWoVKmSQtKUZfTo0YiPj0doaGi+xskWJyIiIlKrffv2YceOHTh06BCaNm0qld+6dQtVq1bNcZms8lu3buVLjFnY4kRERERqVaNGDZQtWxaTJk1CUlKSSsvKZLI8iipnbHHKA2XH71F3CPnuwazW6g6BiIi+USVLlsSWLVvg6uoKd3d37N+/H6ampqhUqRIiIyNzXCYqKgoAULFixfwMlS1OREREpH6lS5dGREQE4uLi4ObmhsTERPTo0QO3b9/Grl27stWfN28eLCws0KJFi3yNk4kTERERaYRSpUohPDwc8fHxcHNzQ6tWrdCxY0f0798fy5Ytw4MHD3DlyhX8+OOP2LlzJ5YuXZqvV9QBTJyIiIhIg5QsWRIRERF49eoVWrRogSVLlsDX1xfz589HlSpV0LhxYzx8+BBhYWHo0KFDvsfHMU5ERETfMxVn8s5vK1euzFZWvHhx3LhxQ3o+evRojB49Oh+j+ji2OBEREREpSeXE6enTp+jbty9KlCgBHR0daGtrKzyIiIiIvlcqd9V5enoiOjoafn5+KF68eL7Pn0BERESkLionTseOHcPRo0dRq1atPAiHiIiISHOp3FVnY2MDIURexEJERESk0VROnIKCgjB+/Hg8ePAgD8IhIiKir8HGjZzJ5fJcWY/KXXXdu3dHamoqypcvDyMjI+jq6iq8/uLFi1wJjIiIiJSnq6sLmUyGZ8+ewdLSkmOQ/z8hBN68eYNnz55BS0sLenp6X7U+lROnoKCgr9ogERER5T5tbW2UKlUKjx49Yq9QDoyMjFC6dGloaX3dTEwqJ079+/f/qg1+KDg4GL///jtiY2NRrVo1BAUFoXHjxh+tv3btWsyZMwe3b9+Gubk53N3dMXfuXFhYWORqXERERN8aExMTVKxYEW/fvlV3KBpFW1sbOjo6udIK90Uzh2dmZmL79u2IioqCTCaDnZ0d2rVrp/I8TiEhIRg5ciSCg4PRqFEj/P333/Dw8EBkZCRKly6drf6xY8fQr18/zJ8/H23btsXjx48xZMgQeHt7Y9u2bV+yK0RERN8VzquYt1ROnO7cuYNWrVrh8ePHqFy5MoQQuHXrFmxsbLBnzx6UL19e6XUFBgbCy8sL3t7eAN51Ax44cACLFy9GQEBAtvqnTp1C2bJlMWLECACAra0tfvzxR8yZM0fV3SAiIiJSmcodfSNGjED58uURExODCxcu4OLFi4iOjoatra2U0CjjzZs3OH/+PNzc3BTK3dzccOLEiRyXcXJywqNHj7B3714IIfD06VNs3rwZrVu3VnU3iIiIiFSmcotTREQETp06hSJFikhlFhYWmDVrFho1aqT0ep4/f47MzExYW1srlFtbW+PJkyc5LuPk5IS1a9eie/fuSEtLQ0ZGBtq1a4c///zzo9tJT09Henq69DwxMVHpGImIiIjep3KLk76+PpKSkrKVJycnf9Elfh8O1BJCfHTwVmRkJEaMGIFJkybh/Pnz2L9/P+7fv48hQ4Z8dP0BAQEwNzeXHjY2NirHSERERAR8QeLUpk0bDB48GKdPn4YQAkIInDp1CkOGDEG7du2UXk/RokWhra2drXUpLi4uWytUloCAADRq1Ai//voratSogZYtWyI4OBjLly9HbGxsjstMmDABCQkJ0iMmJkb5nSUiIiJ6j8qJ04IFC1C+fHk4OjrCwMAABgYGaNSoESpUqIA//vhD6fXo6emhbt26CA0NVSgPDQ2Fk5NTjsukpqZmm38h68qBj82Uqq+vDzMzM4UHERER0ZdQeYxToUKFsGPHDty+fRs3btyAEAJ2dnaoUKGCyhv38fFB37594eDgAEdHRyxZsgTR0dFS19uECRPw+PFjrF69GgDQtm1bDBo0CIsXL0bLli0RGxuLkSNHon79+ihRooTK2yciIiJSxRfN4wQAFStWRMWKFb9q4927d0d8fDymTZuG2NhYVK9eHXv37kWZMmUAALGxsYiOjpbqe3p6IikpCQsXLsTo0aNRqFAhNG3aFLNnz/6qOIiIiIiUIRNK3A3Qx8cH/v7+MDY2ho+PzyfrBgYG5lpweSExMRHm5uZISEjIs267suP35Ml6NdmDWZwS4nvAc5eICiJVcgOlWpwuXrwoTd9+8eLFr4+QiIiI6BukVOIUFhaW4/+JiIiIChKVr6obOHBgjvM4paSkYODAgbkSFBEREZEmUjlxWrVqFV6/fp2t/PXr19LVb0RERETfI6WvqktMTJQmvExKSoKBgYH0WmZmJvbu3QsrK6s8CZKIiIhIEyidOBUqVAgymQwymQyVKlXK9rpMJsPUqVNzNTgiIiIiTaJ04hQWFgYhBJo2bYotW7Yo3ORXT08PZcqU4SSURERE9F1TOnFydnYGANy/fx+lS5f+6I14iYiIiL5XKg8OP3z4MDZv3pytfNOmTVi1alWuBEVERESkiVROnGbNmoWiRYtmK7eyssLMmTNzJSgiIiIiTaRy4vTw4UPY2tpmKy9TpozCfeWIiIiIvjcqJ05WVla4cuVKtvLLly/DwsIiV4IiIiIi0kQqJ049evTAiBEjEBYWhszMTGRmZuLw4cP45Zdf0KNHj7yIkYiIiEgjKH1VXZbp06fj4cOHaNasGXR03i0ul8vRr18/jnEiIiKi75rKiZOenh5CQkLg7++Py5cvw9DQEPb29ihTpkxexEdERESkMVROnLJUqlQpxxnEiYiIiL5XX5Q4PXr0CDt37kR0dDTevHmj8FpgYGCuBEZERESkaVROnA4dOoR27drB1tYWN2/eRPXq1fHgwQMIIVCnTp28iJGIiIhII6h8Vd2ECRMwevRoXLt2DQYGBtiyZQtiYmLg7OyMrl275kWMRERERBpB5cQpKioK/fv3BwDo6Ojg9evXMDExwbRp0zB79uxcD5CIiIhIU6icOBkbGyM9PR0AUKJECdy9e1d67fnz57kXGREREZGGUXmMU8OGDXH8+HHY2dmhdevWGD16NK5evYqtW7eiYcOGeREjERERkUZQOXEKDAxEcnIyAGDKlClITk5GSEgIKlSogPnz5+d6gERERESaQqXEKTMzEzExMahRowYAwMjICMHBwXkSGBEREZGmUWmMk7a2Nlq2bIlXr17lUThEREREmkvlweH29va4d+9eXsRCREREpNFUTpxmzJiBMWPGYPfu3YiNjUViYqLCg4iIiOh7pfLgcHd3dwBAu3btIJPJpHIhBGQyGTIzM3MvOiIiIiINonLiFBYWlhdxEBEREWk8pRKnTp06YeXKlTAzM8PDhw/RvXt36Ovr53VsRERERBpFqTFOu3fvRkpKCgBgwIABSEhIyNOgiIiIiDSRUi1OVapUwYQJE+Dq6gohBDZu3AgzM7Mc6/br1y9XAyQiIiLSFEolTn/99Rd8fHywZ88eyGQyTJw4UWFgeBaZTMbEiYiIiL5bSiVOTk5OOHXqFABAS0sLt27dgpWVVZ4GRkRERKRpVJ7H6f79+7C0tMyLWIiIiIg0msrTEZQpUyYv4iAiIiLSeCq3OOW24OBg2NrawsDAAHXr1sXRo0c/WT89PR2+vr4oU6YM9PX1Ub58eSxfvjyfoiUiIqKCTOUWp9wUEhKCkSNHIjg4GI0aNcLff/8NDw8PREZGonTp0jku061bNzx9+hTLli1DhQoVEBcXh4yMjHyOnIiIiAoitSZOgYGB8PLygre3NwAgKCgIBw4cwOLFixEQEJCt/v79+xEREYF79+6hSJEiAICyZcvmZ8hERERUgKmtq+7Nmzc4f/483NzcFMrd3Nxw4sSJHJfZuXMnHBwcMGfOHJQsWRKVKlXCmDFj8Pr1649uJz09nTciJiIiolyhcotT7dq1PzqHk4GBASpUqABPT0+4urp+cj3Pnz9HZmYmrK2tFcqtra3x5MmTHJe5d+8ejh07BgMDA2zbtg3Pnz/HsGHD8OLFi4+OcwoICMDUqVOV3DsiIiKij1O5xcnd3R337t2DsbExXF1d4eLiAhMTE9y9exf16tVDbGwsmjdvjh07dii1vg+TMCFEjokZAMjlcshkMqxduxb169dHq1atEBgYiJUrV3601WnChAlISEiQHjExMartMBEREdH/p3KL0/PnzzF69Gj4+fkplE+fPh0PHz7EwYMHMXnyZPj7+6N9+/YfXU/RokWhra2drXUpLi4uWytUluLFi6NkyZIwNzeXyqpWrQohBB49eoSKFStmW0ZfX583JCYiIqJcoXKL08aNG9GzZ89s5T169MDGjRsBAD179sTNmzc/uR49PT3UrVsXoaGhCuWhoaFwcnLKcZlGjRrhv//+Q3JyslR269YtaGlpoVSpUqruChEREZFKVE6cDAwMchy8feLECRgYGAB416WmTCuPj48Pli5diuXLlyMqKgqjRo1CdHQ0hgwZAuBdN9v7977r1asXLCwsMGDAAERGRuLIkSP49ddfMXDgQBgaGqq6K0REREQqUbmr7ueff8aQIUNw/vx51KtXDzKZDGfOnMHSpUvx22+/AQAOHDiA2rVrf3Zd3bt3R3x8PKZNm4bY2FhUr14de/fulWYnj42NRXR0tFTfxMQEoaGh+Pnnn+Hg4AALCwt069YN06dPV3U3iIiIiFQmE0IIVRdau3YtFi5cKHXHVa5cGT///DN69eoFAHj9+rV0lZ2mSUxMhLm5ORISEmBmZpYn2yg7fk+erFeTPZjVWt0hUC7guUtEBZEqucEXTYDZu3dv9O7d+6Ovs9uMiIiIvkdfPHP4mzdvEBcXB7lcrlD+sVulEBEREX3rVE6cbt++jYEDB2YbIJ41/1JmZmauBUdERESkSVROnDw9PaGjo4Pdu3ejePHiH52skoiIiOh7o3LidOnSJZw/fx5VqlTJi3iIiIiINJbK8zjZ2dnh+fPneRELERERkUZTOXGaPXs2xo4di/DwcMTHxyMxMVHhQURERPS9Urmrrnnz5gCAZs2aKZRzcDgRERF971ROnMLCwvIiDiIiIiKNp3Li5OzsnBdxEBEREWk8pRKnK1euoHr16tDS0sKVK1c+WbdGjRq5EhgRERGRplEqcapVqxaePHkCKysr1KpVCzKZDDnd4o5jnIiIiOh7plTidP/+fVhaWkr/JyIiIiqIlEqcypQpI/3/4cOHcHJygo6O4qIZGRk4ceKEQl0iIiKi74nK8zi5urrixYsX2coTEhLg6uqaK0ERERERaSKVE6es+Zo+FB8fD2Nj41wJioiIiEgTKT0dQadOnQC8GwDu6ekJfX196bXMzExcuXIFTk5OuR8hERERkYZQOnEyNzcH8K7FydTUFIaGhtJrenp6aNiwIQYNGpT7ERIRERFpCKUTpxUrVgAAypYtizFjxrBbjoiIiAoclWcOnzx5cl7EQURERKTxlEqc6tSpg0OHDqFw4cKoXbt2joPDs1y4cCHXgiMiIiLSJEolTu3bt5cGg3fo0CEv4yEiIiLSWEolTu93z7GrjoiIiAoqledxIiIiIiqolGpxKly48CfHNb0vp1nFiYiIiL4HSiVOQUFB0v/j4+Mxffp0tGzZEo6OjgCAkydP4sCBA/Dz88uTIImIiIg0gVKJU//+/aX/d+7cGdOmTcNPP/0klY0YMQILFy7Ev//+i1GjRuV+lEREREQaQOUxTgcOHIC7u3u28pYtW+Lff//NlaCIiIiINJHKiZOFhQW2bduWrXz79u2wsLDIlaCIiIiINJHKM4dPnToVXl5eCA8Pl8Y4nTp1Cvv378fSpUtzPUAiIiIiTaFy4uTp6YmqVatiwYIF2Lp1K4QQsLOzw/Hjx9GgQYO8iJG+BVPM1R1B/puSoO4IiIgon6mcOAFAgwYNsHbt2tyOhYiIiEijfdEEmHfv3sXEiRPRq1cvxMXFAQD279+P69ev52pwRERERJpE5cQpIiIC9vb2OH36NLZs2YLk5GQAwJUrV3g7FiIiIvquqZw4jR8/HtOnT0doaCj09PSkcldXV5w8eTJXgyMiIiLSJConTlevXkXHjh2zlVtaWiI+Pl7lAIKDg2FrawsDAwPUrVsXR48eVWq548ePQ0dHB7Vq1VJ5m0RERERfQuXEqVChQoiNjc1WfvHiRZQsWVKldYWEhGDkyJHw9fXFxYsX0bhxY3h4eCA6OvqTyyUkJKBfv35o1qyZStsjIiIi+hoqJ069evXCuHHj8OTJE8hkMsjlchw/fhxjxoxBv379VFpXYGAgvLy84O3tjapVqyIoKAg2NjZYvHjxJ5f78ccf0atXL2keKSIiIqL8oHLiNGPGDJQuXRolS5ZEcnIy7Ozs0KRJEzg5OWHixIlKr+fNmzc4f/483NzcFMrd3Nxw4sSJjy63YsUK3L17lwPRiYiIKN+pPI+Trq4u1q5dC39/f1y4cAFyuRy1a9dGxYoVVVrP8+fPkZmZCWtra4Vya2trPHnyJMdlbt++jfHjx+Po0aPQ0VEu9PT0dKSnp0vPExMTVYqTiIiIKMsXTYAJAOXKlUO5cuW+OgCZTKbwXAiRrQwAMjMz0atXL0ydOhWVKlVSev0BAQGYOnXqV8dJRERE9EUTYOaGokWLQltbO1vrUlxcXLZWKABISkrCuXPn8NNPP0FHRwc6OjqYNm0aLl++DB0dHRw+fDjH7UyYMAEJCQnSIyYmJk/2h4iIiL5/X9zi9LX09PRQt25dhIaGKkxvEBoaivbt22erb2ZmhqtXryqUBQcH4/Dhw9i8eTNsbW1z3I6+vj709fVzN3giIiIqkNSWOAGAj48P+vbtCwcHBzg6OmLJkiWIjo7GkCFDALxrLXr8+DFWr14NLS0tVK9eXWF5KysrGBgYZCsnIiIiygtqTZy6d++O+Ph4TJs2DbGxsahevTr27t2LMmXKAABiY2M/O6cTERERUX6RCSGEspUTExNhZmYGANi7dy8yMjKk17S1tdG6devcjzCXJSYmwtzcHAkJCdK+5Lay4/fkyXo12QODXuoOIf9NSVB3BLmuQJ67szT/c4uI8pYquYHSLU67d++Gn58fLl68COBda1FKSor0ukwmQ0hICLp06fKFYRMRERFpNqWvqluyZAl++uknhbI7d+5ALpdDLpcjICAAy5cvz/UAiYiIiDSF0onTlStXULNmzY++7uHhgXPnzuVKUERERESaSOnE6cmTJ7CwsJCeh4WFwcbGRnpuYmKChITvb8wHERERURalE6ciRYrg7t270nMHBwfo6upKz2/fvo0iRYrkbnREREREGkTpxKlJkyZYsGDBR19fsGABmjRpkitBEREREWkipROncePG4eDBg+jatSvOnj0r3cLkzJkz6Ny5M/7991+MGzcuL2MlIiIiUiulpyOoXbs2QkJC4O3tja1btyq8VrhwYWzYsAF16tTJ9QCJiIiINIVKM4e3b98eLVq0wIEDB3D79m0AQMWKFeHm5gZjY+M8CZCIiIhIU6h8yxUjIyOFm/ISERERFRRKj3HKybBhw/D8+fPcioWIiIhIo31V4rRmzRokJibmVixEREREGu2rEicV7g9MRERE9M37qsSJiIiIqCBReXD4+5KSknIrDiIiIiKN98WJ07Nnz3Dz5k3IZDJUqlQJlpaWuRkXERERkcZRuasuJSUFAwcORIkSJdCkSRM0btwYJUqUgJeXF1JTU/MiRiIiIiKNoHLi5OPjg4iICOzcuROvXr3Cq1evsGPHDkRERGD06NF5ESMRERGRRlC5q27Lli3YvHkzXFxcpLJWrVrB0NAQ3bp1w+LFi3MzPiIiIiKNoXKLU2pqKqytrbOVW1lZsauOiIiIvmsqJ06Ojo6YPHky0tLSpLLXr19j6tSpcHR0zNXgiIiIiDSJyl11QUFB8PDwQKlSpVCzZk3IZDJcunQJBgYGOHDgQF7ESERERKQRVE6c7O3tcfv2baxZswY3btyAEAI9evRA7969YWhomBcxEhEREWkElROnI0eOwMnJCYMGDVIoz8jIwJEjR9CkSZNcC46IiIhIk6g8xsnV1RUvXrzIVp6QkABXV9dcCYqIiIhIE6mcOAkhIJPJspXHx8fD2Ng4V4IiIiIi0kRKd9V16tQJACCTyeDp6Ql9fX3ptczMTFy5cgVOTk65HyERERGRhlA6cTI3NwfwrsXJ1NRUYSC4np4eGjZsmG3cExEREdH3ROnEacWKFQCAsmXLYsyYMeyWIyIiogJH5TFOkydPlpKmWbNm4dWrV7kdExEREZFGUjlxet/MmTNzvMKOiIiI6Hv0VYmTECK34iAiIiLSeF+VOBEREREVJF+VOEVGRqJMmTLS80uXLn1tPEREREQa66sSJxsbGyQnJyM4OBh16tRB3bp1cysuIiIiIo3zxYnT4cOH0adPHxQvXhx//vknWrVqhXPnzuVmbEREREQaRaXE6dGjR5g+fTrKlSuHnj17onDhwnj79i22bNmC6dOno3bt2ioHEBwcDFtbWxgYGKBu3bo4evToR+tu3boVLVq0gKWlJczMzODo6IgDBw6ovE0iIiKiL6F04tSqVSvY2dkhMjISf/75J/777z/8+eefX7XxkJAQjBw5Er6+vrh48SIaN24MDw8PREdH51j/yJEjaNGiBfbu3Yvz58/D1dUVbdu2xcWLF78qDiIiIiJlKD1z+MGDBzFixAgMHToUFStWzJWNBwYGwsvLC97e3gCAoKAgHDhwAIsXL0ZAQEC2+kFBQQrPZ86ciR07dmDXrl1f1NpFREREpAqlW5yOHj2KpKQkODg4oEGDBli4cCGePXv2xRt+8+YNzp8/Dzc3N4VyNzc3nDhxQql1yOVyJCUloUiRIl8cBxEREZGylE6cHB0d8c8//yA2NhY//vgjNmzYgJIlS0IulyM0NBRJSUkqbfj58+fIzMyEtbW1Qrm1tTWePHmi1DrmzZuHlJQUdOvW7aN10tPTkZiYqPAgIiIi+hIqX1VnZGSEgQMH4tixY7h69SpGjx6NWbNmwcrKCu3atVM5AJlMpvBcCJGtLCfr16/HlClTEBISAisrq4/WCwgIgLm5ufSwsbFROUYiIiIi4CvncapcuTLmzJmDR48eYf369SotW7RoUWhra2drXYqLi8vWCvWhkJAQeHl5YePGjWjevPkn606YMAEJCQnSIyYmRqU4iYiIiLLkyi1XtLW10aFDB+zcuVPpZfT09FC3bl2EhoYqlIeGhsLJyemjy61fvx6enp5Yt24dWrdu/dnt6Ovrw8zMTOFBRERE9CWUvqouL/j4+KBv375wcHCAo6MjlixZgujoaAwZMgTAu9aix48fY/Xq1QDeJU39+vXDH3/8gYYNG0qtVYaGhjA3N1fbfhAREVHBoNbEqXv37oiPj8e0adMQGxuL6tWrY+/evdL972JjYxXmdPr777+RkZGB4cOHY/jw4VJ5//79sXLlyvwOn4iIiAoYtSZOADBs2DAMGzYsx9c+TIbCw8PzPiAiIiKij8iVMU5EREREBQETJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlqT1xCg4Ohq2tLQwMDFC3bl0cPXr0k/UjIiJQt25dGBgYoFy5cvjrr7/yKVIiIiIq6NSaOIWEhGDkyJHw9fXFxYsX0bhxY3h4eCA6OjrH+vfv30erVq3QuHFjXLx4Eb/99htGjBiBLVu25HPkREREVBCpNXEKDAyEl5cXvL29UbVqVQQFBcHGxgaLFy/Osf5ff/2F0qVLIygoCFWrVoW3tzcGDhyIuXPn5nPkREREVBDpqGvDb968wfnz5zF+/HiFcjc3N5w4cSLHZU6ePAk3NzeFspYtW2LZsmV4+/YtdHV1sy2Tnp6O9PR06XlCQgIAIDEx8Wt34aPk6al5tm5NlSgT6g4h/+XhOaQuBfLc/Q7/jkSkmqzPASE+/12mtsTp+fPnyMzMhLW1tUK5tbU1njx5kuMyT548ybF+RkYGnj9/juLFi2dbJiAgAFOnTs1WbmNj8xXR04fM1R2AOswqkHv93TEPUncERKQpkpKSYG7+6c92tSVOWWQymcJzIUS2ss/Vz6k8y4QJE+Dj4yM9l8vlePHiBSwsLD65HVJeYmIibGxsEBMTAzMzM3WHQ6Q0nrv0LeJ5m/uEEEhKSkKJEiU+W1dtiVPRokWhra2drXUpLi4uW6tSlmLFiuVYX0dHBxYWFjkuo6+vD319fYWyQoUKfXng9FFmZmZ8E9M3iecufYt43uauz7U0ZVHb4HA9PT3UrVsXoaGhCuWhoaFwcnLKcRlHR8ds9Q8ePAgHB4ccxzcRERER5Sa1XlXn4+ODpUuXYvny5YiKisKoUaMQHR2NIUOGAHjXzdavXz+p/pAhQ/Dw4UP4+PggKioKy5cvx7JlyzBmzBh17QIREREVIGod49S9e3fEx8dj2rRpiI2NRfXq1bF3716UKVMGABAbG6swp5OtrS327t2LUaNGYdGiRShRogQWLFiAzp07q2sXCO+6QydPnpytS5RI0/HcpW8Rz1v1kgllrr0jIiIiIvXfcoWIiIjoW8HEiYiIiEhJTJyIiIiIlMTEiYiIiEhJTJzos3j9AH1LeL7St4jn7beDiRN91vu3puGbmzSdTCbD27dv1R0GkUqyPmflcrmaI6HPYeJEn/X69Wv4+fkhLS0NMpmMyRNptE2bNmHAgAHS8/379yM1NVWNERF93tOnT/Hjjz9i27ZtAIBz587hxo0bao6KcsLEiSRCiBx/7Zw6dQr//PMPpk6dKtUj0lSxsbG4desWhgwZAiMjI8ycORMZGRnqDovok1JTU/Ho0SOsXLkSbdu2Rf369XHr1i11h0U5YOJUwF24cAEeHh4A3jUVa2m9OyWePn0q1alXrx5GjBiBdevW4c6dO9DS0mLyRBonMzMTANCkSRPcuXMHy5cvx+TJk3HkyBHeCJU0mhACtra2KFq0KPbs2YNHjx7h2bNnaNeunbpDoxwwcSrgMjMzceDAASxevBgAkJaWhoCAALi7u0utTyYmJmjVqhVsbW0xceJEAIrjnojU5f2WJG1tbQDAw4cP0aFDB9SpU0dK8JnokyaRy+UKrfsymQxxcXGwtbWFm5sbrK2tkZSUBOD/fhCQ5mDiVIDk1A1Xu3Zt+Pn5wdfXFykpKTAwMECxYsUgl8vx559/SvXs7e3Ru3dvHD16FP/+++9H10eUH7LOPR2dd7fbPHbsGK5fvw4AaN++PZYvX4569ephz549OHPmDGQyGc9X0giZmZnQ0tKClpYW7t+/jxMnTuDFixewsrLClClTMHLkSKSkpCA4OBjA//0gIM3BxKkAyeqGi4qKksp0dHTg5eWFwoUL49dffwUAtG7dGs7OzlixYgViY2MBvHvzurq6wtLSEr///rvC+ojyW9a5t2/fPpQvXx5DhgxBkyZN8Msvv+DatWsAgM6dO0NXVxfLly9XWIZInbS1tZGWloYBAwagfv36GDRoEJo1ayadp+7u7qhfvz6OHz+OiIgIAGwx1TT8JClAkpKSYG9vj1q1aiEkJES6ZLtEiRLw8/PD33//jcjISFhZWaFt27YwNjbGrFmzpOUtLS2RmZmJ0NBQzJ07V127QQQhBBYtWoSRI0di8ODBOH36NNauXYvr169j9uzZAAAXFxc4Ozvj2rVr2LRpEwDg/PnzuH37tjpDpwLu/v37aN26NZ48eYJ9+/bh2LFjaNmyJYKCgnD48GEAQLdu3WBmZiYlUwBw+PBhpKWlqStseg8TpwIkLS0NderUQZ8+fTB//nxpigEdHR20b98eP/zwA37++WcAwA8//IC2bdti69at2LVrFwAgPDwc9erVw++//45atWqpcU+oIMlpjIdMJoOFhQWGDh2KcePGwcjICKGhoTh69CgOHz6M1atXAwB69uyJkiVLYuzYsWjTpg3q1auHmzdv5vcuUAEkl8tzPHctLS1RtWpVrF69Gg4ODoiMjMTGjRsRHR0tJf0NGjSAh4cHTp48iZYtW6JQoUIICAhgd7OmEFRgpKamipo1a4rw8HBx/Phx4erqKtzd3UV0dLQQQohDhw4JLS0tsW3bNiGEEHfv3hXe3t5CV1dX1K5dWxgbG4t169apcQ+oIJHL5UIul0vPL126JB4/fiw9z8jIEG/evBERERGiYsWKwtnZWWzevFm0bdtWODo6ipSUFCGEEJGRkeL3338Xo0ePFk+ePMn3/aCCJzMzU/p/bGysuHnzpkhISJDK4uLixNu3b4W3t7coXry48Pf3FwsWLBA2NjZiyZIlQgghXrx4IUJDQ4W3t7fYsGFDvu8DfRwTpwIiIyNDCCHEoEGDhJeXlxDi3Zu3bdu2omPHjiI8PFwIIcSAAQNE2bJlFZbbvn27+PPPP8XTp0/zP3Aq8MLDw4WdnZ2oUaOGsLa2FhMmTBDXrl0TQgiRnJwsOnbsKH799Vfx6tUrIYQQXl5eQkdHR0yePDnH9b19+za/QqcCLDMzUwwbNkxYWloKBwcHUbNmTbF582bp9c2bN4sGDRqIEydOCCGEuH79ujAzMxO1atX66Gdt1uc4qRe76goIbW1tCCFQoUIFaWyTpaUlPDw8sH37dvTt2xc7d+7Ezz//DJlMhilTpkjLtW/fHj/99BOsrKzUuAdUEJ0+fRrDhw9Hx44dER4ejtWrV+P06dNSl/LLly9x8uRJVKxYEebm5khOTsbr16/h5OSEI0eOSJd0Z5HL5dKVeER55e3bt/jll19w6dIlbN68GaGhofDw8ICPjw+2bt0K4N3M4ImJiXB0dAQAXL9+Hfb29hBCIDQ0VGF9WV10vMJOM/ATpIAQQkAmk6Fw4cK4d+8e4uLi4OnpicOHD2PmzJm4c+cOZs2ahWrVqqFz585Yu3YtfH19oaurq+7QqQDIzMxU+FKQy+XQ0tJCWFgYSpUqhenTpwMATp48iePHj8Pb2xtpaWnIyMhAgwYNsGLFCpQsWRLr1q1DSkoK5s+fjzp16mTbDq+so9wk3vXaSJMCZ81vFx8fj8OHD2Pp0qVwdHTExYsXsX37dhgbG6Nw4cIA3l2Uo62tDT8/P5QrVw4LFy7EoEGD0KlTp2w/Unneahh1NndR/ouOjhba2tpCJpOJ9u3biwsXLggh3nV5rFmzRujr64s9e/aoOUoqSN4fx5SYmCj9PyMjQ3To0EEEBweLTZs2ieLFiwsHBwfx77//Kiy/b98+4e7uLmxtbUXz5s1FTEyMwjqI8sL745hSU1MVXtu/f79wcHAQd+/eFZ06dRLm5uZi3Lhx4uXLl1Kdp0+fCl9fX2FnZydsbW3FokWLpNc+HN9HmkUmBCeIKEhu3LiB3r17o1evXhg9enS215OSkmBqaqqGyKigyWpVAt7dD3H8+PHQ0tJCy5Yt0aFDB1SuXBm//vor5s2bhzJlymDs2LEYNGgQdHR0kJaWhp07d6J06dJo2LAhMjMz8fjxY5QuXTrbuoly04fn1pgxY3DkyBFUq1YNrVq1QteuXZGcnAxra2u8fv0aPXr0wOTJk1G5cmUA77rkTp8+jV69esHAwAAxMTEoVaqU1FrFc1fz8a9TwFSpUgUJCQlSF9yHl8syaaL8oqWlhdevXyMsLAzDhw9HjRo1UK1aNaxbtw5t27bF27dv4enpiWLFiqFv374YOnSoND4pLCwMq1atQlJSEuRyObS1taWkKWtmZqK8kHVu3bp1CxMnTkR4eDh69eqF6Oho/PjjjwgODoaJiQmGDh2KIkWKYNGiRVLSlJiYiIULF+LatWtITU0FANjY2EAmk0mfxTx3NR9bnAqgfv364eHDh9KstET5Qbw3BgQArly5gnHjxuHp06do0aKFNIfNnTt30KxZM7Rv3x4LFizA3LlzMXHiRHTo0AFNmzbFyZMnsWXLFvzyyy+YMmUKB8xSvsrIyEBAQAB27NgBQ0ND/P3337Czs8Pr168RGBiIOXPm4OrVqzAwMEDdunVRpkwZODg4oFKlSggMDEThwoXx999/5zgGj74NTG0LoHLlyqFDhw6cxp/yRdYNTT+8MXSpUqVQpEgRREZGom7dugAgXfkZEBCA4OBgxMfHY8yYMVi4cCEAYMeOHYiPj8fp06fh7+/PpIny1Ps3kc6io6ODEiVKICkpCUlJSbCzswMAGBoaYtSoUShZsiTmzJkDKysr7N+/H40aNcLt27cREhKCoUOH4uzZs0yavnFscSqA2IdO+eX9VqaoqCjs3LkTNWvWRLVq1WBjY4PDhw9j8ODBaNy4MVasWCEt9/jxYzRu3Bh+fn4YMGCAVJ6QkABzc3MA/9cl92FCRpTbli9fLnUHu7q6Ij09XZpaYOfOnahXr550ZejUqVOxZ88ehIeHw8jICADw5s0baGtrS4n+h1eR0reF354FEJMmymtZv8dkMhkyMjIwYsQINGjQAAcPHsSvv/4KNzc3XLlyBU2bNkXnzp0RGRmJgwcPSss/f/4cSUlJKF++PID/m8fm/aRJW1ubSRPlqqzzNuvfiIgI2NraIjAwEP/73//QsWNHzJgxA7q6uujXrx8qVKiAJUuWAPi/OZZu3rwJGxsbGBkZSeOWdHV1oa2tzfmYvhP8BiWiXPN+wpRl//79OHHiBMLDw3Ho0CFcvXoVDx48wOTJk5GamoqePXtCX18fI0aMwI4dOxAVFYXg4GBYWlrCxsYGQPZkn188lNve707O+nf+/Pno1KkTrl27hn///Re+vr7w8/PDrl270KBBAzRt2hRbtmyBr68vjh8/jgMHDuD48eNwcHAA8H/nadb6+KP1O6GOORCI6Pvz/rw227ZtE3369BFCCNG+fXsxceJEIcS720zUqlVLlCtXTuzevVuq/8cff4giRYoIOzs70b17d1G7dm1x6dKl/N0BKpDeP2/T0tJE//79xaFDh8SFCxdE6dKlhRBCJCUlCW9vb2FmZiaGDh0q4uLihBBC3Lp1S7i4uIhChQqJjh07ilq1aonffvtNLftB+YczhxNRrtDS0sKrV69w6tQpTJkyBR4eHgCAIkWKICoqCt27d8e///6LkSNHwsfHB8bGxkhMTISZmRk6d+6Mo0ePQgiB8ePHo1atWgA4Ho/yXtb5FRkZib/++gu3b99GkSJFoKOjAz09PUydOhWLFy9G1apVERoaivr16wN4N7VAxYoV0aVLFwCAo6OjdDsVgOfu94x/VSL6IuKD60oiIiLQoUMHTJo0CV5eXggICIBcLkfFihURGhqKly9f4v79+/Dz84OxsTGePXuGiRMn4vr16yhZsiTatGmDx48f4/DhwwDe3e+LXzyU11JTUzFq1Cj07dsX9+7dw7Zt21CrVi1kZmaiaNGiCAwMxMKFCxEWFiYlTf/73//w999/AwC6dOmCEiVK4PDhw4iNjQXAc/d7x78sEalELpdnm5MJeDe9gKmpKa5fv47GjRsDePdrvkmTJqhTpw4MDQ1hZmaGN2/eIDk5GYGBgThx4gRevnwJAOjatStq1qyJNWvW4PLly7xPIuULIyMjlCxZEk+fPoW2trZ0n7gaNWqgcePG0nmd5fDhw/jrr78QFxeH1NRUWFtbo0uXLnj+/DnmzJkDADx3v3PsqiMipWS1MGX9kj558iSOHj2KZs2aoWrVqihfvjx69eqFy5cvY+vWrVJ3W6NGjTBkyBD4+vqiTJkyqFOnDi5evAhTU1MsW7ZM+hVvZGSEdu3aSTejJsprWd1pAwYMwJkzZ3D16lU8fPgQZcqUgUwmg7e3N5KSktC2bVs4OTlBX18fR44cwZgxY+Dv7y+tp23btggLC0Pt2rXVuDeUXziPExGpRAiBMWPG4O+//0bZsmWRnJyMDh06ICgoCElJSRg1ahTu3r2LhQsXolq1atJy0dHRCA0NRXR0NCpXroxevXpJ6xP//w7zRLlJmXFGWa2nW7Zswbx589CyZUtMnjxZoc6mTZsQFxeHuLg4/PjjjyhRooS0fuDdj4mMjAzplkD0fWPiRERKW7hwIR4+fAhtbW0MHToUpqamWLt2LcaOHYs9e/agadOm2LNnD+bMmYO6desiMDAQwMe/wDgRIOWXnLqX3y9/+/YtRo0ahStXriAwMBAODg54+/Ztjt1unHy1YONPPCLKJmsc0/uSkpJw584dzJs3D48fP0bp0qVRpEgR9O7dG23atMHw4cMBAK1bt4aTkxPOnj2LvXv3AkC2L5isdTNporyUlpaGUaNG4cKFCwo30n2fTCaDXC6Hrq4uevToAQMDA2ngd05JU9ZNpZk0FVxMnIhIwfu/puPi4qQvG1NTU/Tr1w81a9ZEWlqa9IVTpEgRjBo1CjExMVi0aBEAoFOnTkhPT8exY8cAZE+c+KVDuSnrfogfunfvHo4dO4ZJkyYB+HiintUa+sMPP6BmzZq4dOkSIiMjP1mXCi521RFRNgkJCRg8eDAuXryI4sWLw8nJCf7+/tDR0UFwcDBGjx6NkydPSgPAX79+jenTpyMoKAiJiYnQ1tbGlStXUKNGDfXuCH237ty5g/bt2yMsLEy6Eg4Abt26hUqVKknP16xZA19fXwQGBqJz584f7R7O6k6OjY2Fjo4OLC0t82U/6NvD1JmIFDx8+BDNmjVDSkoKfv/9dzRv3hwLFy6Ej48P4uPj0bVrVzRu3BijRo2SljE0NESvXr1QqFAhaRLArKQpp5YAoq9laGiI5ORkjB8/HgCQkZGBJUuWwMXFBbdv35bqOTs7w9XVFTNmzADwrtUpp/aCrPsqFi9eHJaWljl2VxMBTJyICqwPvxiy/n/t2jXEx8dj/vz5aN++Pfz8/LBgwQJERERg9+7dsLS0xPDhw3H16lVs2rRJWr5q1aqIjIxE165dFbbDrg36Wjkl3yVLlsS8efOwcuVKnD17Fjo6OrC1tUWlSpUwd+5cqZ6NjQ26deuGpKQkqfzD9WVkZEAmk0FHRwcpKSmIjY3l4G/6KH6iERVAWd0SMpkMycnJAP5v3NHZs2dhZmaGihUrSuObBgwYAGtra+zbtw9yuRyOjo5o06YNvLy8pHVqaWnB3Nycv9Qp130s+W7ZsiU8PDzw888/AwBcXV3RunVrhIeHSzPQA0Djxo1RqVIlLFmyBM+ePYO2tjbkcrl0fmdNIzBr1iyYmZnh33//5TlMH8XEiagAeX/emfT0dAwaNAgeHh4YM2YMLl68CODdANmrV6/i/v370NbWRnp6OoB3M3sfOXIEWlpasLKywqBBgzB//nwAirdf4S91ym0HDhxAq1atcODAAYVyU1NTTJw4EZcuXcLq1auho6MDd3d3VKtWTaHVycTEBBkZGbhz546UZAkhpLFOO3bsQIUKFbBu3TqEhISgb9++PIfpo5g4ERUgWb/co6KiMHbsWNy7dw+tW7fGpk2bMGXKFNy9exc//PADGjRoIE0voK+vLy1jb2+P1NRUAO9mBM9qceKXDOWlLVu24MKFCxg8eDBCQkKkcxAA6tati8GDB2P8+PGQy+Wwt7dHhw4dcOvWLUyaNAkpKSnYt28frK2tsXLlSgwePBjAu7FO169fh5ubGwYNGoTBgwfj9OnT0k17iT6GV9URFSCpqanw9fXFkSNHUKFCBQQHB8PCwgL79+/HrFmzUKNGDSxYsACHDx9Gy5Yt0adPH7Ro0QLp6emYMGECfvvtN4wYMULdu0EFiBACw4cPR4MGDZCeno5t27ZJSVCWW7duwdnZGQMGDMDMmTMRHx+PdevWYdSoUahcuTLu3LmD4OBgha7l169fo1SpUmjdujUCAgJQsmRJNewdfZMEERUov//+uyhVqpTo2rWrVCaXy8XUqVNFrVq1xL///iuEEGLTpk2iZcuWonr16qJcuXJi+fLl6gqZCii5XC6EeHfOOjk5CSGEuHLliihRooQYMWKEOHHihBBCiDdv3oigoCChp6cnYmJipOVPnz4tNmzYIJ4/f66wzrdv3wohhPjvv//ya1foO8IWJ6ICImtAeHx8PIYNG4Zbt25h586dsLGxAQBcvXoVvr6+0NfXV7ha7sGDByhbtmy29RDlJvGJexaeOHEC48aNQ3BwMOzt7fHgwQP07t0bz58/x65du1ChQgWkpqbCxcUFVlZW0oz17+NtUii38NOP6DvxufmStLS0IISAhYUFunbtCiMjIyxfvlx63d7eHq1bt8b58+cRHBwslWclTVlXIDFpotwm/v/94rS0tKSrPN8nl8uRlpYGe3t7xMTEYOjQobh58yZ0dHQwYMAALF68GCYmJhg7diyuXLmCZ8+eZVs/b5NCuYWfgETfifcTms81JLdv3x61a9fGoUOHcO7cOam8VatW6N+/P1xcXLItw/vKUW7LSvazEpqRI0fC0dERvXv3xvbt26V6P/zwA2JiYlC/fn1UrFgRurq6OHv2LE6fPg03Nzf8/PPPmDNnDjp06IBHjx5lm/WbCRPlJnbVEX0n0tLSMGHCBPTt2xd16tT57K0ljh07hmnTpqFMmTL4559/1BAx0TtRUVE4cuQI1qxZA09PT6xYsQLJyckIDAxE06ZNkZycDC8vL0RGRmLOnDnw8PCQlpXL5dixYwccHR1RrFgxAO8mtMyam4kot7HFiegbkx83NOXvKcoPmZmZWLBgAVxcXBASEoK///4bXl5eWLp0KapUqQI/Pz8A7+ZhSktLg4ODAzw8PKRuY+Dd+dyxY0cUK1ZMOm+ZNFFeYuJE9A24c+cOqlWrhri4OGhpaUnJz61bt6Q6dnZ2+OWXX3D16lVs2bIFABS+YN6XlXj5+Phg7969sLOzU3idXRuU2zIyMrKVaWtro1y5cihXrhxevHghnYdVqlRB9+7d8eTJEwQFBQEAWrRogYMHD0rL5YTnLeUHJk5E3wDe0JS+dVmtQKtWrcLWrVtx+fJlAO8SojZt2iA6OhqnTp2S6jdu3Bht27bFkiVLkJSUBBMTExgaGuLGjRtqiZ8oCxMnIg3DG5rS9yg0NBTFixdHQEAAJkyYAFdXVyxZsgT6+vro1KkTGjRogDlz5kj1ixYtii5duuD169eYNm0aOnfujPDwcFSpUkWNe0HExIlI4/CGpvSt+zBZT09Px8yZM9GjRw/cuHEDhw8fxujRozFu3Djs3bsXVatWRadOnRAZGYmNGzdKy9WpUwe+vr5o27YtTE1NUapUqc9Ou0GU15g4EWkY3tCUvlXv30T6/fF158+fx+XLl9G9e3cA71pQfX19UbNmTSxZsgQZGRlo2bIlnJycMH/+fKSlpQEAjIyM4O3tjSZNmkjr4jxipG48A4k0DG9oSt+a9xMmAAgICECbNm3w888/4/HjxyhRogRSUlKkG0ZnndNTpkzBnj178OTJE5QuXRrNmjXDq1evcOTIEYX1s0WUNAnncSLSIII3NKVvWGxsLIKDg7FlyxZ07doVixYtQuPGjdG3b1+EhIRALpcr3M7n4sWLaNGiBbZs2QJnZ2ekpKQgOTkZ1tbWatwLok9j4kSkIbJuOzF37lxs27YNx48fx9WrV+Hu7o4uXbqgR48ecHR0xNu3bxEcHIyxY8fi7t27KFWqFADgzJkzuH//Ppo3bw4LCwtpnZmZmdDR0UFsbCyKFy+uzl2k75i/vz9OnToFAwMDzJkzB+XLl8fZs2fx66+/onLlyqhWrRqCgoIwbdo09OnTBwAwe/Zs7NixA4cOHYKhoaG0rqyvJXYhkyZi4kSkBryhKX1vLl68iJYtW6JixYo4fvy4VO7v74/w8HC0b98e6enpmDBhApycnKCvr4/jx49jwYIF8Pb2ln44EGk6jnEiyme8oSl9j2rXro0ePXogLS0NFy5ckMq9vLxgYGCAkydPYvjw4Th48CDc3d1Ru3ZtPHjwAN7e3gDYukTfDiZORPmENzSl713WLVJ27twpne8lSpRAt27dcOnSJSxYsABNmzbFb7/9hjlz5sDKygqZmZkc/E3fFCZORPkkq1suKioKf//9N86fP4+RI0fi4cOHmDJlijQXU3JyMpydnfH69Wts27YNO3fuhK2tLUxMTODn54ctW7agX79+0NPTA5DzrSyI1MHS0hL9+/dHWFgYDh06JJX37NkTzZo1Q8OGDQH83xgmuVzO1lH65nCME1E+yczMxKJFizBjxgxUq1YNCxcuhJ2dHW7cuIEpU6YgJiZGGhvSvn17FClSBCtWrEBmZmaO9+bimBDSROnp6WjRogXs7e3h7++PIkWKqDskolzFFieiPMAbmlJBpa+vj7Fjx+LQoUO4fv26wmuc9Zu+B0yciPIAb2hKBVnr1q2xdetWNG7cWKGcs37T94BddUR5IDQ0FP369YO5uTmEEHj27BlmzZqFwYMHIyoqCj4+PjA0NMTWrVulZY4dO4a+ffuiS5cumDRpEhISEqQ5moiISDMw/Sf6SryhKRFRwcEWJ6IvJJfLpa6H9wdwnzhxAm3atMHevXulq4gAwMXFBYUKFcLmzZvx33//YcqUKYiKikJYWBgMDAzUsg9ERKQatjgRqYg3NCUiKriYOBGpKCthio2NhZ+fH/73v/+hfv36WL9+PX766SdcuHABHTp0wMyZMwEARkZGAABzc3OYm5vj7t27AIAOHTogPDwcbm5uCuvn1XJERJqLiRPRF/D394e3tzciIyOxa9cuTJ06Ffv27cPLly9x4MABNGrUCOfPn8eaNWukZQ4ePIhKlSqhfv36AABjY2NYW1tL960jIiLNx8SJ6Au0adMGZ8+exZMnT1C+fHkAQL169dCsWTPcuXMHADB06FB4enqiSZMmaNGiBaZOnYqBAwfC0NBQIVGSyWRsZSIi+kYwcSL6AryhKRFRwcSr6oi+0LNnz+Du7o62bdti0qRJ0tinVatWYdasWejfvz/Gjx+vsExmZia0tLSYOBERfaPY4kT0hXhDUyKigoctTkRfgTc0JSIqWNjiRPQVeENTIqKChS1ORF9JCIGoqCjY2dmpOxQiIspjTJyIiIiIlMSuOiIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiIiIiUhITJyIiIiIlMXEiom9eeHg4ZDIZXr16pe5QiOg7x8SJiPKMp6cnZDIZZDIZdHR0ULp0aQwdOhQvX77M1e04OTkhNjYW5ubmubreD7m4uGDkyJF5ug0i0mxMnIgoT7m7uyM2NhYPHjzA0qVLsWvXLgwbNixXt6Gnp4dixYpBJpPl6nqJiD7ExImI8pS+vj6KFSuGUqVKwc3NDd27d8fBgwcV6qxYsQJVq1aFgYEBqlSpguDgYIXXT5w4gVq1asHAwAAODg7Yvn07ZDIZLl26BCDnrrotW7agWrVq0NfXR9myZTFv3jyFdZYtWxYzZ87EwIEDYWpqitKlS2PJkiUf3Q9PT09ERETgjz/+kFrR7t+/jwoVKmDu3LkKda9duwYtLS3cvXsXACCTybB48WJ4eHjA0NAQtra22LRpk8Iyjx8/Rvfu3VG4cGFYWFigffv2ePDggTKHmIjyERMnIso39+7dw/79+6GrqyuV/fPPP/D19cWMGTMQFRWFmTNnws/PD6tWrQIAJCUloW3btrC3t8eFCxfg7++PcePGfXI758+fR7du3dCjRw9cvXoVU6ZMgZ+fH1auXKlQb968eXBwcMDFixcxbNgwDB06FDdu3MhxnX/88QccHR0xaNAgxMbGIjY2FqVLl8bAgQOxYsUKhbrLly9H48aNUb58eanMz88PnTt3xuXLl9GnTx/07NkTUVFRAIDU1FS4urrCxMQER44cwbFjx2BiYgJ3d3e8efNG6eNLRPlAEBHlkf79+wttbW1hbGwsDAwMBAABQAQGBkp1bGxsxLp16xSW8/f3F46OjkIIIRYvXiwsLCzE69evpdf/+ecfAUBcvHhRCCFEWFiYACBevnwphBCiV69eokWLFgrr/PXXX4WdnZ30vEyZMqJPnz7Sc7lcLqysrMTixYs/uj/Ozs7il19+USj777//hLa2tjh9+rQQQog3b94IS0tLsXLlSqkOADFkyBCF5Ro0aCCGDh0qhBBi2bJlonLlykIul0uvp6enC0NDQ3HgwIGPxkNE+Y8tTkSUp1xdXXHp0iWcPn0aP//8M1q2bImff/4ZAPDs2TPExMTAy8sLJiYm0mP69OlSN9fNmzdRo0YNGBgYSOusX7/+J7cZFRWFRo0aKZQ1atQIt2/fRmZmplRWo0YN6f8ymQzFihVDXFycSvtXvHhxtG7dGsuXLwcA7N69G2lpaejatatCPUdHx2zPs1qczp8/jzt37sDU1FQ6BkWKFEFaWpp0HIhIM+ioOwAi+r4ZGxujQoUKAIAFCxbA1dUVU6dOhb+/P+RyOYB33XUNGjRQWE5bWxsAIITINuhbCPHJbSq7zPtdhsC75CkrJlV4e3ujb9++mD9/PlasWIHu3bvDyMjos8tlxSiXy1G3bl2sXbs2Wx1LS0uV4yGivMPEiYjy1eTJk+Hh4YGhQ4eiRIkSKFmyJO7du4fevXvnWL9KlSpYu3Yt0tPToa+vDwA4d+7cJ7dhZ2eHY8eOKZSdOHEClSpVkhKyL6Gnp6fQYpWlVatWMDY2xuLFi7Fv3z4cOXIkW51Tp06hX79+Cs9r164NAKhTpw5CQkJgZWUFMzOzL46PiPIeu+qIKF+5uLigWrVqmDlzJgBgypQpCAgIwB9//IFbt27h6tWrWLFiBQIDAwEAvXr1glwux+DBgxEVFYUDBw5IV7F9bPqB0aNH49ChQ/D398etW7ewatUqLFy4EGPGjPmq2MuWLYvTp0/jwYMHeP78udQ6pa2tDU9PT0yYMAEVKlTI1i0HAJs2bcLy5ctx69YtTJ48GWfOnMFPP/0EAOjduzeKFi2K9u3b4+jRo7h//z4iIiLwyy+/4NGjR18VMxHlLiZORJTvfHx88M8//yAmJgbe3t5YunQpVq5cCXt7ezg7O2PlypWwtbUFAJiZmWHXrl24dOkSatWqBV9fX0yaNAkAFMY9va9OnTrYuHEjNmzYgOrVq2PSpEmYNm0aPD09vyruMWPGQFtbG3Z2drC0tER0dLT0mpeXF968eYOBAwfmuOzUqVOxYcMG1KhRA6tWrcLatWthZ2cHADAyMsKRI0dQunRpdOrUCVWrVsXAgQPx+vVrtkARaRiZ+NxgASIiDbN27VoMGDAACQkJMDQ0VHc4AIDjx4/DxcUFjx49grW1tcJrMpkM27ZtQ4cOHdQTHBHlGo5xIiKNt3r1apQrVw4lS5bE5cuXMW7cOHTr1k0jkqb09HTExMTAz88P3bp1y5Y0EdH3hV11RKTxnjx5gj59+qBq1aoYNWoUunbt+slZvvPT+vXrUblyZSQkJGDOnDnqDoeI8hi76oiIiIiUxBYnIiIiIiUxcSIiIiJSEhMnIiIiIiUxcSIiIiJSEhMnIiIiIiUxcSIiIiJSEhMnIiIiIiUxcSIiIiJSEhMnIiIiIiX9P+UfOfkAN45YAAAAAElFTkSuQmCC",
            "text/plain": [
              "<Figure size 600x400 with 1 Axes>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "#Plot1 - A-G fraction bar plot\n",
        "\n",
        "AG_fraction_pivot = (\n",
        "    AG_fraction_plot\n",
        "    .pivot_table(\n",
        "        index='region_type',\n",
        "        columns='condition',\n",
        "        values='AG_fraction',\n",
        "        fill_value=0\n",
        "    )\n",
        "    .reset_index()\n",
        ")\n",
        "\n",
        "x = np.arange(len(AG_fraction_pivot['region_type']))\n",
        "width = 0.35\n",
        "\n",
        "plt.figure(figsize=(6, 4))\n",
        "\n",
        "plt.bar(\n",
        "    x - width/2,\n",
        "    AG_fraction_pivot['WT'],\n",
        "    width,\n",
        "    label='WT'\n",
        ")\n",
        "\n",
        "plt.bar(\n",
        "    x + width/2,\n",
        "    AG_fraction_pivot['KO'],\n",
        "    width,\n",
        "    label='KO'\n",
        ")\n",
        "\n",
        "plt.xticks(\n",
        "    x,\n",
        "    AG_fraction_pivot['region_type'],\n",
        "    rotation=30,\n",
        "    ha='right'\n",
        ")\n",
        "\n",
        "plt.ylabel('A-to-G editing fraction')\n",
        "plt.xlabel('Region type')\n",
        "plt.title('A-to-G / T-to-C Editing Fraction by Region and Condition')\n",
        "plt.legend()\n",
        "plt.tight_layout()\n",
        "plt.show()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "pcf5aWmgteRm",
        "outputId": "4a7c7e24-89f2-4d13-9556-23c2926d3b0a"
      },
      "outputs": [
        {
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAk4AAAGGCAYAAACNCg6xAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAVMVJREFUeJzt3XlcTfn/B/DXLZX2fSWJIhQhIktlzzIMM8Y2ZBsjW9Zh7L5GmBFf+zD2bcwYxs6EalB2yZJtFKFEUilaP78/+nW+rsK95qYrr+fj0ePhfu45575PXfW6n8/nfI5MCCFARERERO+kUdoFEBEREX0sGJyIiIiIFMTgRERERKQgBiciIiIiBTE4ERERESmIwYmIiIhIQQxORERERApicCIiIiJSEIMTERERkYIYnIiUtH79eshksjd+hYWFvfMYMpkMM2bMKHLMuLg4qW3r1q1YtGiRQvurk7CwsCLfhwMHDqhtvSXp4cOHmDFjBqKiolR2TH9/fxgYGKjseOpgxowZkMlkpV0GkULKlXYBRB+rdevWwcXFpUh7zZo1lT5Whw4dEBkZCVtbW6lt69atuHLlCgIDA4tsHxkZiYoVKyr9Oh9CvXr1EBkZKfd9OHDgAJYtW/bJhaeHDx9i5syZqFy5Mtzd3Uu7HLU1aNAgtGvXrrTLIFIIgxPRe3J1dYWHh4dKjmVpaQlLS0uFt2/UqJFKXrckGBkZqXV99P5evHgBXV1dlR+3YsWKavtBgOh1HKojKkFpaWkYPHgwzM3NYWBggHbt2uHmzZtFtnt9qM7Hxwf79+/H3bt35YYBC71pqC80NBRDhw6FhYUFzM3N0bVrVzx8+FDutbKysjB27FjY2NhAT08PzZs3x/nz51G5cmX4+/u/85xWrFiBOnXqwMDAAIaGhnBxccH3338vPf/6UJ2/vz+WLVsm1V34VXiuQggsX74c7u7u0NXVhampKb744gvcuXNHge8wcP36dfTs2RPW1tbQ0dFBpUqV0LdvX2RlZUnbXLlyBZ07d4apqSnKly8Pd3d3bNiwQe44xQ2XFnc+QMHPx9XVFWfPnkWzZs2gp6eHKlWqYO7cucjPz5f2a9CgAQCgf//+0nkX/tzu3LmDHj16wM7ODjo6OrC2tkbLli0VHta7evUqWrZsCX19fVhaWmL48OHIzMyUnm/ZsiVcXFzw+n3chRBwcnJChw4d3nr8ypUro2PHjti5cyfq1q2L8uXLY+bMmQCAxMREDBkyBBUrVoS2tjYcHR0xc+ZM5Obmyh3j/v37+OKLL2BoaAgTExP07t0bZ8+ehUwmw/r166Xtihuqy8/Px/z58+Hi4gIdHR1YWVmhb9++uH//vtx2ivwsiFSJPU5E7ykvL6/IHwqZTAZNTU0ABX+gunTpgoiICEybNg0NGjTAyZMn4efn985jL1++HN988w3++ecf7Nq1S+GaBg0ahA4dOmDr1q2Ij4/H+PHj0adPHxw7dkzapn///ti+fTsmTJiAFi1a4Nq1a/j888+Rlpb2zuP/+uuvCAgIwIgRI/DTTz9BQ0MDt2/fxrVr1964z9SpU5GRkYEdO3YgMjJSai8clhwyZAjWr1+PkSNHYt68eXj69ClmzZoFLy8vXLp0CdbW1m889qVLl9C0aVNYWFhg1qxZcHZ2RkJCAvbs2YPs7Gzo6Ojgxo0b8PLygpWVFRYvXgxzc3Ns3rwZ/v7+ePToESZMmKDIt7aIxMRE9O7dG2PHjsX06dOxa9cuTJo0CXZ2dujbty/q1auHdevWoX///pgyZYoUVAp7Vtq3b4+8vDzMnz8flSpVwpMnTxAREYFnz56987VzcnLQvn17DBkyBBMnTkRERARmz56Nu3fvYu/evQCAUaNGoXPnzjh69ChatWol7Xvw4EH8888/WLx48Ttf58KFC4iJicGUKVPg6OgIfX19JCYmomHDhtDQ0MC0adNQtWpVREZGYvbs2YiLi8O6desAABkZGfD19cXTp08xb948ODk54dChQ/jqq68U+v4OHToUq1atwvDhw9GxY0fExcVh6tSpCAsLw4ULF2BhYaHwz4JIpQQRKWXdunUCQLFfmpqa0nYHDx4UAMR///tfuf1/+OEHAUBMnz69yDFjY2Oltg4dOggHB4dia3jT/gEBAXLbzZ8/XwAQCQkJQgghrl69KgCI7777Tm67bdu2CQCiX79+bz334cOHCxMTk7duExoaKgCI0NBQqW3YsGGiuF83kZGRAoBYsGCBXHt8fLzQ1dUVEyZMeOtrtWjRQpiYmIikpKQ3btOjRw+ho6Mj7t27J9fu5+cn9PT0xLNnz4QQxf8M3nQ+3t7eAoA4ffq03LY1a9YUbdu2lR6fPXtWABDr1q2T2+7JkycCgFi0aNFbz684/fr1e+v76sSJE0IIIfLy8kSVKlVE586di5x31apVRX5+/ltfx8HBQWhqaoobN27ItQ8ZMkQYGBiIu3fvyrX/9NNPAoC4evWqEEKIZcuWCQDi4MGDRfZ//Xsyffp0ufdHTExMse/n06dPCwDi+++/l9oU/VkQqQqH6oje08aNG3H27Fm5r9OnT0vPh4aGAgB69+4tt1+vXr1KrKbPPvtM7nHt2rUBAHfv3gUAhIeHAwC6d+8ut90XX3yBcuXe3QHdsGFDPHv2DD179sTu3bvx5MmTf1Xvvn37IJPJ0KdPH+Tm5kpfNjY2qFOnzluvUMzMzER4eDi6d+/+1vlhx44dQ8uWLWFvby/X7u/vj8zMTLleMGXY2NigYcOGcm21a9eWvtdvY2ZmhqpVq+LHH39EcHAwLl68qPSw0pveV4XvOw0NDQwfPhz79u3DvXv3AAD//PMPDh06hICAAIWuYqtduzaqVasm17Zv3z74+vrCzs5O7mdW2JNa+B4LDw+HoaFhkUnfPXv2fOfrFp7D60PHDRs2RI0aNXD06FG59n/zsyBSFoMT0XuqUaMGPDw85L7q168vPZ+cnIxy5crB3Nxcbj8bG5sSq+n119LR0QFQMKm3sCYARYa/iquzOF9//TXWrl2Lu3fvolu3brCysoKnpydCQkLeq95Hjx5BCAFra2toaWnJfZ06deqtwSwlJQV5eXnvnFScnJwsd7ViITs7O+n591Hc90tHR0f6Xr+NTCbD0aNH0bZtW8yfPx/16tWDpaUlRo4cifT09Hfu/7b31avnM2DAAOjq6mLlypUAgGXLlkFXVxcDBgx452sAKPb79ujRI+zdu7fIz6tWrVoAIP3MkpOTix1mfdvQa6HCc3jTz+31n9m/+VkQKYtznIhKiLm5OXJzc5GcnCz3iz0xMbFUawIK/vhVqFBBai+sUxH9+/dH//79kZGRgb///hvTp09Hx44dcfPmTTg4OChVj4WFBWQyGY4fPy6FvFcV11bIzMwMmpqaRSYLv87c3BwJCQlF2gsnzRfOlSlfvjwAyE0qB/Cve9XexMHBAWvWrAEA3Lx5E7/99htmzJiB7OxsKei8ydveV6+2GRsbo1+/fvjll18wbtw4rFu3Dr169YKJiYlCNRbXK2VhYYHatWvjhx9+KHafwkBqbm6OM2fOFHlekfd/4TkkJCQUCcYPHz6Um99E9KGxx4mohPj6+gIAtmzZIte+detWhfYviU/MzZs3BwBs375drn3Hjh1FJrq/i76+Pvz8/DB58mRkZ2fj6tWrb9z29Z6vQh07doQQAg8ePCjSe+fh4QE3N7c3HlNXVxfe3t74/fff3xpuWrZsiWPHjhW5unDjxo3Q09OTlk6oXLkyACA6Olpuuz179rzx2O/ypvN+XbVq1TBlyhS4ubnhwoULCh37Te8rHx8fufaRI0fiyZMn+OKLL/Ds2TMMHz5cweqL17FjR1y5cgVVq1Yt9mdWGJy8vb2Rnp6OgwcPyu3/66+/vvM1WrRoAQDYvHmzXPvZs2cRExODli1b/qtzIPo32ONE9J6uXLlSbNioWrUqLC0t0aZNGzRv3hwTJkxARkYGPDw8cPLkSWzatEmh47u5uWHnzp1YsWIF6tevDw0NjX+9blStWrXQs2dPLFiwAJqammjRogWuXr2KBQsWwNjYGBoab/8sNXjwYOjq6qJJkyawtbVFYmIigoKCYGxsLF16/6ZzAYB58+bBz88PmpqaqF27Npo0aYJvvvkG/fv3x7lz59C8eXPo6+sjISEBJ06cgJubG4YOHfrG4wYHB6Np06bw9PTExIkT4eTkhEePHmHPnj34+eefYWhoiOnTp0vzcqZNmwYzMzNs2bIF+/fvx/z582FsbAwAaNCgAapXr45x48YhNzcXpqam2LVrF06cOPEe3+kCVatWha6uLrZs2YIaNWrAwMAAdnZ2ePLkCYYPH44vv/wSzs7O0NbWxrFjxxAdHY2JEye+87ja2tpYsGABnj9/jgYNGkhX1fn5+aFp06Zy21arVg3t2rXDwYMH0bRpU9SpU+e9zwcAZs2ahZCQEHh5eWHkyJGoXr06Xr58ibi4OBw4cAArV65ExYoV0a9fPyxcuBB9+vTB7Nmz4eTkhIMHD+Lw4cMA8Nb3WvXq1fHNN99gyZIl0NDQgJ+fn3RVnb29PUaPHv2vzoHoXynt2elEH5u3XVUHQKxevVra9tmzZ2LAgAHCxMRE6OnpidatW4vr168rdFXd06dPxRdffCFMTEyETCaTu+roTfufPXtWrtbirgh7+fKlGDNmjLCyshLly5cXjRo1EpGRkcLY2FiMHj36ree+YcMG4evrK6ytrYW2traws7MT3bt3F9HR0W99zaysLDFo0CBhaWkpncur57p27Vrh6ekp9PX1ha6urqhataro27evOHfu3FvrEUKIa9euiS+//FKYm5sLbW1tUalSJeHv7y9evnwpbXP58mXRqVMnYWxsLLS1tUWdOnWKXOkmhBA3b94Ubdq0EUZGRsLS0lKMGDFC7N+/v9ir6mrVqlVk/379+hW5EnLbtm3CxcVFaGlpST+3R48eCX9/f+Hi4iL09fWFgYGBqF27tli4cKHIzc196/n269dP6Ovri+joaOHj4yN0dXWFmZmZGDp0qHj+/Hmx+6xfv14AEL/++utbj/0qBwcH0aFDh2Kfe/z4sRg5cqRwdHQUWlpawszMTNSvX19MnjxZroZ79+6Jrl27CgMDA2FoaCi6desmDhw4IACI3bt3S9u9flWdEAVXBc6bN09Uq1ZNaGlpCQsLC9GnTx8RHx8vt50yPwsiVZAJ8drqaET0yYmIiECTJk2wZcuWEr3qj0pHt27dcOrUKcTFxUFLS6tUa5kzZw6mTJmCe/fucbVw+ihxqI7oExMSEoLIyEjUr18furq6uHTpEubOnQtnZ2d07dq1tMsjFcnKysKFCxdw5swZ7Nq1C8HBwR88NC1duhQA4OLigpycHBw7dgyLFy9Gnz59GJroo8XgRPSJMTIywl9//YVFixYhPT0dFhYW8PPzQ1BQkHRlGX38EhIS4OXlBSMjIwwZMgQjRoz44DXo6elh4cKFiIuLQ1ZWFipVqoTvvvsOU6ZM+eC1EKkKh+qIiIiIFMTlCIiIiIgUxOBEREREpCAGJyIiIiIFlerk8KCgIOzcuRPXr1+Hrq4uvLy8MG/ePFSvXl3axt/fHxs2bJDbz9PTE6dOnZIeZ2VlYdy4cdi2bRtevHiBli1bYvny5QpftZGfn4+HDx/C0NBQoRtfEhERUdkhhEB6ejrs7OzeuRBwqU4Ob9euHXr06IEGDRogNzcXkydPxuXLl3Ht2jXo6+sDKAhOjx49wrp166T9tLW1YWZmJj0eOnQo9u7di/Xr18Pc3Bxjx47F06dPcf78eWhqar6zjvv37xe5czoRERF9WuLj49/Z6aJWV9U9fvwYVlZWCA8Pl+6p5e/vj2fPnuHPP/8sdp/U1FRYWlpi06ZN+OqrrwAU3ATS3t4eBw4cQNu2bd/5uqmpqTAxMUF8fDyMjIxUdj5ERESk/tLS0mBvb49nz55Jt2F6E7Vaxyk1NRUA5HqTACAsLAxWVlYwMTGBt7c3fvjhB1hZWQEAzp8/j5ycHLRp00ba3s7ODq6uroiIiFAoOBUOzxkZGTE4ERERfaIUma6jNsFJCIExY8agadOmcHV1ldr9/Pzw5ZdfwsHBAbGxsZg6dSpatGiB8+fPQ0dHB4mJidDW1oapqanc8aytrZGYmFjsa2VlZSErK0t6nJaWVjInRURERGWK2gSn4cOHIzo6usidyAuH3wDA1dUVHh4ecHBwwP79+996ewghxBuTY1BQEGbOnKmawomIiOiToRbLEYwYMQJ79uxBaGjoOydl2drawsHBAbdu3QIA2NjYIDs7GykpKXLbJSUlwdrauthjTJo0CampqdJXfHy8ak6EiIiIyrRS7XESQmDEiBHYtWsXwsLC4Ojo+M59kpOTER8fD1tbWwBA/fr1oaWlhZCQEHTv3h1AwT2arly5gvnz5xd7DB0dHejo6KjuRIiIiNREfn4+srOzS7sMtaKlpaXQVfaKKNXgNGzYMGzduhW7d++GoaGhNCfJ2NgYurq6eP78OWbMmIFu3brB1tYWcXFx+P7772FhYYHPP/9c2nbgwIEYO3YszM3NYWZmhnHjxsHNzQ2tWrUqzdMjIiL6oLKzsxEbG4v8/PzSLkXtmJiYwMbG5l+v11iqwWnFihUAAB8fH7n2devWwd/fH5qamrh8+TI2btyIZ8+ewdbWFr6+vti+fTsMDQ2l7RcuXIhy5cqhe/fu0gKY69evV1m6JCIiUndCCCQkJEBTUxP29vbvXMjxUyGEQGZmJpKSkgBAGrF6X2q1jlNpSUtLg7GxMVJTU7kcARERfZRycnJw+/Zt2NnZvXMtok9RcnIykpKSUK1atSIdK8rkAMZRIiKiMiAvLw9Awd01qCg9PT0ABQHz32BwIiIiKkN4z9Xiqer7wuBEREREpCAGJyIiIiIFqc3K4WVZ5Yn7S7uEDy5ubofSLoGIiPDh/wYp+/t/5cqVGD9+PFJSUlCuXEEsef78OUxNTdGoUSMcP35c2vb48eNo3rz5O49ZeHV+SWBwIiIiolLj6+uL58+f49y5c2jUqBGAgoBkY2ODs2fPIjMzU5rYHRYWBhsbG1y8eFHaf9SoUUhLS8O6deuktpK8qpBDdURERFRqqlevDjs7O4SFhUltYWFh6Ny5M6pWrYqIiAi59pYtW8LGxkb60tXVhY6OTpG2ksLgRERERKXKx8cHoaGh0uPQ0FD4+PjA29tbas/OzkZkZCR8fX1Lq0wADE5ERERUynx8fHDy5Enk5uYiPT0dFy9eRPPmzeHt7S31RJ06dQovXrwo9eDEOU5ERERUqnx9fZGRkYGzZ88iJSUF1apVg5WVFby9vfH1118jIyMDYWFhqFSpEqpUqVKqtTI4ERERUalycnJCxYoVERoaipSUFHh7ewMAbGxs4OjoiJMnTyI0NBQtWrQo5Uo5VEdERERqwNfXF2FhYQgLC4OPj4/U7u3tjcOHD+PUqVOlPkwHMDgRERGRGvD19cWJEycQFRUl9TgBBcFp9erVePnyJYMTEREREVAQnF68eAEnJydYW1tL7d7e3khPT0fVqlVhb29fihUW4BwnIiKiMuxjuZND5cqVIYQo0l6xYsVi2wutX7++BKsqij1ORERERApicCIiIiJSEIMTERERkYIYnIiIiIgUxOBEREREpCAGJyIiIiIFMTgRERERKYjBiYiIiEhBDE5ERERECmJwIiIiIlIQb7lCRERUls0w/sCvl6rU5v7+/nj27Bn+/PNPqW3Hjh3o06cPZs2ahbFjx2Lx4sVYt24dbt68ifLly6Nx48aYMmUKmjRpouLi3409TkRERKQ2fvnlF/Tu3RtLly7F+PHj0aNHD8yaNQsjR45ETEwMwsPDYW9vDx8fH7mw9aGwx4mIiIjUwvz58zFt2jRs3boV3bp1w/bt27Fjxw7s2bMHnTp1krZbtWoVkpOTMWjQILRu3Rr6+vofrEb2OBEREVGpmzhxIv7zn/9g37596NatGwBg69atqFatmlxoKjR27FgkJycjJCTkg9bJHiciIiIqVQcPHsTu3btx9OhRtGjRQmq/efMmatSoUew+he03b978IDUWYo8TERERlaratWujcuXKmDZtGtLT05XaVyaTlVBVxWNwIiIiolJVoUIFhIeHIyEhAe3atZPCU7Vq1XDt2rVi94mJiQEAODs7f7A6AQYnIiIiUgOVKlVCeHg4kpKS0KZNG6SlpaFHjx64desW9u7dW2T7BQsWwNzcHK1bt/6gdTI4ERERkVqoWLEiwsLCkJycjDZt2qB9+/b4/PPP0a9fP6xZswZxcXGIjo7GkCFDsGfPHvzyyy8f9Io6gMGJiIiI1EjhsN2zZ8/QunVrrFq1CpMnT8bChQvh4uKCZs2a4e7duwgNDUWXLl0+eH0yIYT44K+qZtLS0mBsbIzU1FQYGRmp/PiVJ+5X+THVXdzcDqVdAhHRJ+Xly5eIjY2Fo6MjypcvX9rlqJ23fX+UyQHscSIiIiJSEIMTERERkYIYnIiIiIgUxOBEREREpCAGJyIiIiIFMTgRERGVIbxYvnj5+fkqOQ5v8ktERFQGaGlpQSaT4fHjx7C0tPzg93BTV0IIZGdn4/Hjx9DQ0IC2tva/Oh6DExERURmgqamJihUr4v79+4iLiyvtctSOnp4eKlWqBA2NfzfYxuBERERURhgYGMDZ2Rk5OTmlXYpa0dTURLly5VTSC1eqwSkoKAg7d+7E9evXoaurCy8vL8ybNw/Vq1eXthFCYObMmVi1ahVSUlLg6emJZcuWoVatWtI2WVlZGDduHLZt24YXL16gZcuWWL58OSpWrFgap0VERFRqNDU1oampWdpllFmlOjk8PDwcw4YNw6lTpxASEoLc3Fy0adMGGRkZ0jbz589HcHAwli5dirNnz8LGxgatW7dGenq6tE1gYCB27dqFX3/9FSdOnMDz58/RsWNH5OXllcZpERERURmlVveqe/z4MaysrBAeHo7mzZtDCAE7OzsEBgbiu+++A1DQu2RtbY158+ZhyJAhSE1NhaWlJTZt2oSvvvoKAPDw4UPY29vjwIEDaNu27Ttfl/eqUz3eq46IiD4WH+296lJTUwEAZmZmAIDY2FgkJiaiTZs20jY6Ojrw9vZGREQEAOD8+fPIycmR28bOzg6urq7SNkRERESqoDaTw4UQGDNmDJo2bQpXV1cAQGJiIgDA2tpabltra2vcvXtX2kZbWxumpqZFtinc/3VZWVnIysqSHqelpansPIiIiKjsUpsep+HDhyM6Ohrbtm0r8tzrs+CFEO+cGf+2bYKCgmBsbCx92dvbv3/hRERE9MlQSXB69uzZv9p/xIgR2LNnD0JDQ+WuhLOxsQGAIj1HSUlJUi+UjY0NsrOzkZKS8sZtXjdp0iSkpqZKX/Hx8f+qfiIiIvo0KB2c5s2bh+3bt0uPu3fvDnNzc1SoUAGXLl1S6lhCCAwfPhw7d+7EsWPH4OjoKPe8o6MjbGxsEBISIrVlZ2cjPDwcXl5eAID69etDS0tLbpuEhARcuXJF2uZ1Ojo6MDIykvsiIiIiehelg9PPP/8sDW2FhIQgJCQEBw8ehJ+fH8aPH6/UsYYNG4bNmzdj69atMDQ0RGJiIhITE/HixQsABUN0gYGBmDNnDnbt2oUrV67A398fenp66NWrFwDA2NgYAwcOxNixY3H06FFcvHgRffr0gZubG1q1aqXs6RERERG9kdKTwxMSEqTgtG/fPnTv3h1t2rRB5cqV4enpqdSxVqxYAQDw8fGRa1+3bh38/f0BABMmTMCLFy8QEBAgLYD5119/wdDQUNp+4cKFKFeuHLp37y4tgLl+/XouAEZEREQqpXRwMjU1RXx8POzt7XHo0CHMnj0bQMGwm7ILTiqyhJRMJsOMGTMwY8aMN25Tvnx5LFmyBEuWLFHq9YmIiIiUoXRw6tq1K3r16gVnZ2ckJyfDz88PABAVFQUnJyeVF0hERESkLpQOTgsXLkTlypURHx+P+fPnw8DAAEDBEF5AQIDKCyQiIiJSF0oHJy0tLYwbN65Ie2BgoCrqISIiIlJb77WO06ZNm9C0aVPY2dlJK3gvWrQIu3fvVmlxREREROpE6eC0YsUKjBkzBn5+fnj27Jk0IdzExASLFi1SdX1EREREakPp4LRkyRKsXr0akydPlrvc38PDA5cvX1ZpcURERETqROngFBsbi7p16xZp19HRQUZGhkqKIiIiIlJHSgcnR0dHREVFFWk/ePAgatasqYqaiIiIiNSS0lfVjR8/HsOGDcPLly8hhMCZM2ewbds2BAUF4ZdffimJGomIiIjUgtLBqX///sjNzcWECROQmZmJXr16oUKFCvjvf/+LHj16lESNRERERGpB6eAEAIMHD8bgwYPx5MkT5Ofnw8rKStV1EREREakdpec4tWjRAs+ePQMAWFhYSKEpLS0NLVq0UGlxREREROpE6eAUFhaG7OzsIu0vX77E8ePHVVIUERERkTpSeKguOjpa+ve1a9eQmJgoPc7Ly8OhQ4dQoUIF1VZHREREpEYUDk7u7u6QyWSQyWTFDsnp6upiyZIlKi2OiIiISJ0oHJxiY2MhhECVKlVw5swZWFpaSs9pa2vDyspKbiVxIiIiorJG4eDk4OAAAMjPzy+xYoiIiIjUmULBac+ePfDz84OWlhb27Nnz1m0/++wzlRRGREREpG4UCk5dunRBYmIirKys0KVLlzduJ5PJkJeXp6raiIiIiNSKQsHp1eE5DtURERHRp0rpdZyKU7ggJhEREVFZpnRwmjdvHrZv3y49/vLLL2FmZoYKFSrg0qVLKi2OiIiISJ0oHZx+/vln2NvbAwBCQkJw5MgRHDp0CH5+fhg/frzKCyQiIiJSF0rf5DchIUEKTvv27UP37t3Rpk0bVK5cGZ6eniovkIiIiEhdKN3jZGpqivj4eADAoUOH0KpVKwCAEIJX1BEREVGZpnSPU9euXdGrVy84OzsjOTkZfn5+AICoqCg4OTmpvEAiIiIidaF0cFq4cCEqV66M+Ph4zJ8/HwYGBgAKhvACAgJUXiARERGRulA6OGlpaWHcuHFF2gMDA1VRDxEREZHaUsk6TkRERESfAgYnIiIiIgUxOBEREREpiMGJiIiISEEMTkREREQKUvqqOlNTU8hksiLtMpkM5cuXh5OTE/z9/dG/f3+VFEhERESkLpQOTtOmTcMPP/wAPz8/NGzYEEIInD17FocOHcKwYcMQGxuLoUOHIjc3F4MHDy6JmomIiIhKhdLB6cSJE5g9eza+/fZbufaff/4Zf/31F/744w/Url0bixcvZnAiIiKiMkXpOU6HDx+W7k/3qpYtW+Lw4cMAgPbt2+POnTv/vjoiIiIiNaJ0cDIzM8PevXuLtO/duxdmZmYAgIyMDBgaGv776oiIiIjUiNJDdVOnTsXQoUMRGhqKhg0bQiaT4cyZMzhw4ABWrlwJAAgJCYG3t7fKiyUiIiIqTUoHp8GDB6NmzZpYunQpdu7cCSEEXFxcEB4eDi8vLwDA2LFjVV4oERERUWlTOjgBQJMmTdCkSRNV10JERESk1t4rOOXn5+P27dtISkpCfn6+3HPNmzdXSWFERERE6kbp4HTq1Cn06tULd+/ehRBC7jmZTIa8vDyVFUdERESkTpQOTt9++y08PDywf/9+2NraFruKOBEREVFZpHRwunXrFnbs2AEnJ6eSqIeIiIhIbSm9jpOnpydu376tkhf/+++/0alTJ9jZ2UEmk+HPP/+Ue97f3x8ymUzuq1GjRnLbZGVlYcSIEbCwsIC+vj4+++wz3L9/XyX1EREREb1K6R6nESNGYOzYsUhMTISbmxu0tLTknq9du7bCx8rIyECdOnXQv39/dOvWrdht2rVrh3Xr1kmPtbW15Z4PDAzE3r178euvv8Lc3Bxjx45Fx44dcf78eWhqaipxZkRERERvp3RwKgw4AwYMkNpkMhmEEEpPDvfz84Ofn99bt9HR0YGNjU2xz6WmpmLNmjXYtGmTdBuYzZs3w97eHkeOHEHbtm0VroWIiIjoXZQOTrGxsSVRxxuFhYXBysoKJiYm8Pb2xg8//AArKysAwPnz55GTk4M2bdpI29vZ2cHV1RUREREMTkRERKRSSgcnBweHkqijWH5+fvjyyy/h4OCA2NhYTJ06FS1atMD58+eho6ODxMREaGtrw9TUVG4/a2trJCYmvvG4WVlZyMrKkh6npaWV2DkQERFR2aFQcNqzZw/8/PygpaWFPXv2vHXbzz77TCWFAcBXX30l/dvV1RUeHh5wcHDA/v370bVr1zfuVzhs+CZBQUGYOXOmyuokIiKiT4NCwalLly5ITEyElZUVunTp8sbtSnoBTFtbWzg4OODWrVsAABsbG2RnZyMlJUWu1ykpKUm6b15xJk2ahDFjxkiP09LSYG9vX2J1ExERUdmg0HIE+fn50ryi/Pz8N36V9KrhycnJiI+Ph62tLQCgfv360NLSQkhIiLRNQkICrly58tbgpKOjAyMjI7kvIiIiondReh2njRs3ys0PKpSdnY2NGzcqdaznz58jKioKUVFRAAomnkdFReHevXt4/vw5xo0bh8jISMTFxSEsLAydOnWChYUFPv/8cwCAsbExBg4ciLFjx+Lo0aO4ePEi+vTpAzc3N+kqOyIiIiJVUTo49e/fH6mpqUXa09PT0b9/f6WOde7cOdStWxd169YFAIwZMwZ169bFtGnToKmpicuXL6Nz586oVq0a+vXrh2rVqiEyMhKGhobSMRYuXIguXbqge/fuaNKkCfT09LB3716u4UREREQqp/RVdW+aeH3//n0YGxsrdSwfH58iNwp+1eHDh995jPLly2PJkiVYsmSJUq9NREREpCyFg1PdunWl2560bNkS5cr9b9e8vDzExsaiXbt2JVIkERERkTpQODgVXk0XFRWFtm3bwsDAQHpOW1sblStXfuNtU4iIiIjKAoWD0/Tp0wEAlStXxldffYXy5cuXWFFERERE6kjpOU79+vUriTqIiIiI1J5CwcnMzAw3b96EhYUFTE1N37oq99OnT1VWHBEREZE6USg4LVy4UFoCYOHChW8NTkRERERllULB6dXhOX9//5KqhYiIiEitKRSc0tLSFD4gb19CREREZZVCwcnExETh4bmSvl8dERERUWlRKDiFhoZK/46Li8PEiRPh7++Pxo0bAwAiIyOxYcMGBAUFlUyVRERERGpAoeDk7e0t/XvWrFkIDg5Gz549pbbPPvsMbm5uWLVqFZcrICIiojJL6Zv8RkZGwsPDo0i7h4cHzpw5o5KiiIiIiNSR0sHJ3t4eK1euLNL+888/w97eXiVFEREREakjpVcOX7hwIbp164bDhw+jUaNGAIBTp07hn3/+wR9//KHyAomIiIjUhdI9Tu3bt8fNmzfx2Wef4enTp0hOTkbnzp1x8+ZNtG/fviRqJCIiIlILSvc4AQXDdXPmzFF1LURERERqTekeJwA4fvw4+vTpAy8vLzx48AAAsGnTJpw4cUKlxRERERGpE6WD0x9//IG2bdtCV1cXFy5cQFZWFgAgPT2dvVBERERUpikdnGbPno2VK1di9erV0NLSktq9vLxw4cIFlRZHREREpE6UDk43btxA8+bNi7QbGRnh2bNnqqiJiIiISC0pHZxsbW1x+/btIu0nTpxAlSpVVFIUERERkTpSOjgNGTIEo0aNwunTpyGTyfDw4UNs2bIF48aNQ0BAQEnUSERERKQWlF6OYMKECUhNTYWvry9evnyJ5s2bQ0dHB+PGjcPw4cNLokYiIiIitfBe6zj98MMPmDx5Mq5du4b8/HzUrFkTBgYGqq6NiIiISK28V3ACAD09vWJv9ktERERUVr3XAphEREREnyIGJyIiIiIFMTgRERERKUih4FSvXj2kpKQAAGbNmoXMzMwSLYqIiIhIHSkUnGJiYpCRkQEAmDlzJp4/f16iRRERERGpI4WuqnN3d0f//v3RtGlTCCHw008/vXH5gWnTpqm0QCIiIiJ1oVBwWr9+PaZPn459+/ZBJpPh4MGDKFeu6K4ymYzBiYiIiMoshYJT9erV8euvvwIANDQ0cPToUVhZWZVoYURERETqRukFMPPz80uiDiIiIiK1914rh//zzz9YtGgRYmJiIJPJUKNGDYwaNQpVq1ZVdX1EREREakPpdZwOHz6MmjVr4syZM6hduzZcXV1x+vRp1KpVCyEhISVRIxEREZFaULrHaeLEiRg9ejTmzp1bpP27775D69atVVYcERERkTpRuscpJiYGAwcOLNI+YMAAXLt2TSVFEREREakjpYOTpaUloqKiirRHRUXxSjsiIiIq05Qeqhs8eDC++eYb3LlzB15eXpDJZDhx4gTmzZuHsWPHlkSNRERERGpB6eA0depUGBoaYsGCBZg0aRIAwM7ODjNmzMDIkSNVXiARERGRulA6OMlkMowePRqjR49Geno6AMDQ0FDlhRERERGpm/dax6kQAxMRERF9SpSeHE5ERET0qWJwIiIiIlJQqQanv//+G506dYKdnR1kMhn+/PNPueeFEJgxYwbs7Oygq6sLHx8fXL16VW6brKwsjBgxAhYWFtDX18dnn32G+/fvf8CzICIiok+FUsEpJycHvr6+uHnzpkpePCMjA3Xq1MHSpUuLfX7+/PkIDg7G0qVLcfbsWdjY2KB169bSpHQACAwMxK5du/Drr7/ixIkTeP78OTp27Ii8vDyV1EhERERUSKnJ4VpaWrhy5QpkMplKXtzPzw9+fn7FPieEwKJFizB58mR07doVALBhwwZYW1tj69atGDJkCFJTU7FmzRps2rQJrVq1AgBs3rwZ9vb2OHLkCNq2bauSOomIiIiA9xiq69u3L9asWVMStciJjY1FYmIi2rRpI7Xp6OjA29sbERERAIDz588jJydHbhs7Ozu4urpK2xARERGpitLLEWRnZ+OXX35BSEgIPDw8oK+vL/d8cHCwSgpLTEwEAFhbW8u1W1tb4+7du9I22traMDU1LbJN4f7FycrKQlZWlvQ4LS1NJTUTERFR2aZ0cLpy5Qrq1asHAEXmOqlqCO9txxRCvPN13rVNUFAQZs6cqZL6iIiI6NOhdHAKDQ0tiTqKsLGxAVDQq2Rrayu1JyUlSb1QNjY2yM7ORkpKilyvU1JSEry8vN547EmTJmHMmDHS47S0NNjb26v6FIiIiKiMee/lCG7fvo3Dhw/jxYsXAAp6eVTJ0dERNjY2CAkJkdqys7MRHh4uhaL69etDS0tLbpuEhARcuXLlrcFJR0cHRkZGcl9ERERE76J0j1NycjK6d++O0NBQyGQy3Lp1C1WqVMGgQYNgYmKCBQsWKHys58+f4/bt29Lj2NhYREVFwczMDJUqVUJgYCDmzJkDZ2dnODs7Y86cOdDT00OvXr0AAMbGxhg4cCDGjh0Lc3NzmJmZYdy4cXBzc5OusiMiIiJSFaV7nEaPHg0tLS3cu3cPenp6UvtXX32FQ4cOKXWsc+fOoW7duqhbty4AYMyYMahbty6mTZsGAJgwYQICAwMREBAADw8PPHjwAH/99ZfcPfIWLlyILl26oHv37mjSpAn09PSwd+9eaGpqKntqRERERG8lE0qOsdnY2ODw4cOoU6cODA0NcenSJVSpUgWxsbFwc3PD8+fPS6rWEpOWlgZjY2OkpqaWyLBd5Yn7VX5MdRc3t0Npl0BERKQQZXKA0j1OGRkZcj1NhZ48eQIdHR1lD0dERET00VA6ODVv3hwbN26UHstkMuTn5+PHH3+Er6+vSosjIiIiUidKTw7/8ccf4ePjg3PnziE7OxsTJkzA1atX8fTpU5w8ebIkaiQiIiJSC0r3ONWsWRPR0dFo2LAhWrdujYyMDHTt2hUXL15E1apVS6JGIiIiIrWgdI8TUDBBnCtvExER0afmvYJTSkoK1qxZg5iYGMhkMtSoUQP9+/eHmZmZqusjIiIiUhtKD9WFh4fD0dERixcvRkpKCp4+fYrFixfD0dER4eHhJVEjERERkVpQusdp2LBh6N69O1asWCEtMpmXl4eAgAAMGzYMV65cUXmRREREROpA6R6nf/75B2PHjpVbmVtTUxNjxozBP//8o9LiiIiIiNSJ0sGpXr16iImJKdIeExMDd3d3VdREREREpJYUGqqLjo6W/j1y5EiMGjUKt2/fRqNGjQAAp06dwrJlyzB37tySqZKIiIhIDSh0rzoNDQ3IZDK8a1OZTIa8vDyVFfeh8F51qsd71RER0cdCmRygUI9TbGysSgojIiIi+pgpFJwcHBxKug4iIiIitfdeC2A+ePAAJ0+eRFJSEvLz8+WeGzlypEoKIyIiIlI3SgendevW4dtvv4W2tjbMzc0hk8mk52QyGYMTERERlVlKB6dp06Zh2rRpmDRpEjQ0lF7NgIiIiOijpXTyyczMRI8ePRiaiIiI6JOjdPoZOHAgfv/995KohYiIiEitKT1UFxQUhI4dO+LQoUNwc3ODlpaW3PPBwcEqK46IiIhInSgdnObMmYPDhw+jevXqAFBkcjgRERFRWaV0cAoODsbatWvh7+9fAuUQERERqS+l5zjp6OigSZMmJVELERERkVpTOjiNGjUKS5YsKYlaiIiIiNSa0kN1Z86cwbFjx7Bv3z7UqlWryOTwnTt3qqw4IiIiInWidHAyMTFB165dS6IWIiIiIrX2XrdcISIiIvoUcflvIiIiIgUp3ePk6Oj41vWa7ty5868KIiIiIlJXSgenwMBAucc5OTm4ePEiDh06hPHjx6uqLiIiIiK1o3RwGjVqVLHty5Ytw7lz5/51QURERETqSmVznPz8/PDHH3+o6nBEREREakdlwWnHjh0wMzNT1eGIiIiI1I7SQ3V169aVmxwuhEBiYiIeP36M5cuXq7Q4IiIiInWidHDq0qWL3GMNDQ1YWlrCx8cHLi4uqqqLiIiISO0oHZymT59eEnUQERERqT0ugElERESkIIV7nDQ0NN668CUAyGQy5Obm/uuiqAyYYVzaFXxYM1JLuwIiIvoAFA5Ou3bteuNzERERWLJkCYQQKimKiIiISB0pHJw6d+5cpO369euYNGkS9u7di969e+M///mPSosjIiIiUifvNcfp4cOHGDx4MGrXro3c3FxERUVhw4YNqFSpkqrrIyIiIlIbSgWn1NRUfPfdd3BycsLVq1dx9OhR7N27F66uriVVHxEREZHaUHiobv78+Zg3bx5sbGywbdu2YofuiIiIiMoyhYPTxIkToaurCycnJ2zYsAEbNmwodrudO3eqrDgiIiIidaJwcOrbt+87lyMgIiIiKssUDk7r168vwTKKN2PGDMycOVOuzdraGomJiQAK7pM3c+ZMrFq1CikpKfD09MSyZctQq1atD14rERERlX1qv3J4rVq1kJCQIH1dvnxZem7+/PkIDg7G0qVLcfbsWdjY2KB169ZIT08vxYqJiIiorFL74FSuXDnY2NhIX5aWlgAKepsWLVqEyZMno2vXrnB1dcWGDRuQmZmJrVu3lnLVREREVBapfXC6desW7Ozs4OjoiB49euDOnTsAgNjYWCQmJqJNmzbStjo6OvD29kZERERplUtERERlmMJznEqDp6cnNm7ciGrVquHRo0eYPXs2vLy8cPXqVWmek7W1tdw+1tbWuHv37luPm5WVhaysLOlxWlqa6osnIiKiMketg5Ofn5/0bzc3NzRu3BhVq1bFhg0b0KhRIwAocqWfEOKdV/8FBQUVmXRORERE9C5qP1T3Kn19fbi5ueHWrVuwsbEBAKnnqVBSUlKRXqjXTZo0CampqdJXfHx8idVMREREZcdHFZyysrIQExMDW1tbODo6wsbGBiEhIdLz2dnZCA8Ph5eX11uPo6OjAyMjI7kvIiIiondR66G6cePGoVOnTqhUqRKSkpIwe/ZspKWloV+/fpDJZAgMDMScOXPg7OwMZ2dnzJkzB3p6eujVq1dpl05ERERlkFoHp/v376Nnz5548uQJLC0t0ahRI5w6dQoODg4AgAkTJuDFixcICAiQFsD866+/YGhoWMqVExERUVkkE0KI0i6itKWlpcHY2BipqaklMmxXeeJ+lR9T3cWV/8R6/WaklnYFRET0npTJAR/VHCciIiKi0sTgRERERKQgBiciIiIiBTE4ERERESmIwYmIiIhIQQxORERERApicCIiIiJSEIMTERERkYIYnIiIiIgUxOBEREREpCAGJyIiIiIFMTgRERERKYjBiYiIiEhBDE5ERERECmJwIiIiIlIQgxMRERGRghiciIiIiBTE4ERERESkIAYnIiIiIgUxOBEREREpiMGJiIiISEEMTkREREQKYnAiIiIiUhCDExEREZGCGJyIiIiIFMTgRERERKQgBiciIiIiBTE4ERERESmIwYmIiIhIQQxORERERApicCIiIiJSEIMTERERkYIYnIiIiIgUxOBEREREpCAGJyIiIiIFMTgRERERKYjBiYiIiEhBDE5ERERECmJwIiIiIlIQgxMRERGRghiciIiIiBTE4ERERESkIAYnIiIiIgUxOBEREREpiMGJiIiISEHlSrsAVVm+fDl+/PFHJCQkoFatWli0aBGaNWtW2mUREdEHUHni/tIu4YOLm9uhtEv4JJWJHqft27cjMDAQkydPxsWLF9GsWTP4+fnh3r17pV0aERERlSFlIjgFBwdj4MCBGDRoEGrUqIFFixbB3t4eK1asKO3SiIiIqAz56INTdnY2zp8/jzZt2si1t2nTBhEREaVUFREREZVFH/0cpydPniAvLw/W1tZy7dbW1khMTCx2n6ysLGRlZUmPU1NTAQBpaWklUmN+VmaJHFedpclEaZfwYZXQe4eIFPNJ/p7l7x2VKfxeCvHuv10ffXAqJJPJ5B4LIYq0FQoKCsLMmTOLtNvb25dIbZ8i49Iu4EOb+8mdMRGVMuNFpV1B2ZOeng5j47f/Pv/og5OFhQU0NTWL9C4lJSUV6YUqNGnSJIwZM0Z6nJ+fj6dPn8Lc3PyNYYsUl5aWBnt7e8THx8PIyKi0yyF6J75n6WPD96xqCSGQnp4OOzu7d2770QcnbW1t1K9fHyEhIfj888+l9pCQEHTu3LnYfXR0dKCjoyPXZmJiUpJlfpKMjIz4H5o+KnzP0seG71nVeVdPU6GPPjgBwJgxY/D111/Dw8MDjRs3xqpVq3Dv3j18++23pV0aERERlSFlIjh99dVXSE5OxqxZs5CQkABXV1ccOHAADg4OpV0aERERlSFlIjgBQEBAAAICAkq7DELBUOj06dOLDIcSqSu+Z+ljw/ds6ZEJRa69IyIiIqKPfwFMIiIiog+FwYmIiIhIQQxORERERApicCIiUnNCCIVuBUH0MftY3uMMTkREak4mk8nd1SAvL68UqyEqGa++x3Nzc0uxkrdjcCKVUeSX+d27dwEU3OaGiBSTlJSESZMm4eeffwYAaGpqlnJFRKp37NgxTJ48GQBQrpz6rpbE4ET/WlZWFrp164aJEycCKD5A5eXlYfHixahXrx5ycnKgocG3HpGidHR0cOXKFZw4cQKHDh1Cq1atsHPnztIui0ilIiIicPjwYRw/fhw///wzJk+ejLS0tNIuqwj+9SKlvNpTVDgeraWlhcqVK2PHjh0ACj4N5+TkYNeuXXj58qXU5uTkBHNzc+zZs0dufyL6n7y8vCL/z7S1tZGeno4tW7agb9++qFKlCurXr1+KVRK9PyGE3AfsnJwcAECFChVw48YNtG7dGnPnzoW1tTX09fVLq8w3YnAipRT2FMXHx0vj0RoaGujRowfi4+Nx6tQpAMCePXvwzTffSGEKANzc3ODq6orNmzcDYHAielXh/wdNTU1oaGjg5cuXyM/Ph0wmw7Vr1+Ds7Izq1atj9OjRWLVqFW8pRR+lwvd04XBzTk4OtLS0kJmZibi4OFSsWBGenp64efMmRo4cqZbD0gxOpJSnT5+iatWqqFOnDi5evCi1V69eHQ0aNMCKFSsAAHXq1IG3tzcOHjwobVOhQgV4e3vj/PnzePLkCYfriF4hk8kghMCmTZvQtGlT+Pn5YcKECXjw4AHq16+PlStXomHDhjh16hQyMzNLu1yi96KhoYEnT55g1qxZaNmyJcaPH4+zZ89CV1cXM2fOxNixY5GdnY3Dhw8DUM8LIfiXi5Ry9+5dNG/eHO7u7pg1axYiIyMBAHp6eujZsyf+/PNPAICTkxM8PDxw8eJFPHjwAEDBf5i6deuifPny+P333wFwkjh9enJzc4vtbX369Cm6d++OhQsXwsfHBwMGDMCBAwcwceJEPHz4EJqamnB3d8ejR49w4sQJAPz/Q+opPz//jVfFbd68Gc2aNcORI0fQtm1bXLp0CaNHj5ZGJzw8PKCvr4+jR48CgFp+wFa/ikgtFf6iz87OxoULF7Bx40ZUrVoVgYGByMrKQrly5dC2bVvk5uZKk1br1asHPT09uUmsDg4OyMvLw4EDBwCo538KopJUrlw5yGQyxMfHIzk5WWo3MzNDzZo1sXPnTsyePRtff/01+vfvj/DwcOzduxcA0KxZM+jp6SE0NBSA/OXbROpCQ0NDuiru8ePHcs/p6elh0KBB+PvvvzFhwgRs2bIFWlpa2LhxIwDA3d0d1atXx+XLl5GQkKCW73H+1SKFFL55K1eujJSUFMhkMsydOxdZWVkIDAzE/fv3Ub16dfj6+kqXTLu5ucHNzU2a0wQAUVFRyM/Px/79+3H16tVSORei0iCEQG5uLlavXg1nZ2c0adIEXbp0weLFi/H8+XMAwHfffQcrKytMmzYNVapUwfz586Gnpyf10Hp4eKBWrVoICwvDw4cPIYTAlStX1HrNG/p0FH7AfvDgAQICAlCxYkX07NkTAQEBUoDy8/PDmDFjcPHiRfTu3Ruurq64fv06oqOjERERAQDw8vJCRkaGdCHRkydPkJ6eXjonVQwGJ1LK5cuXUa1aNSQmJqJcuXLYvXs3Hj58iOnTpwMAevXqhdDQUKSmpsLW1hYDBgxAdHQ0unbtim+//RaLFy/G999/j5UrV8Lc3LyUz4ZI9YoLMUIIyGQyHDlyBKtWrUJgYCAiIyPRq1cv7N69G/PnzwdQ8Gn8P//5D44fP46ZM2fi4cOH+Prrr3Hjxg1ERUUBADp06ACZTIZWrVqhXLlymDFjhnT1KtGHkJ+fLzf3qDAwyWQy5OTkYOLEibh37x4WL16MiRMn4uLFixg/fjweP34MXV1dXLx4ESNGjICGhgYOHz6MnTt3wsDAALt37wYA+Pj4oGbNmpgzZw68vLxgbW2N6OjoUjnXYgkiJURGRgpnZ2chhBBhYWGiS5cuQlNTU8hkMrF+/Xpx9epVUaVKFbF48WJpnwMHDogBAwaI9u3bi4MHD5ZW6UQlJi8vr0hbdHS0SEhIkB6npqaKrl27ipkzZ0pt27dvFxYWFqJu3boiPT1d3Lt3T9jY2IilS5dK24wdO1bIZDIxbdo0IYQQubm54tatW2LFihXiypUrJXhWRPJyc3PlHufn54u4uDi5tt27d4tatWqJJ0+eSPv07t1blCtXTmzbtk0IIcSYMWNEjRo1REZGhhBCiFu3bgkLCwvh6ekpcnJyhBAF/18WLFggFi9eLJKSkkr61JTCHidSioWFBW7fvg19fX107twZ5cqVw5kzZzBr1iysWrUKe/fuRadOnTBv3jxpHz8/P/zyyy/Yv38/2rVrV4rVE5WMwrl6z549w86dO2FqaopWrVqhU6dO0twNfX19hIaGwsTEBF9//TVMTEwwefJkBAQEYMuWLTAwMEBiYiKcnZ1x//59AMDZs2dx9+5dtGjRQposW7gm2rfffotatWrxPnb0wRQuDZCQkICvv/4aRkZGaN++PebNmyddBPTXX3+hevXq+Pvvv9G0aVOYmprizp07WLlyJdq3b4+cnBzk5ubC1tZW6rX6/fffUaNGDdy9e1d6nxsZGWHMmDEYMWIELC0t1epCCPVd05zU0vXr11GzZk0MHjwYffv2hampKQDA2dkZmpqamDRpElatWgVbW1vk5+dLf1DUcYIfkbLy8/MhhCiytkxmZiamTp2KK1euwNjYGCtXrkTdunURFBSEwMBAWFpaws/PDx4eHggMDMTQoUPx559/wsvLC9ra2gCA58+fw93dHb6+vli+fDn27t2LuLg4TJ48Gb169UKlSpWK1CP+fwiQSJXy8vIgk8mKXLxz5MgRrFu3Dtra2tDT08ORI0dw4MABrF27FufPn8dvv/2GRo0aoW/fvoiJiUHXrl3xyy+/wMXFBcD/3q8tWrTA3r170aJFCzx58gQVK1bE9OnTUaVKFVSpUkXuNQvXfVKrC4lKtb+LPjqnTp0SRkZGIi0trchz2dnZ4tq1a6VQFdGHd+7cOWlYQQghJkyYIExNTUWfPn2ktqysLNG9e3fRvHlzkZ+fL+bOnSvMzMxEenq63LF27twpgoODpaGL/fv3i19++UU8evRIbrvihgSJSsqzZ8+kITchCqZd1KpVS9jY2IjTp09L7bt37xYymUzExMSIK1euCENDQ7Fp0ya5Y8XFxYlFixaJixcvCiGEiIqKElOnThUbN278IOeiSgxOpJSzZ8+K8uXLi6tXrwohCsa4icqi/Pz8Iu/vuLg40bdvX2FqaiqqV68uWrZsKbZs2SKEEOL48eOiYsWKYsSIEUKI/4WcY8eOCU1NTfH48WPx8uVLUaFCBeHt7S1Wr14twsLCxLfffitq1qwpFi9eLDIzM4vUkZeXx/9nVGJenbdU+O+9e/cKT09PUbFiRdG6dWsxadIkIYQQmZmZolevXsLKykraJz8/X2RlZYlq1aqJWbNmCSGE8Pf3F05OTmLWrFni1q1bYvfu3aJDhw6ic+fO4saNG8XW8TF9KFCjvi/6GAgh0LVrV+jp6QHgEByVPeKVK4RkMhkuXbqEa9euAQCWL1+OxMREHDlyBEePHkW7du0wZswYXLhwAQ0aNEDjxo1x584dvHjxQhpacHR0hImJCU6fPg0dHR3s2LEDLi4u2LhxI7766ivcv38fy5cvx4gRI6CrqyvVUTgsqKGhwf9npFLilTlxhcPOFy9ehKamJq5evYpZs2ahRYsWOHjwIIYMGYLVq1dj0aJF0NXVhbe3N6ysrKS1+MT/30uxadOm0vpiP/30k7QgcseOHTFw4EA4Ozvjv//9L6pVqyZXS+E8J7UainuXUo1tRERqKCMjQ1y6dElMmTJFyGQysWTJEnH79m1hZWUlHj58KIQouBJoxowZQiaTiR9//FEIIURwcLBwdnYWBw4ckI61ZcsWYW9vL86ePSv3Gvfv3/+oPmXTx624XsvHjx+LkydPCicnJ+lq6c6dO4thw4ZJ29y5c0dYWFiIhg0bioSEBHH58mXRtGlT4e/vL23z/Plz0bBhQzFmzBi5HqyEhARpdKIs4eRwIqJXPHjwAIMGDUJCQgLc3NykG49GRETA0NAQa9euxY4dO3Dnzh14eXlh27ZtaN++PQDA19cXq1evRkBAAIYPHw4XFxcsXboUnp6e8PDwkF5DCIEKFSoA+N8nbnW8mSl9/PLy8qCpqVmk1zI4OBjbt2+HlZUVxo0bhy+//BI5OTnSRT1z5szBkiVLAACdOnXCkCFDYGNjAxsbG7i6umL16tWwtbXFF198gePHjyMhIQFdunSBpqamdIzC7YGC9c2Kq+NjxOBERJ8MocBVaBUqVECtWrUQHh6OgIAA6Wq2Fy9eQFNTEz///DOmTp2Ktm3bSs9lZWUhNzcXtWvXRpMmTRAREYFbt27h999/R61atTBz5ky513+1BgYmUrVX3+eampoQQuDo0aOoWrUqHB0dARQMIaekpCAnJwcDBgyAlpYWUlNTYWhoiJUrV6J169ZYsmQJ2rRpAyMjIwBARkYG9PX14e3tjaNHjyIyMhI3btzAmTNnMGXKFDRu3FgaXn5d4S1YyoKycyZERO/wrtBU+EnZ19cX27dvl26FAgA1atRAjRo18PLlSwwePFhqT0tLw7Rp09ClSxf4+PjAw8MD//zzD7y9vbFy5UqlXp9IFQrfZ0lJSTh16hR69OgBY2Nj2Nraol+/fhg1ahSaNWuG6tWr48GDB9DS0gIAGBsbo0aNGnB2dsYPP/wAd3d36Zh79uzBtWvXMHHiRHh4eMDFxQUVKlRAcHCw3Ny8T8FHNBuLiOjNwsPDAUBaEPL1BfNyc3OxceNGaaJ3cQvqFf7BadCgAerVq4fz589Lz1WsWBH9+vXD+fPn0bVrV/z2229Ys2YNOnfujIiICOl4vr6+0NbWxqlTpwBAGv4gUrW8vDy5W58UevHiBUaPHg03NzeEhYXh8OHDOH/+PDw8PDB37lwkJSXBwsICnp6e0NLSkt6rANClSxfY29ujd+/e2LhxI6KjozFr1izMmjULSUlJyM7OhpOTE1xcXBATE4NHjx4BKOh1/VQwOBHRR+/AgQPw9fXF/fv3paGw14cLbt68iaCgICxcuBAAil1tuzA4WVlZoW7duoiLi8PFixel5z///HOsXbsWBgYGmDt3LubPnw9fX18cPHgQLVq0AABUq1YNjo6OOHbsGGJjY6GlpfVxXTFEHw1NTU1oamoiJSUFp0+flkKUrq4u6tati5cvXyIjIwNNmjSBnZ0dJk2aBD09PSxbtgwApAVYjxw5Ih3T1dUV69evR61atbBixQq0bdsW+/btw9ixYxEcHCwt2NqkSRM8fvwYO3bsAACp/VPAoToi+mi8uhr9q+rVq4cqVapg69atmDBhAtLT07Fy5Uo8ffoUQUFBAArmdHz55Zf4448/ALx5blHhazRv3hyhoaE4duwY6tatK02y7dSpEzp16oTHjx/D0tJS2q+wl0tTUxOff/45mjVrBjs7uxL4LtCn5k1z86KiojBu3DhERESgQoUKsLW1xfz589GoUSM0atQIFhYWMDU1hYaGBoQQsLW1RefOnbFhwwbMnDkTnp6ecHR0xPnz55GZmQk9PT3k5eXBxsYGv/32G+7duwcjIyOYmJhIr1n4/8DLywtffvklmjZtCuDTGobmxyAi+mhoaGggOTkZycnJAAqG34CCHqLPPvsMv/zyC4CCX+K3b9/GgQMHpG10dXXRsGFDZGRkYP/+/QBQ7DBHYTDz9PSEsbEx9uzZg+zsbLmgJYSApaUlhBDIzc2V/rAVbtOmTRv06tULOjo6JfSdoE9B4bDz66GkcOj3p59+Qvny5XHjxg2sW7cOhoaGGDhwIKKjo+Hs7Iz27dtj165dAAr+T5QvXx4dO3bEvXv3cOnSJejr68Pd3R1Xr16Vep1efZ9XqlQJJiYmyM/PL3L1p6WlJWbMmIFGjRqV+PdB3TA4EdFHY+XKlbC0tERAQACA/12po6GhgW7duuH27duIioqCgYEBWrVqhby8PBw8eFDa39XVFa6urti8eTOAN39KFkJAX18f/fr1w9SpU4sMQxTuJ5PJUK5cuU/q0zaVPPHaIqwxMTHYvHkzbt++DaDg/R4ZGYkjR45g4MCBsLe3R9OmTaX39a5duyCTydC+fXskJibi7Nmz0rFdXV3h7u4u9cT6+PhgwIABcHNze2M9GhoavPrzFQxORPRRyMvLw+PHjzFq1CicOnUKy5YtQ2pqqvR8rVq14O7ujlWrVgEA3NzcULlyZWkOBlDwCdrV1RURERF4/vx5scN+hb1IANCtWze0atWqhM+MSJ5MJkNmZiYuX76MqVOnon79+pgzZw46dOiAP//8EwCgp6eHpKQkNGvWDEBB76upqSkaNmyI8+fP4+nTp6hTp440ZwkoeG+bmZmha9euePnyJYCCYe6JEydKyxTQuzE4EZHaef0qNCEENDU1cfr0aTg7O2PdunX4888/sWjRImkbIyMjfPXVV/j9998BAM7OzvD09MS5c+eQnp4O4H9DffHx8dizZw+A/w33FV6hJJPJpMuzX758ySvi6IN78OABunXrhr59+yIpKQm3b99GWFgYqlatihUrVuDhw4ews7ODhYUF9u3bB+B/w85NmzbF5cuXYWJiAlNTU7Rq1UpaFqOwh3TSpElSACvE97niGJyISG0U/vJ+vSeosN3FxQXnzp1DixYtMGzYMCxevBh79+6V9mnfvj3S09Nx8OBBaGpqomnTpsjPz8eUKVOQk5OD6OhoPH/+HO7u7tiwYYPca7x6hdKCBQtQsWJFLF26lH9QSKWKu5rzdYWLsN6+fRteXl6ws7ODlZUVAgICkJKSgoMHD8LS0hLt2rXDTz/9BADSfLozZ87A1NQUQMG8vnbt2mH69OnIyMiQjl+4KOar721e+ak4fqeISG1oaGggJSUF69atk3qEgIJf9Lm5ucjIyICVlRWAgvVmvv/+e8ycORNHjx4FAFSpUgXNmjWThuuaNGmC0aNHY+vWrXB1dYW3tzfq1KmDP//8U7pJaeE8qV9//RXe3t6oVKkSdu7ciTFjxqB///5lasVjKn2KLMIKFMw9srS0lHpLAaB+/fpwdHSU5u1NnToViYmJaNWqFX777Tf88MMPOHz4MGbMmCEFocaNG2PatGnQ19cvUgfD0vvhbwQi+qDedGl14SXR3bp1g42NDZKSkvDXX38hKCgIhoaGKFeuHOLj46XVuxcuXIglS5YgMTERU6ZMgYWFBerUqYNevXph5MiR0uXV33zzDdzd3XHjxg20bt1aundWoZCQELRt2xY1atTA559/jhUrVqBmzZof6ttBZUh4eDi8vb2lXqXXbz+Sm5uLrVu3wsPDAzVr1ix2eY1XF2GtXbs2zpw5Iz1na2uL+vXr47fffsOFCxdQr149bNu2DTt27MD06dOhp6eHuXPn4rPPPpM7ZuHVeQxKKvJBbiVMRJ+83NzcYu/QLoQQCxYsEGZmZqJdu3bi0KFDUpu7u7tYuXKltN3QoUOFTCYT2trawt3dXfz3v/8VMTExol27dqJVq1bi4cOH4t69e0Imk4lt27YV+1p5eXkiPz9fquXFixciLCxM5OTkqPiM6VOyf/9+IZPJRHx8/Bu3uXr1qnBxcRGDBg0SQhT8n3ib6dOnCy8vL3HhwgWp7e+//xYNGjQQ48ePl9ry8/NFenr6vzwDUhTjJxF9EIV3Rr916xa2bt2K+/fvSxNaa9WqhYoVKyI3Nxdt27YFAPTr1w/16tWTLrHOycnBkydP0LlzZ0RGRuL8+fMYOXIkXFxcMG/ePNy+fRvjx4+HlZUVrl27hh49esi9vnjlU/erN9otX748vL29OSRHCnnTnLdXF2EFgPT0dPz444+YNGmStE3hIqwREREA3r4IKwA0b94c5cqVw7Fjx6TnateujaZNm8LZ2VluHwMDA7n1lqjkMDgRkUq96Zd3TEwMfHx8UKdOHcyePRstW7bEtGnTAEC6aejTp0+l7c3NzdG8eXPExsbi0qVL0NLSwuPHj+Hg4IB69erJTbKtXbs2QkNDsX79eujo6MDFxaXI678alojeV2kuwgoU3Ig3ODhY7kbThe9rrrf0YTA4EdG/Jl65QufVX945OTkACsLU2rVrkZubi7i4OBw9ehQjR45EUFAQ9u/fD3Nzc/j6+iIjIwNRUVHScd3d3WFvb4/ffvsNAGBoaIhbt25Jr/OqypUrs9eISpy6LMLKqz1LD4MTEf1rr16hc/PmTXTr1g0eHh4YP348nj17htzcXKxduxaDBg2ClZUVbGxsMGzYMLRo0QJr165FZmYmGjVqBDMzM+lecgBQtWpVeHp6YuPGjQAKglO3bt3eOMGcqCSp0yKsnOhdevidJ6J/TQiBtWvXYuHChQgODoaZmRkGDBiANWvWYOLEiYiJiYGFhYX0Kblw1eKBAwfi5MmTSElJgZOTEzw8PBAWFiYd18DAAD4+PujUqRMyMzOxZcsWDBgwgKGJShwXYaU3YXAion9NJpMhKioK06ZNQ1ZWFpYvX46AgAAsXrwYV69exV9//QVvb29plePCYYfGjRsjOTkZL168gIGBAerUqSNtX6hLly5Yvnw59PT0ABT8cREKLCJI9D64CCu9C4MTEanEyJEjoaenB2dnZ+nTcocOHWBgYICbN2+ibt26CA8Px/Xr16U5UGvWrEGNGjWk7b29vbFq1So0btxY7tivzqEqvDqPqCRwEVZ6F/40iEglqlSpAmdnZyQkJCA7Oxva2tqwsrKCq6sroqOj0aNHD7Ru3Rq+vr7o3bs3hBDYvn07vvvuOzg4OAAAnJyc4OTkVOTYvCKOVOlNc+S4CCspgj1ORKQSGhoa6NmzJ27cuIHr169L7W3btkV6ejru3r2LzZs3Y/Lkybh06RKuX7+OVatWYcSIEUWOxaE4KgmFw7zFhabg4GC0adMGc+bMwaZNmxAdHY0JEybg5MmT0tpMQMHk7sDAQBgbG+O3337DxIkTce3aNZiYmGDcuHFISEhAq1atkJGRIddj1bBhQ3z99dewsbFBfn6+tK4YADRr1gyhoaG4dOkSZs+ezdCk5mSCv6GISEUePnyIjh07YujQodI6My9fvpQ+ZW/cuBG6urrF3mqC6EO5desWzp49i+bNm8PW1haampo4fPgwJkyYACsrK4SEhAAAkpOTMWHCBNy8eRPHjx9HTk4OevfujZycHEydOhXu7u7S+zg6OhqdO3dGkyZNsGbNGsTGxhZZT6zwzy17Tz9u/M1FRCpjZ2cHe3t7nDx5EmlpaQAKVub+6aefsGrVKrnQVHgFEZGqcRFWKkkMTkSkUj169JBbWRkoWAHZ1NQUwP+uViq8gohIFbgIK30oDE5EpFLdu3fH3r174ejoWNql0CeEi7DSh8JoTEQqVfhJn/OY6EMSQmDdunVITU3FjRs3YGZmhpYtW+K7777Dy5cvMXTo0CKLsOrq6mLgwIEYPXr0Oxdhzc7OlhZhpU8bf6sRUYlgaKIPiYuw0ofC32xERFQmcBFW+hA4VEdERGUCF2GlD4E9TkREVCZwEVb6ELgAJhERlRlchJVKGoMTERGVKZ07d4apqSkWL14MIyMjAMDp06dRrVo1mJqayi3CCoDriZFSGLeJiKhM4SKsVJLY40RERGVKXl4eAxGVGPY4ERFRmfLqIqxEqsYeJyIiIiIFsceJiIiISEEMTkREREQKYnAiIiIiUhCDExEREZGCGJyIiIiIFMTgRERERKQgBiciIiIiBTE4ERERESmIwYmIiIhIQQxORERERAr6P+LE1cFkHB8EAAAAAElFTkSuQmCC",
            "text/plain": [
              "<Figure size 600x400 with 1 Axes>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "#Plot2 - Editing site counts bar plot\n",
        "\n",
        "wt_counts = count_plot[count_plot['condition'] == 'WT']['site_count'].values\n",
        "ko_counts = count_plot[count_plot['condition'] == 'KO']['site_count'].values\n",
        "\n",
        "plt.figure(figsize=(6, 4))\n",
        "plt.bar(x - width/2, wt_counts, width, label='WT')\n",
        "plt.bar(x + width/2, ko_counts, width, label='KO')\n",
        "\n",
        "plt.xticks(x, regions, rotation=20)\n",
        "plt.ylabel('Number of editing sites')\n",
        "plt.title('Editing site counts by region')\n",
        "plt.legend()\n",
        "plt.tight_layout()\n",
        "plt.show()"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "General",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.18"
    },
    "colab": {
      "provenance": [],
      "toc_visible": true
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}